// @metamask/chain-agnostic-permission/dist/operators/caip-permission-operator-permittedChains.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setChainIdsInCaip25CaveatValue = exports.addCaipChainIdInCaip25CaveatValue = exports.getAllScopesFromPermission = exports.getAllNamespacesFromCaip25CaveatValue = exports.getAllScopesFromCaip25CaveatValue = exports.getAllScopesFromScopesObjects = exports.setPermittedEthChainIds = exports.addPermittedEthChainId = exports.getPermittedEthChainIds = void 0;
const controller_utils_1 = require("@metamask/controller-utils");
const utils_1 = require("@metamask/utils");
const caip25Permission_1 = require("../caip25Permission.cjs");
const transform_1 = require("../scope/transform.cjs");
const types_1 = require("../scope/types.cjs");
/*
 *
 *
 * EVM SPECIFIC GETTERS AND SETTERS
 *
 *
 */
/**
 * Gets the Ethereum (EIP155 namespaced) chainIDs from internal scopes.
 *
 * @param scopes - The internal scopes from which to get the Ethereum chainIDs.
 * @returns An array of Ethereum chainIDs.
 */
const getPermittedEthChainIdsFromScopes = (scopes) => {
    const ethChainIds = [];
    Object.keys(scopes).forEach((scopeString) => {
        const { namespace, reference } = (0, types_1.parseScopeString)(scopeString);
        if (namespace === utils_1.KnownCaipNamespace.Eip155 && reference) {
            ethChainIds.push((0, controller_utils_1.toHex)(reference));
        }
    });
    return ethChainIds;
};
/**
 * Gets the Ethereum (EIP155 namespaced) chainIDs from the required and optional scopes.
 *
 * @param caip25CaveatValue - The CAIP-25 caveat value from which to get the Ethereum chainIDs.
 * @returns An array of Ethereum chainIDs.
 */
const getPermittedEthChainIds = (caip25CaveatValue) => {
    const { requiredScopes, optionalScopes } = caip25CaveatValue;
    const ethChainIds = [
        ...getPermittedEthChainIdsFromScopes(requiredScopes),
        ...getPermittedEthChainIdsFromScopes(optionalScopes),
    ];
    return (0, transform_1.getUniqueArrayItems)(ethChainIds);
};
exports.getPermittedEthChainIds = getPermittedEthChainIds;
/**
 * Adds an Ethereum (EIP155 namespaced) chainID to the optional scopes if it is not already present
 * in either the pre-existing required or optional scopes.
 *
 * @param caip25CaveatValue - The CAIP-25 caveat value to add the Ethereum chainID to.
 * @param chainId - The Ethereum chainID to add.
 * @returns The updated CAIP-25 caveat value with the added Ethereum chainID.
 */
const addPermittedEthChainId = (caip25CaveatValue, chainId) => {
    const scopeString = `eip155:${(0, utils_1.hexToBigInt)(chainId).toString(10)}`;
    if (Object.keys(caip25CaveatValue.requiredScopes).includes(scopeString) ||
        Object.keys(caip25CaveatValue.optionalScopes).includes(scopeString)) {
        return caip25CaveatValue;
    }
    return {
        ...caip25CaveatValue,
        optionalScopes: {
            ...caip25CaveatValue.optionalScopes,
            [scopeString]: {
                accounts: [],
            },
        },
    };
};
exports.addPermittedEthChainId = addPermittedEthChainId;
/**
 * Filters the scopes object to only include:
 * - Scopes without references (e.g. "wallet:")
 * - EIP155 scopes for the given chainIDs
 * - Non EIP155 scopes (e.g. "bip122:" or any other non ethereum namespaces)
 *
 * @param scopesObject - The scopes object to filter.
 * @param chainIds - The chainIDs to filter EIP155 scopes by.
 * @returns The filtered scopes object.
 */
const filterEthScopesObjectByChainId = (scopesObject, chainIds) => {
    const updatedScopesObject = {};
    Object.entries(scopesObject).forEach(([key, scopeObject]) => {
        // Cast needed because index type is returned as `string` by `Object.entries`
        const scopeString = key;
        const { namespace, reference } = (0, types_1.parseScopeString)(scopeString);
        if (!reference) {
            updatedScopesObject[scopeString] = scopeObject;
            return;
        }
        if (namespace === utils_1.KnownCaipNamespace.Eip155) {
            const chainId = (0, controller_utils_1.toHex)(reference);
            if (chainIds.includes(chainId)) {
                updatedScopesObject[scopeString] = scopeObject;
            }
        }
        else {
            updatedScopesObject[scopeString] = scopeObject;
        }
    });
    return updatedScopesObject;
};
/**
 * Sets the permitted Ethereum (EIP155 namespaced) chainIDs for the required and optional scopes.
 *
 * @param caip25CaveatValue - The CAIP-25 caveat value to set the permitted Ethereum chainIDs for.
 * @param chainIds - The Ethereum chainIDs to set as permitted.
 * @returns The updated CAIP-25 caveat value with the permitted Ethereum chainIDs.
 */
const setPermittedEthChainIds = (caip25CaveatValue, chainIds) => {
    let updatedCaveatValue = {
        ...caip25CaveatValue,
        requiredScopes: filterEthScopesObjectByChainId(caip25CaveatValue.requiredScopes, chainIds),
        optionalScopes: filterEthScopesObjectByChainId(caip25CaveatValue.optionalScopes, chainIds),
    };
    chainIds.forEach((chainId) => {
        updatedCaveatValue = (0, exports.addPermittedEthChainId)(updatedCaveatValue, chainId);
    });
    return updatedCaveatValue;
};
exports.setPermittedEthChainIds = setPermittedEthChainIds;
/*
 *
 *
 * GENERALIZED GETTERS AND SETTERS
 *
 *
 */
/*
 *
 * GETTERS
 *
 */
/**
 * Gets all scopes from a CAIP-25 caveat value
 *
 * @param scopesObjects - The scopes objects to get the scopes from.
 * @returns An array of InternalScopeStrings.
 */
function getAllScopesFromScopesObjects(scopesObjects) {
    const scopeSet = new Set();
    for (const scopeObject of scopesObjects) {
        for (const key of Object.keys(scopeObject)) {
            scopeSet.add(key);
        }
    }
    return Array.from(scopeSet);
}
exports.getAllScopesFromScopesObjects = getAllScopesFromScopesObjects;
/**
 * Gets all scopes (chain IDs) from a CAIP-25 caveat
 * This extracts all scopes from both required and optional scopes
 * and returns a unique set.
 *
 * @param caip25CaveatValue - The CAIP-25 caveat value to extract scopes from
 * @returns Array of unique scope strings (chain IDs)
 */
function getAllScopesFromCaip25CaveatValue(caip25CaveatValue) {
    return getAllScopesFromScopesObjects([
        caip25CaveatValue.requiredScopes,
        caip25CaveatValue.optionalScopes,
    ]);
}
exports.getAllScopesFromCaip25CaveatValue = getAllScopesFromCaip25CaveatValue;
/**
 * Gets all non-wallet namespaces from a CAIP-25 caveat value
 * This extracts all namespaces from both required and optional scopes
 * and returns a unique set.
 *
 * @param caip25CaveatValue - The CAIP-25 caveat value to extract namespaces from
 * @returns Array of unique namespace strings
 */
function getAllNamespacesFromCaip25CaveatValue(caip25CaveatValue) {
    const allScopes = getAllScopesFromCaip25CaveatValue(caip25CaveatValue);
    const namespaceSet = new Set();
    for (const scope of allScopes) {
        const { namespace, reference } = (0, types_1.parseScopeString)(scope);
        if (namespace === utils_1.KnownCaipNamespace.Wallet) {
            namespaceSet.add(reference ?? namespace);
        }
        else if (namespace) {
            namespaceSet.add(namespace);
        }
    }
    return Array.from(namespaceSet);
}
exports.getAllNamespacesFromCaip25CaveatValue = getAllNamespacesFromCaip25CaveatValue;
/**
 * Gets all scopes (chain IDs) from a CAIP-25 permission
 * This extracts all scopes from both required and optional scopes
 * and returns a unique set.
 *
 * @param caip25Permission - The CAIP-25 permission object
 * @param caip25Permission.caveats - The caveats of the CAIP-25 permission
 * @returns Array of unique scope strings (chain IDs)
 */
function getAllScopesFromPermission(caip25Permission) {
    const caip25Caveat = caip25Permission.caveats.find((caveat) => caveat.type === caip25Permission_1.Caip25CaveatType);
    if (!caip25Caveat) {
        return [];
    }
    return getAllScopesFromCaip25CaveatValue(caip25Caveat.value);
}
exports.getAllScopesFromPermission = getAllScopesFromPermission;
/*
 *
 * SETTERS
 *
 */
/**
 * Adds a chainID to the optional scopes if it is not already present
 * in either the pre-existing required or optional scopes.
 *
 * @param caip25CaveatValue - The CAIP-25 caveat value to add the chainID to.
 * @param chainId - The chainID to add.
 * @returns The updated CAIP-25 caveat value with the added chainID.
 */
const addCaipChainIdInCaip25CaveatValue = (caip25CaveatValue, chainId) => {
    if (caip25CaveatValue.requiredScopes[chainId] ||
        caip25CaveatValue.optionalScopes[chainId]) {
        return caip25CaveatValue;
    }
    return {
        ...caip25CaveatValue,
        optionalScopes: {
            ...caip25CaveatValue.optionalScopes,
            [chainId]: {
                accounts: [],
            },
        },
    };
};
exports.addCaipChainIdInCaip25CaveatValue = addCaipChainIdInCaip25CaveatValue;
/**
 * Sets the CAIP-2 chainIds for the required and optional scopes.
 * If the caip25CaveatValue contains chainIds not in the chainIds array arg they are filtered out
 *
 * @param caip25CaveatValue - The CAIP-25 caveat value to set the permitted CAIP-2 chainIDs for.
 * @param chainIds - The CAIP-2 chainIDs to set.
 * @returns The updated CAIP-25 caveat value with the CAIP-2 chainIDs.
 */
const setChainIdsInCaip25CaveatValue = (caip25CaveatValue, chainIds) => {
    const chainIdSet = new Set(chainIds);
    const result = {
        requiredScopes: {},
        optionalScopes: {},
        sessionProperties: caip25CaveatValue.sessionProperties,
        isMultichainOrigin: caip25CaveatValue.isMultichainOrigin,
    };
    for (const [key, value] of Object.entries(caip25CaveatValue.requiredScopes)) {
        const scopeString = key;
        if ((0, types_1.isWalletScope)(scopeString) || chainIdSet.has(scopeString)) {
            result.requiredScopes[scopeString] = value;
        }
    }
    for (const [key, value] of Object.entries(caip25CaveatValue.optionalScopes)) {
        const scopeString = key;
        if ((0, types_1.isWalletScope)(scopeString) || chainIdSet.has(scopeString)) {
            result.optionalScopes[scopeString] = value;
        }
    }
    for (const chainId of chainIds) {
        if (!result.requiredScopes[chainId] && !result.optionalScopes[chainId]) {
            result.optionalScopes[chainId] = { accounts: [] };
        }
    }
    return result;
};
exports.setChainIdsInCaip25CaveatValue = setChainIdsInCaip25CaveatValue;
