// @metamask/connect-multichain/dist/src/multichain/transports/mwp/index.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { SessionStore } from '@metamask/mobile-wallet-protocol-core';
import { TransportTimeoutError, } from '@metamask/multichain-api-client';
import { JsonRpcError, providerErrors, rpcErrors } from '@metamask/rpc-errors';
import { createLogger, } from '../../../domain';
import { addValidAccounts, getOptionalScopes, getUniqueRequestId, getValidAccounts, isSameScopesAndAccounts, } from '../../utils';
import { MULTICHAIN_PROVIDER_STREAM_NAME } from '../constants';
const DEFAULT_REQUEST_TIMEOUT = 60 * 1000;
const CONNECTION_GRACE_PERIOD = 60 * 1000;
const DEFAULT_CONNECTION_TIMEOUT = DEFAULT_REQUEST_TIMEOUT + CONNECTION_GRACE_PERIOD;
const DEFAULT_RESUME_TIMEOUT = 10 * 1000;
const SESSION_STORE_KEY = 'cache_wallet_getSession';
const ACCOUNTS_STORE_KEY = 'cache_eth_accounts';
const CHAIN_STORE_KEY = 'cache_eth_chainId';
const PENDING_SESSION_REQUEST_KEY = 'pending_session_request';
const CACHED_METHOD_LIST = [
    'wallet_getSession',
    'wallet_createSession',
    'wallet_sessionChanged',
];
const CACHED_RESET_METHOD_LIST = [
    'wallet_revokeSession',
    'wallet_revokePermissions',
];
const logger = createLogger('metamask-sdk:transport');
/**
 * Mobile Wallet Protocol transport implementation
 * Bridges the MWP DappClient with the multichain API client Transport interface
 */
export class MWPTransport {
    get pendingRequests() {
        return this.__pendingRequests;
    }
    set pendingRequests(pendingRequests) {
        this.__pendingRequests = pendingRequests;
    }
    get sessionRequest() {
        return this.currentSessionRequest;
    }
    constructor(dappClient, kvstore, options = {
        requestTimeout: DEFAULT_REQUEST_TIMEOUT,
        connectionTimeout: DEFAULT_CONNECTION_TIMEOUT,
        resumeTimeout: DEFAULT_RESUME_TIMEOUT,
    }) {
        this.dappClient = dappClient;
        this.kvstore = kvstore;
        this.options = options;
        this.__pendingRequests = new Map();
        this.notificationCallbacks = new Set();
        this.dappClient.on('message', this.handleMessage.bind(this));
        // We store the pending session request in the KVStore so that we can:
        // 1. Use it to regenerate the deeplink to re-prompt the user with if they
        // attempt to connect while there is a pending connection attempt.
        // 2. Determine if there was a pending connection attempt when MultichainConnect
        // is initializing itself for the first time on page load and correctly set
        // the appropriate timeout duration for that connection attempt.
        this.dappClient.on('session_request', (sessionRequest) => {
            this.currentSessionRequest = sessionRequest;
            this.kvstore
                .set(PENDING_SESSION_REQUEST_KEY, JSON.stringify(sessionRequest))
                .catch((err) => {
                logger('Failed to store pending session request', err);
            });
        });
        if (typeof window !== 'undefined' &&
            typeof window.addEventListener !== 'undefined') {
            this.windowFocusHandler = this.onWindowFocus.bind(this);
            window.addEventListener('focus', this.windowFocusHandler);
        }
    }
    /**
     * Returns the stored pending session request from the dappClient session_request event, if any.
     *
     * @returns The stored SessionRequest, or null if none or invalid.
     */
    getStoredPendingSessionRequest() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const raw = yield this.kvstore.get(PENDING_SESSION_REQUEST_KEY);
                if (!raw) {
                    return null;
                }
                return JSON.parse(raw);
            }
            catch (_a) {
                return null;
            }
        });
    }
    /**
     * Removes the stored pending session request from the KVStore.
     * This is necessary to ensure that ConnectMultichain is able to correctly
     * infer the MWP Transport connection attempt status.
     */
    removeStoredPendingSessionRequest() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.kvstore.delete(PENDING_SESSION_REQUEST_KEY);
        });
    }
    onWindowFocus() {
        if (!this.isConnected()) {
            this.dappClient.reconnect();
        }
    }
    notifyCallbacks(data) {
        this.notificationCallbacks.forEach((callback) => callback(data));
    }
    rejectRequest(id, error = new Error('Request rejected')) {
        const request = this.pendingRequests.get(id);
        if (request) {
            this.pendingRequests.delete(id);
            clearTimeout(request.timeout);
            request.reject(error);
        }
    }
    parseWalletError(errorPayload) {
        const errorData = errorPayload;
        if (typeof errorData.code === 'number' &&
            typeof errorData.message === 'string') {
            const { code, message } = errorData;
            if (code >= 1000 && code <= 4999) {
                return providerErrors.custom({ code, message });
            }
            return new JsonRpcError(code, message);
        }
        const message = errorPayload instanceof Error
            ? errorPayload.message
            : JSON.stringify(errorPayload);
        return rpcErrors.internal({ message });
    }
    handleMessage(message) {
        if (typeof message === 'object' && message !== null) {
            if ('data' in message) {
                const messagePayload = message.data;
                if ('id' in messagePayload && typeof messagePayload.id === 'string') {
                    const request = this.pendingRequests.get(messagePayload.id);
                    if (request) {
                        clearTimeout(request.timeout);
                        // Check if the message contains an error (e.g., user rejected)
                        if ('error' in messagePayload && messagePayload.error) {
                            this.pendingRequests.delete(messagePayload.id);
                            request.reject(this.parseWalletError(messagePayload.error));
                            return;
                        }
                        // Success case - resolve the promise
                        const requestWithName = Object.assign(Object.assign({}, messagePayload), { method: request.method === 'wallet_getSession' ||
                                request.method === 'wallet_createSession'
                                ? 'wallet_sessionChanged'
                                : request.method });
                        const notification = Object.assign(Object.assign({}, messagePayload), { method: request.method === 'wallet_getSession' ||
                                request.method === 'wallet_createSession'
                                ? 'wallet_sessionChanged'
                                : request.method, params: requestWithName.result });
                        this.notifyCallbacks(notification);
                        request.resolve(requestWithName);
                        this.pendingRequests.delete(messagePayload.id);
                    }
                }
                else {
                    if (message.data.method ===
                        'metamask_chainChanged') {
                        this.kvstore.set(CHAIN_STORE_KEY, JSON.stringify(message.data.params
                            .chainId));
                    }
                    if (message.data.method ===
                        'metamask_accountsChanged') {
                        this.kvstore.set(ACCOUNTS_STORE_KEY, JSON.stringify(message.data.params));
                    }
                    // Ensure session changes are always persisted to the store
                    if (message.data.method ===
                        'wallet_sessionChanged') {
                        const notification = message.data;
                        const response = {
                            result: notification.params,
                        };
                        this.kvstore.set(SESSION_STORE_KEY, JSON.stringify(response));
                    }
                    this.notifyCallbacks(message.data);
                }
            }
        }
    }
    onResumeSuccess(resumeResolve, resumeReject, options) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b, _c, _d, _e, _f, _g;
            try {
                yield this.waitForWalletSessionIfNotCached();
                const sessionRequest = yield this.request({
                    method: 'wallet_getSession',
                });
                // TODO: verify if this branching logic can ever be hit
                if (sessionRequest.error) {
                    return resumeReject(new Error(sessionRequest.error.message));
                }
                let walletSession = sessionRequest.result;
                if (walletSession && options) {
                    const currentScopes = Object.keys((_a = walletSession === null || walletSession === void 0 ? void 0 : walletSession.sessionScopes) !== null && _a !== void 0 ? _a : {});
                    const proposedScopes = (_b = options === null || options === void 0 ? void 0 : options.scopes) !== null && _b !== void 0 ? _b : [];
                    const proposedCaipAccountIds = (_c = options === null || options === void 0 ? void 0 : options.caipAccountIds) !== null && _c !== void 0 ? _c : [];
                    const hasSameScopesAndAccounts = isSameScopesAndAccounts(currentScopes, proposedScopes, walletSession, proposedCaipAccountIds);
                    if (options.forceRequest || !hasSameScopesAndAccounts) {
                        const optionalScopes = addValidAccounts(getOptionalScopes((_d = options === null || options === void 0 ? void 0 : options.scopes) !== null && _d !== void 0 ? _d : []), getValidAccounts((_e = options === null || options === void 0 ? void 0 : options.caipAccountIds) !== null && _e !== void 0 ? _e : []));
                        const sessionRequest = {
                            optionalScopes,
                        };
                        const response = yield this.request({
                            method: 'wallet_createSession',
                            params: sessionRequest,
                        });
                        if (response.error) {
                            return resumeReject(new Error(response.error.message));
                        }
                        // TODO: Maybe find a better way to revoke sessions on wallet without triggering an empty notification
                        // Issue of this is it will send a session update event with an empty session and right after we may get the session recovered
                        // await this.request({ method: 'wallet_revokeSession', params: walletSession });
                        walletSession = response.result;
                    }
                }
                else if (!walletSession) {
                    // TODO: verify if this branching logic can ever be hit
                    const optionalScopes = addValidAccounts(getOptionalScopes((_f = options === null || options === void 0 ? void 0 : options.scopes) !== null && _f !== void 0 ? _f : []), getValidAccounts((_g = options === null || options === void 0 ? void 0 : options.caipAccountIds) !== null && _g !== void 0 ? _g : []));
                    const sessionRequest = { optionalScopes };
                    const response = yield this.request({
                        method: 'wallet_createSession',
                        params: sessionRequest,
                    });
                    if (response.error) {
                        return resumeReject(new Error(response.error.message));
                    }
                    walletSession = response.result;
                }
                yield this.removeStoredPendingSessionRequest();
                this.notifyCallbacks({
                    method: 'wallet_sessionChanged',
                    params: walletSession,
                });
                return resumeResolve();
            }
            catch (err) {
                return resumeReject(err);
            }
        });
    }
    init() {
        return __awaiter(this, void 0, void 0, function* () {
            // no-op for MWP — passive init is only relevant for DefaultTransport
        });
    }
    // TODO: Rename this
    sendEip1193Message(payload, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = Object.assign({ jsonrpc: '2.0', id: String(getUniqueRequestId()) }, payload);
            const cachedWalletSession = yield this.getCachedResponse(request);
            if (cachedWalletSession) {
                this.notifyCallbacks(cachedWalletSession);
                return cachedWalletSession;
            }
            return new Promise((resolve, reject) => {
                var _a;
                const timeout = setTimeout(() => {
                    this.rejectRequest(request.id, new TransportTimeoutError());
                }, (_a = options === null || options === void 0 ? void 0 : options.timeout) !== null && _a !== void 0 ? _a : this.options.requestTimeout);
                this.pendingRequests.set(request.id, {
                    request,
                    method: request.method,
                    resolve: (response) => __awaiter(this, void 0, void 0, function* () {
                        yield this.storeWalletSession(request, response);
                        return resolve(response);
                    }),
                    reject,
                    timeout,
                });
                this.dappClient
                    .sendRequest({
                    name: 'metamask-provider',
                    data: request,
                })
                    .catch(reject);
            });
        });
    }
    connect(options) {
        return __awaiter(this, void 0, void 0, function* () {
            const { dappClient } = this;
            const session = yield this.getActiveSession();
            if (session) {
                logger('active session found', {
                    id: session.id,
                    channel: session.channel,
                    expiresAt: session.expiresAt,
                });
            }
            const storedSessionRequestBeforeConnectionAttempt = yield this.getStoredPendingSessionRequest();
            let timeout;
            let initialConnectionMessageHandler;
            const connectionPromise = new Promise((resolve, reject) => __awaiter(this, void 0, void 0, function* () {
                let connection;
                if (session) {
                    connection = new Promise((resumeResolve, resumeReject) => {
                        var _a;
                        if (this.dappClient.state === 'CONNECTED') {
                            this.onResumeSuccess(resumeResolve, resumeReject, options);
                        }
                        else {
                            this.dappClient.once('connected', () => __awaiter(this, void 0, void 0, function* () {
                                this.onResumeSuccess(resumeResolve, resumeReject, options);
                            }));
                            dappClient.resume((_a = session === null || session === void 0 ? void 0 : session.id) !== null && _a !== void 0 ? _a : '');
                        }
                    });
                }
                else {
                    connection = new Promise((resolveConnection, rejectConnection) => {
                        var _a, _b;
                        const optionalScopes = addValidAccounts(getOptionalScopes((_a = options === null || options === void 0 ? void 0 : options.scopes) !== null && _a !== void 0 ? _a : []), getValidAccounts((_b = options === null || options === void 0 ? void 0 : options.caipAccountIds) !== null && _b !== void 0 ? _b : []));
                        const sessionRequest = {
                            optionalScopes,
                            sessionProperties: options === null || options === void 0 ? void 0 : options.sessionProperties,
                        };
                        const request = {
                            jsonrpc: '2.0',
                            id: String(getUniqueRequestId()),
                            method: 'wallet_createSession',
                            params: sessionRequest,
                        };
                        // Handler for initial connection messages - checks for error responses
                        // and properly rejects the connection promise with EIP-1193 error codes
                        initialConnectionMessageHandler = (message) => __awaiter(this, void 0, void 0, function* () {
                            if (typeof message !== 'object' || message === null) {
                                return;
                            }
                            if (!('data' in message)) {
                                return;
                            }
                            const messagePayload = message.data;
                            // Match by ID (preferred) or by method (backward compatibility for notifications without ID)
                            const isMatchingId = messagePayload.id === request.id;
                            const isMatchingMethod = messagePayload.method === 'wallet_createSession' ||
                                messagePayload.method === 'wallet_sessionChanged';
                            if (!isMatchingId && !isMatchingMethod) {
                                return;
                            }
                            // Handle error response (e.g., user rejected the connection)
                            if (messagePayload.error) {
                                return rejectConnection(this.parseWalletError(messagePayload.error));
                            }
                            // Success case - store session, notify, and resolve
                            yield this.storeWalletSession(request, messagePayload);
                            yield this.removeStoredPendingSessionRequest();
                            this.notifyCallbacks(messagePayload);
                            return resolveConnection();
                        });
                        this.dappClient.on('message', initialConnectionMessageHandler);
                        dappClient
                            .connect({
                            mode: 'trusted',
                            initialPayload: {
                                name: MULTICHAIN_PROVIDER_STREAM_NAME,
                                data: request,
                            },
                        })
                            .catch((error) => {
                            if (initialConnectionMessageHandler) {
                                this.dappClient.off('message', initialConnectionMessageHandler);
                            }
                            rejectConnection(error);
                        });
                    });
                }
                timeout = setTimeout(() => {
                    reject(new TransportTimeoutError());
                }, storedSessionRequestBeforeConnectionAttempt
                    ? this.options.resumeTimeout
                    : this.options.connectionTimeout);
                connection.then(resolve).catch(reject);
            }));
            return connectionPromise
                .catch((error) => __awaiter(this, void 0, void 0, function* () {
                // Clean up the MWP session from the KVStore so stale sessions
                // don't cause subsequent connect attempts to enter the resume path
                yield this.dappClient.disconnect();
                throw error;
            }))
                .finally(() => {
                if (timeout) {
                    clearTimeout(timeout);
                }
                if (initialConnectionMessageHandler) {
                    this.dappClient.off('message', initialConnectionMessageHandler);
                    initialConnectionMessageHandler = undefined;
                }
                this.removeStoredPendingSessionRequest();
            });
        });
    }
    /**
     * Disconnects from the Mobile Wallet Protocol
     *
     * @param [scopes] - The scopes to revoke. If not provided or empty, all scopes will be revoked.
     * @returns Nothing
     */
    disconnect() {
        return __awaiter(this, arguments, void 0, function* (scopes = []) {
            var _a, _b;
            const cachedSession = yield this.getCachedResponse({
                jsonrpc: '2.0',
                id: '0',
                method: 'wallet_getSession',
            });
            const cachedSessionScopes = (_b = (_a = cachedSession === null || cachedSession === void 0 ? void 0 : cachedSession.result) === null || _a === void 0 ? void 0 : _a.sessionScopes) !== null && _b !== void 0 ? _b : {};
            const remainingScopes = scopes.length === 0
                ? []
                : Object.keys(cachedSessionScopes).filter((scope) => !scopes.includes(scope));
            const newSessionScopes = Object.fromEntries(Object.entries(cachedSessionScopes).filter(([key]) => remainingScopes.includes(key)));
            // NOTE: Purposely not awaiting this to avoid blocking the disconnect flow.
            // This might not actually get executed on the wallet if the user doesn't open
            // their wallet before the message TTL or if the underlying transport isn't actually connected
            this.request({ method: 'wallet_revokeSession', params: { scopes } }).catch((err) => {
                console.error('error revoking session', err);
            });
            // Clear the cached values for eth_accounts and eth_chainId if all eip155 scopes were removed.
            const remainingScopesIncludeEip155 = remainingScopes.some((scope) => scope.includes('eip155'));
            if (!remainingScopesIncludeEip155) {
                this.kvstore.delete(ACCOUNTS_STORE_KEY);
                this.kvstore.delete(CHAIN_STORE_KEY);
            }
            if (remainingScopes.length > 0) {
                this.kvstore.set(SESSION_STORE_KEY, JSON.stringify({
                    result: {
                        sessionScopes: newSessionScopes,
                    },
                }));
            }
            else {
                this.kvstore.delete(SESSION_STORE_KEY);
                // Clean up window focus event listener
                if (typeof window !== 'undefined' &&
                    typeof window.removeEventListener !== 'undefined' &&
                    this.windowFocusHandler) {
                    window.removeEventListener('focus', this.windowFocusHandler);
                    this.windowFocusHandler = undefined;
                }
                yield this.dappClient.disconnect();
            }
            this.notifyCallbacks({
                method: 'wallet_sessionChanged',
                params: {
                    sessionScopes: newSessionScopes,
                },
            });
        });
    }
    /**
     * Checks if the transport is connected
     *
     * @returns True if transport is connected, false otherwise
     */
    isConnected() {
        // biome-ignore lint/suspicious/noExplicitAny:  required if state is not made public in dappClient
        return this.dappClient.state === 'CONNECTED';
    }
    /**
     * Attempts to re-establish a connection via DappClient
     *
     * @returns Nothing
     */
    // TODO: We should re-evaluate adding this to the WebSocketTransport layer from `@metamask/mobile-wallet-protocol-core`
    // ticket: https://consensyssoftware.atlassian.net/browse/WAPI-862
    attemptResumeSession() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                yield this.dappClient.reconnect();
                // Wait for connection to be established
                yield new Promise((resolve, reject) => {
                    const timeout = setTimeout(() => {
                        reject(new Error('Resume timeout'));
                    }, 2000);
                    if (this.isConnected()) {
                        clearTimeout(timeout);
                        resolve();
                    }
                    else {
                        this.dappClient.once('connected', () => {
                            clearTimeout(timeout);
                            resolve();
                        });
                    }
                });
            }
            catch (error) {
                return Promise.reject(new Error(`Failed to resume session: ${error.message}`));
            }
        });
    }
    getCachedResponse(request) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a;
            if (request.method === 'wallet_getSession') {
                const walletGetSession = yield this.kvstore.get(SESSION_STORE_KEY);
                if (walletGetSession) {
                    const walletSession = JSON.parse(walletGetSession);
                    return {
                        id: request.id,
                        jsonrpc: '2.0',
                        result: (_a = walletSession.params) !== null && _a !== void 0 ? _a : walletSession.result, // "what?... why walletSession.params?.."
                        method: request.method,
                    };
                }
            }
            else if (request.method === 'eth_accounts') {
                const ethAccounts = yield this.kvstore.get(ACCOUNTS_STORE_KEY);
                if (ethAccounts) {
                    return {
                        id: request.id,
                        jsonrpc: '2.0',
                        result: JSON.parse(ethAccounts),
                        method: request.method,
                    };
                }
            }
            else if (request.method === 'eth_chainId') {
                const ethChainId = yield this.kvstore.get(CHAIN_STORE_KEY);
                if (ethChainId) {
                    return {
                        id: request.id,
                        jsonrpc: '2.0',
                        result: JSON.parse(ethChainId),
                        method: request.method,
                    };
                }
            }
        });
    }
    storeWalletSession(request, response) {
        return __awaiter(this, void 0, void 0, function* () {
            if (response.error) {
                return;
            }
            if (CACHED_METHOD_LIST.includes(request.method)) {
                yield this.kvstore.set(SESSION_STORE_KEY, JSON.stringify(response));
            }
            else if (request.method === 'eth_accounts') {
                yield this.kvstore.set(ACCOUNTS_STORE_KEY, JSON.stringify(response.result));
            }
            else if (request.method === 'eth_chainId') {
                yield this.kvstore.set(CHAIN_STORE_KEY, JSON.stringify(response.result));
            }
            else if (CACHED_RESET_METHOD_LIST.includes(request.method)) {
                yield this.kvstore.delete(SESSION_STORE_KEY);
                yield this.kvstore.delete(ACCOUNTS_STORE_KEY);
                yield this.kvstore.delete(CHAIN_STORE_KEY);
            }
        });
    }
    request(payload, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = Object.assign({ jsonrpc: '2.0', id: String(getUniqueRequestId()) }, payload);
            const cachedWalletSession = yield this.getCachedResponse(request);
            if (cachedWalletSession) {
                this.notifyCallbacks(cachedWalletSession);
                return cachedWalletSession;
            }
            if (!this.isConnected()) {
                yield this.attemptResumeSession();
            }
            return new Promise((resolve, reject) => {
                var _a;
                const timeout = setTimeout(() => {
                    this.rejectRequest(request.id, new TransportTimeoutError());
                }, (_a = options === null || options === void 0 ? void 0 : options.timeout) !== null && _a !== void 0 ? _a : this.options.requestTimeout);
                this.pendingRequests.set(request.id, {
                    request,
                    method: request.method,
                    resolve: (response) => __awaiter(this, void 0, void 0, function* () {
                        yield this.storeWalletSession(request, response);
                        return resolve(response);
                    }),
                    reject,
                    timeout,
                });
                this.dappClient
                    .sendRequest({
                    name: MULTICHAIN_PROVIDER_STREAM_NAME,
                    data: request,
                })
                    .catch(reject);
            });
        });
    }
    onNotification(callback) {
        this.notificationCallbacks.add(callback);
        return () => {
            this.notificationCallbacks.delete(callback);
        };
    }
    getActiveSession() {
        return __awaiter(this, void 0, void 0, function* () {
            const { kvstore } = this;
            const sessionStore = yield SessionStore.create(kvstore);
            try {
                const [activeSession] = yield sessionStore.list();
                return activeSession;
            }
            catch (error) {
                // TODO: verify if this try catch is necessary
                logger('error getting active session', error);
                return undefined;
            }
        });
    }
    // This method checks if an existing CAIP session response is cached or waits for one
    // to be received from the wallet if not cached. This is necessary because there is an edge
    // case during the initial connection flow where after the user has accepted the permission approval
    // and returned back to the dapp from the wallet, the dapp page may have gotten unloaded and refreshed.
    // When it is unloaded and refreshed, it will try to resume the session by making a request for wallet_getSession
    // which should resolve from cache, but because a race condition makes it possible for the response from the wallet
    // for the initial wallet_createSession connection request to not have been handled and cached yet. This results
    // in the wallet_getSession request never resolving unless we wait for it explicitly as done in this method.
    waitForWalletSessionIfNotCached() {
        return __awaiter(this, void 0, void 0, function* () {
            const cachedWalletGetSessionResponse = yield this.kvstore.get(SESSION_STORE_KEY);
            if (cachedWalletGetSessionResponse) {
                return;
            }
            let unsubscribe;
            const responsePromise = new Promise((resolve) => {
                unsubscribe = this.onNotification((message) => {
                    if (typeof message === 'object' && message !== null) {
                        if ('data' in message) {
                            const messagePayload = message.data;
                            if (messagePayload.method === 'wallet_getSession' ||
                                messagePayload.method === 'wallet_sessionChanged') {
                                unsubscribe();
                                resolve();
                            }
                        }
                    }
                });
            });
            const timeoutPromise = new Promise((_resolve, reject) => {
                setTimeout(() => {
                    unsubscribe();
                    this.removeStoredPendingSessionRequest();
                    reject(new TransportTimeoutError());
                }, this.options.resumeTimeout);
            });
            return Promise.race([responsePromise, timeoutPromise]);
        });
    }
}
