// @metamask/controller-utils/dist/constants.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CHAIN_IDS_WITH_NO_NATIVE_TOKEN = exports.DAYS = exports.DAY = exports.HOURS = exports.HOUR = exports.MINUTES = exports.MINUTE = exports.SECONDS = exports.SECOND = exports.CHAIN_ID_TO_ETHERS_NETWORK_NAME_MAP = exports.ApprovalType = exports.ORIGIN_METAMASK = exports.NFT_API_TIMEOUT = exports.NFT_API_VERSION = exports.NFT_API_BASE_URL = exports.OPENSEA_PROXY_URL = exports.NETWORKS_BYPASSING_VALIDATION = exports.BUILT_IN_NETWORKS = exports.BUILT_IN_CUSTOM_NETWORKS_RPC = exports.TESTNET_TICKER_SYMBOLS = exports.ASSET_TYPES = exports.GWEI = exports.ERC1155_TOKEN_RECEIVER_INTERFACE_ID = exports.ERC1155_METADATA_URI_INTERFACE_ID = exports.ERC1155_INTERFACE_ID = exports.ERC721_ENUMERABLE_INTERFACE_ID = exports.ERC721_METADATA_INTERFACE_ID = exports.ERC721_INTERFACE_ID = exports.ERC20 = exports.ERC1155 = exports.ERC721 = exports.MAX_SAFE_CHAIN_ID = exports.GANACHE_CHAIN_ID = exports.IPFS_DEFAULT_GATEWAY_URL = exports.FALL_BACK_VS_CURRENCY = exports.RPC = void 0;
const types_1 = require("./types.cjs");
exports.RPC = 'rpc';
exports.FALL_BACK_VS_CURRENCY = 'ETH';
exports.IPFS_DEFAULT_GATEWAY_URL = 'https://cloudflare-ipfs.com/ipfs/';
// NETWORKS ID
// `toHex` not invoked to avoid cyclic dependency
exports.GANACHE_CHAIN_ID = '0x539'; // toHex(1337)
/**
 * The largest possible chain ID we can handle.
 * Explanation: https://gist.github.com/rekmarks/a47bd5f2525936c4b8eee31a16345553
 */
exports.MAX_SAFE_CHAIN_ID = 4503599627370476;
// TOKEN STANDARDS
exports.ERC721 = 'ERC721';
exports.ERC1155 = 'ERC1155';
exports.ERC20 = 'ERC20';
// TOKEN INTERFACE IDS
exports.ERC721_INTERFACE_ID = '0x80ac58cd';
exports.ERC721_METADATA_INTERFACE_ID = '0x5b5e139f';
exports.ERC721_ENUMERABLE_INTERFACE_ID = '0x780e9d63';
exports.ERC1155_INTERFACE_ID = '0xd9b67a26';
exports.ERC1155_METADATA_URI_INTERFACE_ID = '0x0e89341c';
exports.ERC1155_TOKEN_RECEIVER_INTERFACE_ID = '0x4e2312e0';
// UNITS
exports.GWEI = 'gwei';
// ASSET TYPES
exports.ASSET_TYPES = {
    NATIVE: 'NATIVE',
    TOKEN: 'TOKEN',
    NFT: 'NFT',
    UNKNOWN: 'UNKNOWN',
};
// TICKER SYMBOLS
exports.TESTNET_TICKER_SYMBOLS = {
    GOERLI: 'GoerliETH',
    SEPOLIA: 'SepoliaETH',
    LINEA_GOERLI: 'LineaETH',
    LINEA_SEPOLIA: 'LineaETH',
    MEGAETH_TESTNET: 'MegaETH',
    MEGAETH_TESTNET_V2: 'MegaETH',
};
/**
 * Map of all built-in custom networks to their RPC endpoints.
 */
exports.BUILT_IN_CUSTOM_NETWORKS_RPC = {
    /**
     * @deprecated Please use `megaeth-testnet` instead.
     */
    MEGAETH_TESTNET: 'https://carrot.megaeth.com/rpc',
    /**
     * @deprecated Please use `megaeth-testnet-v2` instead.
     */
    'megaeth-testnet': 'https://carrot.megaeth.com/rpc',
    'megaeth-testnet-v2': 'https://carrot.megaeth.com/rpc',
    'monad-testnet': 'https://testnet-rpc.monad.xyz',
};
/**
 * Map of all build-in Infura networks to their network, ticker and chain IDs.
 */
exports.BUILT_IN_NETWORKS = {
    [types_1.NetworkType.goerli]: {
        chainId: types_1.ChainId.goerli,
        ticker: types_1.NetworksTicker.goerli,
        rpcPrefs: {
            blockExplorerUrl: types_1.BlockExplorerUrl.goerli,
        },
    },
    [types_1.NetworkType.sepolia]: {
        chainId: types_1.ChainId.sepolia,
        ticker: types_1.NetworksTicker.sepolia,
        rpcPrefs: {
            blockExplorerUrl: types_1.BlockExplorerUrl.sepolia,
        },
    },
    [types_1.NetworkType.mainnet]: {
        chainId: types_1.ChainId.mainnet,
        ticker: types_1.NetworksTicker.mainnet,
        rpcPrefs: {
            blockExplorerUrl: types_1.BlockExplorerUrl.mainnet,
        },
    },
    [types_1.NetworkType['linea-goerli']]: {
        chainId: types_1.ChainId['linea-goerli'],
        ticker: types_1.NetworksTicker['linea-goerli'],
        rpcPrefs: {
            blockExplorerUrl: types_1.BlockExplorerUrl['linea-goerli'],
        },
    },
    [types_1.NetworkType['linea-sepolia']]: {
        chainId: types_1.ChainId['linea-sepolia'],
        ticker: types_1.NetworksTicker['linea-sepolia'],
        rpcPrefs: {
            blockExplorerUrl: types_1.BlockExplorerUrl['linea-sepolia'],
        },
    },
    [types_1.NetworkType['linea-mainnet']]: {
        chainId: types_1.ChainId['linea-mainnet'],
        ticker: types_1.NetworksTicker['linea-mainnet'],
        rpcPrefs: {
            blockExplorerUrl: types_1.BlockExplorerUrl['linea-mainnet'],
        },
    },
    [types_1.NetworkType['megaeth-testnet']]: {
        chainId: types_1.ChainId['megaeth-testnet'],
        ticker: types_1.NetworksTicker['megaeth-testnet'],
        rpcPrefs: {
            blockExplorerUrl: types_1.BlockExplorerUrl['megaeth-testnet'],
        },
    },
    [types_1.NetworkType['megaeth-testnet-v2']]: {
        chainId: types_1.ChainId['megaeth-testnet-v2'],
        ticker: types_1.NetworksTicker['megaeth-testnet-v2'],
        rpcPrefs: {
            blockExplorerUrl: types_1.BlockExplorerUrl['megaeth-testnet-v2'],
        },
    },
    [types_1.NetworkType['monad-testnet']]: {
        chainId: types_1.ChainId['monad-testnet'],
        ticker: types_1.NetworksTicker['monad-testnet'],
        rpcPrefs: {
            blockExplorerUrl: types_1.BlockExplorerUrl['monad-testnet'],
        },
    },
    [types_1.NetworkType['base-mainnet']]: {
        chainId: types_1.ChainId['base-mainnet'],
        ticker: types_1.NetworksTicker['base-mainnet'],
        rpcPrefs: {
            blockExplorerUrl: types_1.BlockExplorerUrl['base-mainnet'],
        },
    },
    [types_1.NetworkType['arbitrum-mainnet']]: {
        chainId: types_1.ChainId['arbitrum-mainnet'],
        ticker: types_1.NetworksTicker['arbitrum-mainnet'],
        rpcPrefs: {
            blockExplorerUrl: types_1.BlockExplorerUrl['arbitrum-mainnet'],
        },
    },
    [types_1.NetworkType['bsc-mainnet']]: {
        chainId: types_1.ChainId['bsc-mainnet'],
        ticker: types_1.NetworksTicker['bsc-mainnet'],
        rpcPrefs: {
            blockExplorerUrl: types_1.BlockExplorerUrl['bsc-mainnet'],
        },
    },
    [types_1.NetworkType['optimism-mainnet']]: {
        chainId: types_1.ChainId['optimism-mainnet'],
        ticker: types_1.NetworksTicker['optimism-mainnet'],
        rpcPrefs: {
            blockExplorerUrl: types_1.BlockExplorerUrl['optimism-mainnet'],
        },
    },
    [types_1.NetworkType['polygon-mainnet']]: {
        chainId: types_1.ChainId['polygon-mainnet'],
        ticker: types_1.NetworksTicker['polygon-mainnet'],
        rpcPrefs: {
            blockExplorerUrl: types_1.BlockExplorerUrl['polygon-mainnet'],
        },
    },
    [types_1.NetworkType['sei-mainnet']]: {
        chainId: types_1.ChainId['sei-mainnet'],
        ticker: types_1.NetworksTicker['sei-mainnet'],
        rpcPrefs: {
            blockExplorerUrl: types_1.BlockExplorerUrl['sei-mainnet'],
        },
    },
    [types_1.NetworkType.rpc]: {
        chainId: undefined,
        blockExplorerUrl: undefined,
        ticker: undefined,
        rpcPrefs: undefined,
    },
};
/**
 * When a user adds a custom network to MetaMask, we perform some basic
 * validations on the network. For instance, usually a network cannot share the
 * same chain as another. In some cases, however, we want to allow networks that
 * would normally be invalid. This mapping contains networks that should bypass
 * validation.
 */
exports.NETWORKS_BYPASSING_VALIDATION = {
    // HyperEVM uses the same chain ID as Wanchain
    '0x3e7': {
        name: 'HyperEVM',
        symbol: 'HYPE',
        rpcUrl: 'https://rpc.hyperliquid.xyz',
    },
};
// APIs
exports.OPENSEA_PROXY_URL = 'https://proxy.api.cx.metamask.io/opensea/v1/api/v2';
exports.NFT_API_BASE_URL = 'https://nft.api.cx.metamask.io';
exports.NFT_API_VERSION = '1';
exports.NFT_API_TIMEOUT = 15000;
// Default origin for controllers
exports.ORIGIN_METAMASK = 'metamask';
/**
 * Approval request types for various operations.
 * These types are used by different controllers to create and manage
 * approval requests consistently.
 */
var ApprovalType;
(function (ApprovalType) {
    ApprovalType["AddEthereumChain"] = "wallet_addEthereumChain";
    ApprovalType["ConnectAccounts"] = "connect_accounts";
    ApprovalType["EthDecrypt"] = "eth_decrypt";
    ApprovalType["EthGetEncryptionPublicKey"] = "eth_getEncryptionPublicKey";
    ApprovalType["EthSignTypedData"] = "eth_signTypedData";
    ApprovalType["PersonalSign"] = "personal_sign";
    ApprovalType["ResultError"] = "result_error";
    ApprovalType["ResultSuccess"] = "result_success";
    ApprovalType["SnapDialogAlert"] = "snap_dialog:alert";
    ApprovalType["SnapDialogConfirmation"] = "snap_dialog:confirmation";
    ApprovalType["SnapDialogPrompt"] = "snap_dialog:prompt";
    ApprovalType["SnapDialogDefault"] = "snap_dialog";
    ApprovalType["SwitchEthereumChain"] = "wallet_switchEthereumChain";
    ApprovalType["Transaction"] = "transaction";
    ApprovalType["TransactionBatch"] = "transaction_batch";
    ApprovalType["Unlock"] = "unlock";
    ApprovalType["WalletConnect"] = "wallet_connect";
    ApprovalType["WalletRequestPermissions"] = "wallet_requestPermissions";
    ApprovalType["WatchAsset"] = "wallet_watchAsset";
})(ApprovalType || (exports.ApprovalType = ApprovalType = {}));
/**
 * Mapping of chain IDs to their network names for ENS functionality.
 * Note: MegaETH-testnet is intentionally excluded from this mapping as it doesn't support ENS.
 */
exports.CHAIN_ID_TO_ETHERS_NETWORK_NAME_MAP = {
    [types_1.ChainId.goerli]: types_1.BuiltInNetworkName.Goerli,
    [types_1.ChainId.sepolia]: types_1.BuiltInNetworkName.Sepolia,
    [types_1.ChainId.mainnet]: types_1.BuiltInNetworkName.Mainnet,
    [types_1.ChainId['linea-goerli']]: types_1.BuiltInNetworkName.LineaGoerli,
    [types_1.ChainId['linea-sepolia']]: types_1.BuiltInNetworkName.LineaSepolia,
    [types_1.ChainId['linea-mainnet']]: types_1.BuiltInNetworkName.LineaMainnet,
    [types_1.ChainId.aurora]: types_1.BuiltInNetworkName.Aurora,
};
/**
 * The number of milliseconds in a second.
 */
exports.SECOND = 1000;
/**
 * The number of milliseconds in a second.
 */
exports.SECONDS = exports.SECOND;
/**
 * The number of milliseconds in a minute.
 */
exports.MINUTE = exports.SECONDS * 60;
/**
 * The number of milliseconds in a minute.
 */
exports.MINUTES = exports.MINUTE;
/**
 * The number of milliseconds in a hour.
 */
exports.HOUR = exports.MINUTES * 60;
/**
 * The number of milliseconds in a hour.
 */
exports.HOURS = exports.HOUR;
/**
 * The number of milliseconds in a day.
 */
exports.DAY = exports.HOURS * 24;
/**
 * The number of milliseconds in a day.
 */
exports.DAYS = exports.DAY;
/**
 * Special "EVM-ish" chains with no native tokens.
 * Created for Tempo, but can be extended to others.
 * - For hidding the native token from the token list and Send list.
 * - For excluding the native token from the total wallet value calculation.
 */
exports.CHAIN_IDS_WITH_NO_NATIVE_TOKEN = [
    'eip155:42431', // Tempo Testnet
    'eip155:4217', // Tempo Mainnet
];
