// @metamask/controller-utils/dist/create-service-policy.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createServicePolicy = exports.DEFAULT_DEGRADED_THRESHOLD = exports.DEFAULT_CIRCUIT_BREAK_DURATION = exports.DEFAULT_MAX_CONSECUTIVE_FAILURES = exports.DEFAULT_MAX_RETRIES = exports.handleWhen = exports.handleAll = exports.ExponentialBackoff = exports.ConstantBackoff = exports.CircuitState = exports.CockatielEventEmitter = exports.BrokenCircuitError = void 0;
const cockatiel_1 = require("cockatiel");
Object.defineProperty(exports, "BrokenCircuitError", { enumerable: true, get: function () { return cockatiel_1.BrokenCircuitError; } });
Object.defineProperty(exports, "CircuitState", { enumerable: true, get: function () { return cockatiel_1.CircuitState; } });
Object.defineProperty(exports, "CockatielEventEmitter", { enumerable: true, get: function () { return cockatiel_1.EventEmitter; } });
Object.defineProperty(exports, "ExponentialBackoff", { enumerable: true, get: function () { return cockatiel_1.ExponentialBackoff; } });
Object.defineProperty(exports, "ConstantBackoff", { enumerable: true, get: function () { return cockatiel_1.ConstantBackoff; } });
Object.defineProperty(exports, "handleAll", { enumerable: true, get: function () { return cockatiel_1.handleAll; } });
Object.defineProperty(exports, "handleWhen", { enumerable: true, get: function () { return cockatiel_1.handleWhen; } });
/**
 * Availability statuses that the service can be in.
 *
 * Used to keep track of whether the `onAvailable` event should be fired.
 */
const AVAILABILITY_STATUSES = {
    Available: 'available',
    Degraded: 'degraded',
    Unavailable: 'unavailable',
    Unknown: 'unknown',
};
/**
 * The maximum number of times that a failing service should be re-run before
 * giving up.
 */
exports.DEFAULT_MAX_RETRIES = 3;
/**
 * The maximum number of times that the service is allowed to fail before
 * pausing further retries. This is set to a value such that if given a
 * service that continually fails, the policy needs to be executed 3 times
 * before further retries are paused.
 */
exports.DEFAULT_MAX_CONSECUTIVE_FAILURES = (1 + exports.DEFAULT_MAX_RETRIES) * 3;
/**
 * The default length of time (in milliseconds) to temporarily pause retries of
 * the service after enough consecutive failures.
 */
exports.DEFAULT_CIRCUIT_BREAK_DURATION = 30 * 60 * 1000;
/**
 * The default length of time (in milliseconds) that governs when the service is
 * regarded as degraded (affecting when `onDegraded` is called).
 */
exports.DEFAULT_DEGRADED_THRESHOLD = 5000;
const isServiceFailure = (error) => {
    if (typeof error === 'object' &&
        error !== null &&
        'httpStatus' in error &&
        typeof error.httpStatus === 'number') {
        return error.httpStatus >= 500;
    }
    // If the error is not an object, or doesn't have a numeric code property,
    // consider it a service failure (e.g., network errors, timeouts, etc.)
    return true;
};
/**
 * The circuit breaker policy inside of the Cockatiel library exposes some of
 * its state, but not all of it. Notably, the time that the circuit opened is
 * not publicly accessible. So we have to record this ourselves.
 *
 * This function therefore allows us to obtain the circuit breaker state that we
 * wish we could access.
 *
 * @param state - The public state of a circuit breaker policy.
 * @returns if the circuit is open, the state of the circuit breaker policy plus
 * the time that it opened, otherwise just the circuit state.
 */
function getInternalCircuitState(state) {
    if (state === cockatiel_1.CircuitState.Open) {
        return { state, openedAt: Date.now() };
    }
    return { state };
}
/**
 * Constructs an object exposing an `execute` method which, given a function —
 * hereafter called the "service" — will retry that service with ever increasing
 * delays until it succeeds. If the policy detects too many consecutive
 * failures, it will block further retries until a designated time period has
 * passed; this particular behavior is primarily designed for services that wrap
 * API calls so as not to make needless HTTP requests when the API is down and
 * to be able to recover when the API comes back up. In addition, hooks allow
 * for responding to certain events, one of which can be used to detect when an
 * HTTP request is performing slowly.
 *
 * Internally, this function makes use of the retry and circuit breaker policies
 * from the [Cockatiel](https://www.npmjs.com/package/cockatiel) library; see
 * there for more.
 *
 * @param options - The options to this function. See
 * {@link CreateServicePolicyOptions}.
 * @returns The service policy.
 * @example
 * This function is designed to be used in the context of a service class like
 * this:
 * ``` ts
 * class Service {
 *   constructor() {
 *     this.#policy = createServicePolicy({
 *       maxRetries: 3,
 *       retryFilterPolicy: handleWhen((error) => {
 *         return error.message.includes('oops');
 *       }),
 *       maxConsecutiveFailures: 3,
 *       circuitBreakDuration: 5000,
 *       degradedThreshold: 2000,
 *       onBreak: () => {
 *         console.log('Circuit broke');
 *       },
 *       onDegraded: () => {
 *         console.log('Service is degraded');
 *       },
 *     });
 *   }
 *
 *   async fetch() {
 *     return await this.#policy.execute(async () => {
 *       const response = await fetch('https://some/url');
 *       return await response.json();
 *     });
 *   }
 * }
 * ```
 */
function createServicePolicy(options = {}) {
    const { maxRetries = exports.DEFAULT_MAX_RETRIES, retryFilterPolicy = cockatiel_1.handleAll, maxConsecutiveFailures = exports.DEFAULT_MAX_CONSECUTIVE_FAILURES, circuitBreakDuration = exports.DEFAULT_CIRCUIT_BREAK_DURATION, degradedThreshold = exports.DEFAULT_DEGRADED_THRESHOLD, backoff = new cockatiel_1.ExponentialBackoff(), } = options;
    let availabilityStatus = AVAILABILITY_STATUSES.Unknown;
    const retryPolicy = (0, cockatiel_1.retry)(retryFilterPolicy, {
        // Note that although the option here is called "max attempts", it's really
        // maximum number of *retries* (attempts past the initial attempt).
        maxAttempts: maxRetries,
        // Retries of the service will be executed following ever increasing delays,
        // determined by a backoff formula.
        backoff,
    });
    const onRetry = retryPolicy.onRetry.bind(retryPolicy);
    const consecutiveBreaker = new cockatiel_1.ConsecutiveBreaker(maxConsecutiveFailures);
    const circuitBreakerPolicy = (0, cockatiel_1.circuitBreaker)((0, cockatiel_1.handleWhen)(isServiceFailure), {
        // While the circuit is open, any additional invocations of the service
        // passed to the policy (either via automatic retries or by manually
        // executing the policy again) will result in a BrokenCircuitError. This
        // will remain the case until `circuitBreakDuration` passes, after which the
        // service will be allowed to run again. If the service succeeds, the
        // circuit will close, otherwise it will remain open.
        halfOpenAfter: circuitBreakDuration,
        breaker: consecutiveBreaker,
    });
    let internalCircuitState = getInternalCircuitState(circuitBreakerPolicy.state);
    circuitBreakerPolicy.onStateChange((state) => {
        internalCircuitState = getInternalCircuitState(state);
    });
    circuitBreakerPolicy.onBreak(() => {
        availabilityStatus = AVAILABILITY_STATUSES.Unavailable;
    });
    const onBreak = circuitBreakerPolicy.onBreak.bind(circuitBreakerPolicy);
    const onDegradedEventEmitter = new cockatiel_1.EventEmitter();
    const onDegraded = onDegradedEventEmitter.addListener;
    const onAvailableEventEmitter = new cockatiel_1.EventEmitter();
    const onAvailable = onAvailableEventEmitter.addListener;
    retryPolicy.onGiveUp((data) => {
        if (circuitBreakerPolicy.state === cockatiel_1.CircuitState.Closed) {
            availabilityStatus = AVAILABILITY_STATUSES.Degraded;
            onDegradedEventEmitter.emit(data);
        }
    });
    retryPolicy.onSuccess(({ duration }) => {
        if (circuitBreakerPolicy.state === cockatiel_1.CircuitState.Closed) {
            if (duration > degradedThreshold) {
                availabilityStatus = AVAILABILITY_STATUSES.Degraded;
                onDegradedEventEmitter.emit();
            }
            else if (availabilityStatus !== AVAILABILITY_STATUSES.Available) {
                availabilityStatus = AVAILABILITY_STATUSES.Available;
                onAvailableEventEmitter.emit();
            }
        }
    });
    // Every time the retry policy makes an attempt, it executes the circuit
    // breaker policy, which executes the service.
    //
    // Calling:
    //
    //   policy.execute(() => {
    //     // do what the service does
    //   })
    //
    // is equivalent to:
    //
    //   retryPolicy.execute(() => {
    //     circuitBreakerPolicy.execute(() => {
    //       // do what the service does
    //     });
    //   });
    //
    // So if the retry policy succeeds or fails, it is because the circuit breaker
    // policy succeeded or failed. And if there are any event listeners registered
    // on the retry policy, by the time they are called, the state of the circuit
    // breaker will have already changed.
    const policy = (0, cockatiel_1.wrap)(retryPolicy, circuitBreakerPolicy);
    const getRemainingCircuitOpenDuration = () => {
        if (internalCircuitState.state === cockatiel_1.CircuitState.Open) {
            return internalCircuitState.openedAt + circuitBreakDuration - Date.now();
        }
        return null;
    };
    const getCircuitState = () => {
        return circuitBreakerPolicy.state;
    };
    const reset = () => {
        // Set the state of the policy to "isolated" regardless of its current state
        const { dispose } = circuitBreakerPolicy.isolate();
        // Reset the state to "closed"
        dispose();
        // Reset the counter on the breaker as well
        consecutiveBreaker.success();
        // Re-initialize the availability status so that if the service is executed
        // successfully, onAvailable listeners will be called again
        availabilityStatus = AVAILABILITY_STATUSES.Unknown;
    };
    return {
        ...policy,
        circuitBreakerPolicy,
        circuitBreakDuration,
        getCircuitState,
        getRemainingCircuitOpenDuration,
        reset,
        retryPolicy,
        onBreak,
        onDegraded,
        onAvailable,
        onRetry,
    };
}
exports.createServicePolicy = createServicePolicy;
