// @metamask/controller-utils/dist/types.mjs
/**
 * The names of built-in Infura networks
 */
export const InfuraNetworkType = {
    mainnet: 'mainnet',
    goerli: 'goerli',
    sepolia: 'sepolia',
    'linea-goerli': 'linea-goerli',
    'linea-sepolia': 'linea-sepolia',
    'linea-mainnet': 'linea-mainnet',
    'base-mainnet': 'base-mainnet',
    'arbitrum-mainnet': 'arbitrum-mainnet',
    'bsc-mainnet': 'bsc-mainnet',
    'optimism-mainnet': 'optimism-mainnet',
    'polygon-mainnet': 'polygon-mainnet',
    'sei-mainnet': 'sei-mainnet',
};
/**
 * Custom network types that are not part of Infura.
 */
export const CustomNetworkType = {
    /**
     * @deprecated `megaeth-testnet` is migrated to `megaeth-testnet-v2`.
     */
    'megaeth-testnet': 'megaeth-testnet',
    'megaeth-testnet-v2': 'megaeth-testnet-v2',
    'monad-testnet': 'monad-testnet',
};
/**
 * The "network type"; either the name of a built-in network, or "rpc" for custom networks.
 */
export const NetworkType = {
    ...InfuraNetworkType,
    ...CustomNetworkType,
    rpc: 'rpc',
};
/**
 * A helper to determine whether a given input is NetworkType.
 *
 * @param val - the value to check whether it is NetworkType or not.
 * @returns boolean indicating whether or not the argument is NetworkType.
 */
export function isNetworkType(val) {
    return Object.values(NetworkType).includes(val);
}
/**
 * A type guard to determine whether the input is an InfuraNetworkType.
 *
 * @param value - The value to check.
 * @returns True if the given value is within the InfuraNetworkType enum,
 * false otherwise.
 */
export function isInfuraNetworkType(value) {
    const infuraNetworkTypes = Object.keys(InfuraNetworkType);
    return infuraNetworkTypes.includes(value);
}
/**
 * Names of networks built into the wallet.
 *
 * This includes both Infura and non-Infura networks.
 */
export var BuiltInNetworkName;
(function (BuiltInNetworkName) {
    BuiltInNetworkName["Mainnet"] = "mainnet";
    BuiltInNetworkName["Goerli"] = "goerli";
    BuiltInNetworkName["Sepolia"] = "sepolia";
    BuiltInNetworkName["LineaGoerli"] = "linea-goerli";
    BuiltInNetworkName["LineaSepolia"] = "linea-sepolia";
    BuiltInNetworkName["LineaMainnet"] = "linea-mainnet";
    BuiltInNetworkName["Aurora"] = "aurora";
    /**
     * @deprecated `MegaETHTestnet` is migrated to `MegaETHTestnetV2`.
     */
    BuiltInNetworkName["MegaETHTestnet"] = "megaeth-testnet";
    BuiltInNetworkName["MegaETHTestnetV2"] = "megaeth-testnet-v2";
    BuiltInNetworkName["MonadTestnet"] = "monad-testnet";
    BuiltInNetworkName["BaseMainnet"] = "base-mainnet";
    BuiltInNetworkName["ArbitrumOne"] = "arbitrum-mainnet";
    BuiltInNetworkName["BscMainnet"] = "bsc-mainnet";
    BuiltInNetworkName["OptimismMainnet"] = "optimism-mainnet";
    BuiltInNetworkName["PolygonMainnet"] = "polygon-mainnet";
    BuiltInNetworkName["SeiMainnet"] = "sei-mainnet";
    BuiltInNetworkName["MegaETHMainnet"] = "megaeth-mainnet";
})(BuiltInNetworkName || (BuiltInNetworkName = {}));
/**
 * Decimal string chain IDs of built-in networks, by name.
 *
 * `toHex` not invoked to avoid cyclic dependency
 */
export const ChainId = {
    [BuiltInNetworkName.Mainnet]: '0x1', // toHex(1)
    [BuiltInNetworkName.Goerli]: '0x5', // toHex(5)
    [BuiltInNetworkName.Sepolia]: '0xaa36a7', // toHex(11155111)
    [BuiltInNetworkName.Aurora]: '0x4e454152', // toHex(1313161554)
    [BuiltInNetworkName.LineaGoerli]: '0xe704', // toHex(59140)
    [BuiltInNetworkName.LineaSepolia]: '0xe705', // toHex(59141)
    [BuiltInNetworkName.LineaMainnet]: '0xe708', // toHex(59144)
    /**
     * @deprecated `MegaETHTestnet` is migrated to `MegaETHTestnetV2`.
     */
    [BuiltInNetworkName.MegaETHTestnet]: '0x18c6', // toHex(6342)
    [BuiltInNetworkName.MegaETHTestnetV2]: '0x18c7', // toHex(6343)
    [BuiltInNetworkName.MonadTestnet]: '0x279f', // toHex(10143)
    [BuiltInNetworkName.BaseMainnet]: '0x2105', // toHex(8453)
    [BuiltInNetworkName.ArbitrumOne]: '0xa4b1', // toHex(42161)
    [BuiltInNetworkName.BscMainnet]: '0x38', // toHex(56)
    [BuiltInNetworkName.OptimismMainnet]: '0xa', // toHex(10)
    [BuiltInNetworkName.PolygonMainnet]: '0x89', // toHex(137)
    [BuiltInNetworkName.SeiMainnet]: '0x531', // toHex(1329)
    [BuiltInNetworkName.MegaETHMainnet]: '0x10e6', // toHex(4326)
};
/* eslint-disable @typescript-eslint/naming-convention */
export var NetworksTicker;
(function (NetworksTicker) {
    NetworksTicker["mainnet"] = "ETH";
    NetworksTicker["goerli"] = "GoerliETH";
    NetworksTicker["sepolia"] = "SepoliaETH";
    NetworksTicker["linea-goerli"] = "LineaETH";
    // eslint-disable-next-line @typescript-eslint/no-duplicate-enum-values
    NetworksTicker["linea-sepolia"] = "LineaETH";
    // eslint-disable-next-line @typescript-eslint/no-duplicate-enum-values
    NetworksTicker["linea-mainnet"] = "ETH";
    /**
     * @deprecated `megaeth-testnet` is migrated to `megaeth-testnet-v2`.
     */
    NetworksTicker["megaeth-testnet"] = "MegaETH";
    // eslint-disable-next-line @typescript-eslint/no-duplicate-enum-values
    NetworksTicker["megaeth-testnet-v2"] = "MegaETH";
    NetworksTicker["monad-testnet"] = "MON";
    // eslint-disable-next-line @typescript-eslint/no-duplicate-enum-values
    NetworksTicker["base-mainnet"] = "ETH";
    // eslint-disable-next-line @typescript-eslint/no-duplicate-enum-values
    NetworksTicker["arbitrum-mainnet"] = "ETH";
    NetworksTicker["bsc-mainnet"] = "BNB";
    // eslint-disable-next-line @typescript-eslint/no-duplicate-enum-values
    NetworksTicker["optimism-mainnet"] = "ETH";
    NetworksTicker["polygon-mainnet"] = "POL";
    NetworksTicker["sei-mainnet"] = "SEI";
    NetworksTicker["rpc"] = "";
})(NetworksTicker || (NetworksTicker = {}));
/* eslint-enable @typescript-eslint/naming-convention */
export const BlockExplorerUrl = {
    [BuiltInNetworkName.Mainnet]: 'https://etherscan.io',
    [BuiltInNetworkName.Goerli]: 'https://goerli.etherscan.io',
    [BuiltInNetworkName.Sepolia]: 'https://sepolia.etherscan.io',
    [BuiltInNetworkName.LineaGoerli]: 'https://goerli.lineascan.build',
    [BuiltInNetworkName.LineaSepolia]: 'https://sepolia.lineascan.build',
    [BuiltInNetworkName.LineaMainnet]: 'https://lineascan.build',
    /**
     * @deprecated `MegaETHTestnet` is migrated to `MegaETHTestnetV2`.
     */
    [BuiltInNetworkName.MegaETHTestnet]: 'https://megaexplorer.xyz',
    [BuiltInNetworkName.MegaETHTestnetV2]: 'https://megaeth-testnet-v2.blockscout.com',
    [BuiltInNetworkName.MonadTestnet]: 'https://testnet.monadexplorer.com',
    [BuiltInNetworkName.BaseMainnet]: 'https://basescan.org',
    [BuiltInNetworkName.ArbitrumOne]: 'https://arbiscan.io',
    [BuiltInNetworkName.BscMainnet]: 'https://bscscan.com',
    [BuiltInNetworkName.OptimismMainnet]: 'https://optimistic.etherscan.io',
    [BuiltInNetworkName.PolygonMainnet]: 'https://polygonscan.com',
    [BuiltInNetworkName.SeiMainnet]: 'https://seitrace.com',
};
export const NetworkNickname = {
    [BuiltInNetworkName.Mainnet]: 'Ethereum Mainnet',
    [BuiltInNetworkName.Goerli]: 'Goerli',
    [BuiltInNetworkName.Sepolia]: 'Sepolia',
    [BuiltInNetworkName.LineaGoerli]: 'Linea Goerli',
    [BuiltInNetworkName.LineaSepolia]: 'Linea Sepolia',
    [BuiltInNetworkName.LineaMainnet]: 'Linea',
    /**
     * @deprecated `MegaETHTestnet` is migrated to `MegaETHTestnetV2`.
     */
    [BuiltInNetworkName.MegaETHTestnet]: 'Mega Testnet',
    [BuiltInNetworkName.MegaETHTestnetV2]: 'MegaETH Testnet',
    [BuiltInNetworkName.MonadTestnet]: 'Monad Testnet',
    [BuiltInNetworkName.BaseMainnet]: 'Base Mainnet',
    [BuiltInNetworkName.ArbitrumOne]: 'Arbitrum One',
    [BuiltInNetworkName.BscMainnet]: 'BSC Mainnet',
    [BuiltInNetworkName.OptimismMainnet]: 'Optimism Mainnet',
    [BuiltInNetworkName.PolygonMainnet]: 'Polygon Mainnet',
    [BuiltInNetworkName.SeiMainnet]: 'Sei Mainnet',
};
