// @metamask/controller-utils/dist/util.cjs
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isEqualCaseInsensitive = exports.isValidJson = exports.isNonEmptyArray = exports.isPlainObject = exports.convertHexToDecimal = exports.query = exports.normalizeEnsName = exports.timeoutFetch = exports.fetchWithErrorHandling = exports.handleFetch = exports.successfulFetch = exports.HttpError = exports.isSmartContractCode = exports.isValidHexAddress = exports.toChecksumHexAddress = exports.safelyExecuteWithTimeout = exports.safelyExecute = exports.toHex = exports.fromHex = exports.hexToText = exports.hexToBN = exports.getBuyURL = exports.weiHexToGweiDec = exports.gweiDecToWEIBN = exports.fractionBN = exports.BNToHex = exports.isSafeChainId = exports.isSafeDynamicKey = exports.PROTOTYPE_POLLUTION_BLOCKLIST = void 0;
const ethjs_unit_1 = require("@metamask/ethjs-unit");
const utils_1 = require("@metamask/utils");
const bn_js_1 = __importDefault(require("bn.js"));
const eth_ens_namehash_1 = __importDefault(require("eth-ens-namehash"));
const fast_deep_equal_1 = __importDefault(require("fast-deep-equal"));
const lodash_1 = require("lodash");
const constants_1 = require("./constants.cjs");
const TIMEOUT_ERROR = new Error('timeout');
exports.PROTOTYPE_POLLUTION_BLOCKLIST = [
    '__proto__',
    'constructor',
    'prototype',
];
/**
 * Checks whether a dynamic property key could be used in
 * a [prototype pollution attack](https://portswigger.net/web-security/prototype-pollution).
 *
 * @param key - The dynamic key to validate.
 * @returns Whether the given dynamic key is safe to use.
 */
function isSafeDynamicKey(key) {
    return (typeof key === 'string' &&
        !exports.PROTOTYPE_POLLUTION_BLOCKLIST.some((blockedKey) => key === blockedKey));
}
exports.isSafeDynamicKey = isSafeDynamicKey;
/**
 * Checks whether the given number primitive chain ID is safe.
 * Because some cryptographic libraries we use expect the chain ID to be a
 * number primitive, it must not exceed a certain size.
 *
 * @param chainId - The chain ID to check for safety.
 * @returns Whether the given chain ID is safe.
 */
function isSafeChainId(chainId) {
    if (!(0, utils_1.isHexString)(chainId)) {
        return false;
    }
    const decimalChainId = Number.parseInt(chainId, (0, utils_1.isStrictHexString)(chainId) ? 16 : 10);
    return (Number.isSafeInteger(decimalChainId) &&
        decimalChainId > 0 &&
        decimalChainId <= constants_1.MAX_SAFE_CHAIN_ID);
}
exports.isSafeChainId = isSafeChainId;
/**
 * Converts a BN or BigNumber object to a hex string with a '0x' prefix.
 *
 * @param inputBn - BN|BigNumber instance to convert to a hex string.
 * @returns A '0x'-prefixed hex string.
 */
// TODO: Fix naming convention.
// eslint-disable-next-line @typescript-eslint/naming-convention
function BNToHex(inputBn) {
    return (0, utils_1.add0x)(inputBn.toString(16));
}
exports.BNToHex = BNToHex;
/**
 * Used to multiply a BN by a fraction.
 *
 * @param targetBN - Number to multiply by a fraction.
 * @param numerator - Numerator of the fraction multiplier.
 * @param denominator - Denominator of the fraction multiplier.
 * @returns Product of the multiplication.
 */
function fractionBN(targetBN, numerator, denominator) {
    const numBN = new bn_js_1.default(numerator);
    const denomBN = new bn_js_1.default(denominator);
    return targetBN.mul(numBN).div(denomBN);
}
exports.fractionBN = fractionBN;
/**
 * Used to convert a base-10 number from GWEI to WEI. Can handle numbers with decimal parts.
 *
 * @param value - The base 10 number to convert to WEI.
 * @returns The number in WEI, as a BN.
 */
function gweiDecToWEIBN(value) {
    if (Number.isNaN(value)) {
        return new bn_js_1.default(0);
    }
    const parts = value.toString().split('.');
    const wholePart = parts[0] || '0';
    let decimalPart = parts[1] || '';
    if (!decimalPart) {
        return (0, ethjs_unit_1.toWei)(wholePart, 'gwei');
    }
    if (decimalPart.length <= 9) {
        return (0, ethjs_unit_1.toWei)(`${wholePart}.${decimalPart}`, 'gwei');
    }
    const decimalPartToRemove = decimalPart.slice(9);
    const decimalRoundingDigit = decimalPartToRemove[0];
    decimalPart = decimalPart.slice(0, 9);
    let wei = (0, ethjs_unit_1.toWei)(`${wholePart}.${decimalPart}`, 'gwei');
    if (Number(decimalRoundingDigit) >= 5) {
        wei = wei.add(new bn_js_1.default(1));
    }
    return wei;
}
exports.gweiDecToWEIBN = gweiDecToWEIBN;
/**
 * Used to convert values from wei hex format to dec gwei format.
 *
 * @param hexValue - The value in hex wei.
 * @returns The value in dec gwei as string.
 */
function weiHexToGweiDec(hexValue) {
    const hexWei = new bn_js_1.default((0, utils_1.remove0x)(hexValue), 16);
    return (0, ethjs_unit_1.fromWei)(hexWei, 'gwei');
}
exports.weiHexToGweiDec = weiHexToGweiDec;
/**
 * Return a URL that can be used to obtain ETH for a given network.
 *
 * @param networkCode - Network code of desired network.
 * @param address - Address to deposit obtained ETH.
 * @param amount - How much ETH is desired.
 * @returns URL to buy ETH based on network.
 */
function getBuyURL(networkCode = '1', address, amount = 5) {
    switch (networkCode) {
        case '1':
            return `https://buy.coinbase.com/?code=9ec56d01-7e81-5017-930c-513daa27bb6a&amount=${amount}&address=${address}&crypto_currency=ETH`;
        case '5':
            return 'https://goerli-faucet.slock.it/';
        case '11155111':
            return 'https://sepoliafaucet.net/';
        default:
            return undefined;
    }
}
exports.getBuyURL = getBuyURL;
/**
 * Converts a hex string to a BN object.
 *
 * @param inputHex - Number represented as a hex string.
 * @returns A BN instance.
 */
function hexToBN(inputHex) {
    return inputHex ? new bn_js_1.default((0, utils_1.remove0x)(inputHex), 16) : new bn_js_1.default(0);
}
exports.hexToBN = hexToBN;
/**
 * A helper function that converts hex data to human readable string.
 *
 * @param hexValue - The hex string to convert to string.
 * @returns A human readable string conversion.
 */
function hexToText(hexValue) {
    try {
        const stripped = (0, utils_1.remove0x)(hexValue);
        // TODO: Use `@metamask/utils` version of this function to avoid use of
        // Buffer.
        // eslint-disable-next-line no-restricted-globals
        const buff = Buffer.from(stripped, 'hex');
        return buff.toString('utf8');
    }
    catch {
        /* istanbul ignore next */
        return hexValue;
    }
}
exports.hexToText = hexToText;
/**
 * Parses a hex string and converts it into a number that can be operated on in a bignum-safe,
 * base-10 way.
 *
 * @param value - A base-16 number encoded as a string.
 * @returns The number as a BN object in base-16 mode.
 */
function fromHex(value) {
    if (bn_js_1.default.isBN(value)) {
        return value;
    }
    return new bn_js_1.default(hexToBN(value).toString(10));
}
exports.fromHex = fromHex;
/**
 * Converts an integer to a hexadecimal representation.
 *
 * @param value - An integer, an integer encoded as a base-10 string, or a BN.
 * @returns The integer encoded as a hex string.
 */
function toHex(value) {
    if (typeof value === 'string' && (0, utils_1.isStrictHexString)(value)) {
        return value;
    }
    const hexString = bn_js_1.default.isBN(value) || typeof value === 'bigint'
        ? value.toString(16)
        : new bn_js_1.default(value.toString(), 10).toString(16);
    return `0x${hexString}`;
}
exports.toHex = toHex;
/**
 * Execute and return an asynchronous operation without throwing errors.
 *
 * @param operation - Function returning a Promise.
 * @param logError - Determines if the error should be logged.
 * @template Result - Type of the result of the async operation
 * @returns Promise resolving to the result of the async operation.
 */
async function safelyExecute(operation, logError = false) {
    try {
        return await operation();
    }
    catch (error) {
        /* istanbul ignore next */
        if (logError) {
            console.error(error);
        }
        return undefined;
    }
}
exports.safelyExecute = safelyExecute;
/**
 * Execute and return an asynchronous operation with a timeout.
 *
 * @param operation - Function returning a Promise.
 * @param logError - Determines if the error should be logged.
 * @param timeout - Timeout to fail the operation.
 * @template Result - Type of the result of the async operation
 * @returns Promise resolving to the result of the async operation.
 */
async function safelyExecuteWithTimeout(operation, logError = false, timeout = 500) {
    try {
        return await Promise.race([
            operation(),
            new Promise((_resolve, reject) => setTimeout(() => {
                reject(TIMEOUT_ERROR);
            }, timeout)),
        ]);
    }
    catch (error) {
        /* istanbul ignore next */
        if (logError) {
            console.error(error);
        }
        return undefined;
    }
}
exports.safelyExecuteWithTimeout = safelyExecuteWithTimeout;
function toChecksumHexAddressUnmemoized(address) {
    if (typeof address !== 'string') {
        // Mimic behavior of `addHexPrefix` from `ethereumjs-util` (which this
        // function was previously using) for backward compatibility.
        return address;
    }
    const hexPrefixed = (0, utils_1.add0x)(address);
    if (!(0, utils_1.isHexString)(hexPrefixed)) {
        // Version 5.1 of ethereumjs-util would have returned '0xY' for input 'y'
        // but we shouldn't waste effort trying to change case on a clearly invalid
        // string. Instead just return the hex prefixed original string which most
        // closely mimics the original behavior.
        return hexPrefixed;
    }
    try {
        return (0, utils_1.getChecksumAddress)(hexPrefixed);
    }
    catch (error) {
        // This is necessary for backward compatibility with the old behavior of
        // `ethereumjs-util` which would return the original string if the address
        // was invalid.
        if (error instanceof Error && error.message === 'Invalid hex address.') {
            return hexPrefixed;
        }
        throw error;
    }
}
/**
 * Convert an address to a checksummed hexadecimal address.
 *
 * @param address - The address to convert. For backward compatibility reasons,
 * this can be anything, even a non-hex string with an 0x prefix, but that usage
 * is deprecated. Please use a valid hex string (with or without the `0x`
 * prefix).
 * @returns A 0x-prefixed checksummed version of `address` if it is a valid hex
 * string, or the address as given otherwise.
 */
exports.toChecksumHexAddress = (0, lodash_1.memoize)(toChecksumHexAddressUnmemoized);
function isValidHexAddressUnmemoized(possibleAddress, { allowNonPrefixed = true } = {}) {
    const addressToCheck = allowNonPrefixed
        ? (0, utils_1.add0x)(possibleAddress)
        : possibleAddress;
    if (!(0, utils_1.isStrictHexString)(addressToCheck)) {
        return false;
    }
    // We used to rely on `isValidAddress` from `@ethereumjs/util` which allows
    // for upper-case characters too. So we preserve this behavior and use our
    // faster and memoized validation function instead.
    return (0, utils_1.isHexChecksumAddress)(addressToCheck);
}
/**
 * Validates that the input is a hex address. This utility method is a thin
 * wrapper around `isValidHexAddress` from `@metamask/utils`, with the exception
 * that it may return true for non-0x-prefixed hex strings (depending on the
 * option below).
 *
 * @param possibleAddress - Input parameter to check against.
 * @param options - The validation options.
 * @param options.allowNonPrefixed - If true will regard addresses without a
 * `0x` prefix as valid.
 * @returns Whether or not the input is a valid hex address.
 */
exports.isValidHexAddress = (0, lodash_1.memoize)(isValidHexAddressUnmemoized, (possibleAddress, { allowNonPrefixed = true } = {}) => `${possibleAddress}-${allowNonPrefixed}`);
/**
 * Returns whether the given code corresponds to a smart contract.
 *
 * @param code - The potential smart contract code.
 * @returns Whether the code was smart contract code or not.
 */
function isSmartContractCode(code) {
    /* istanbul ignore if */
    if (!code) {
        return false;
    }
    // Geth will return '0x', and ganache-core v2.2.1 will return '0x0'
    const smartContractCode = code !== '0x' && code !== '0x0';
    return smartContractCode;
}
exports.isSmartContractCode = isSmartContractCode;
/**
 * An error representing a non-200 HTTP response.
 */
class HttpError extends Error {
    /**
     * Construct an HTTP error.
     *
     * @param status - The HTTP response status.
     * @param message - The error message.
     */
    constructor(status, message) {
        super(message ?? `Fetch failed with status '${status}'`);
        this.httpStatus = status;
    }
}
exports.HttpError = HttpError;
/**
 * Execute fetch and verify that the response was successful.
 *
 * @param request - Request information.
 * @param options - Fetch options.
 * @returns The fetch response.
 */
async function successfulFetch(request, options) {
    const response = await fetch(request, options);
    if (!response.ok) {
        throw new HttpError(response.status, 
        // TODO: Replace `String` with more specific conversion.
        // eslint-disable-next-line @typescript-eslint/no-base-to-string
        `Fetch failed with status '${response.status}' for request '${String(request)}'`);
    }
    return response;
}
exports.successfulFetch = successfulFetch;
/**
 * Execute fetch and return object response.
 *
 * @param request - The request information.
 * @param options - The fetch options.
 * @returns The fetch response JSON data.
 */
async function handleFetch(request, options) {
    const response = await successfulFetch(request, options);
    const object = await response.json();
    return object;
}
exports.handleFetch = handleFetch;
/**
 * Execute fetch and return object response, log if known error thrown, otherwise rethrow error.
 *
 * @param request - the request options object
 * @param request.url - The request url to query.
 * @param request.options - The fetch options.
 * @param request.timeout - Timeout to fail request
 * @param request.errorCodesToCatch - array of error codes for errors we want to catch in a particular context
 * @returns The fetch response JSON data or undefined (if error occurs).
 */
async function fetchWithErrorHandling({ url, options, timeout, errorCodesToCatch, }) {
    let result;
    try {
        if (timeout) {
            result = Promise.race([
                await handleFetch(url, options),
                new Promise((_resolve, reject) => setTimeout(() => {
                    reject(TIMEOUT_ERROR);
                }, timeout)),
            ]);
        }
        else {
            result = await handleFetch(url, options);
        }
    }
    catch (error) {
        logOrRethrowError(error, errorCodesToCatch);
    }
    return result;
}
exports.fetchWithErrorHandling = fetchWithErrorHandling;
/**
 * Fetch that fails after timeout.
 *
 * @param url - Url to fetch.
 * @param options - Options to send with the request.
 * @param timeout - Timeout to fail request.
 * @returns Promise resolving the request.
 */
async function timeoutFetch(url, options, timeout = 500) {
    return Promise.race([
        successfulFetch(url, options),
        new Promise((_resolve, reject) => setTimeout(() => {
            reject(TIMEOUT_ERROR);
        }, timeout)),
    ]);
}
exports.timeoutFetch = timeoutFetch;
/**
 * Normalizes the given ENS name.
 *
 * @param ensName - The ENS name.
 * @returns The normalized ENS name string.
 */
function normalizeEnsName(ensName) {
    // `.` refers to the registry root contract
    if (ensName === '.') {
        return ensName;
    }
    if (ensName && typeof ensName === 'string') {
        try {
            const normalized = eth_ens_namehash_1.default.normalize(ensName.trim());
            // this regex is only sufficient with the above call to ensNamehash.normalize
            // TODO: change 7 in regex to 3 when shorter ENS domains are live
            if (normalized.match(/^(([\w\d-]+)\.)*[\w\d-]{7,}\.(eth|test)$/u)) {
                return normalized;
            }
        }
        catch {
            // do nothing
        }
    }
    return null;
}
exports.normalizeEnsName = normalizeEnsName;
/**
 * Wrapper method to handle EthQuery requests.
 *
 * @param ethQuery - EthQuery object initialized with a provider.
 * @param method - Method to request.
 * @param args - Arguments to send.
 * @returns Promise resolving the request.
 */
function query(ethQuery, method, 
// TODO: Replace `any` with type
// eslint-disable-next-line @typescript-eslint/no-explicit-any
args = []) {
    return new Promise((resolve, reject) => {
        const callback = (error, result) => {
            if (error) {
                // We don't control the error object returned by eth-query, so
                // we can't guarantee it's an instance of Error.
                // eslint-disable-next-line @typescript-eslint/prefer-promise-reject-errors
                reject(error);
                return;
            }
            resolve(result);
        };
        // Using `in` rather than `hasProperty` so that we look up the prototype
        // chain for the method.
        if (method in ethQuery && typeof ethQuery[method] === 'function') {
            ethQuery[method](...args, callback);
        }
        else {
            ethQuery.sendAsync({ method, params: args }, callback);
        }
    });
}
exports.query = query;
/**
 * Converts valid hex strings to decimal numbers, and handles unexpected arg types.
 *
 * @param value - a string that is either a hexadecimal with `0x` prefix or a decimal string.
 * @returns a decimal number.
 */
const convertHexToDecimal = (value = '0x0') => {
    if ((0, utils_1.isStrictHexString)(value)) {
        return parseInt(value, 16);
    }
    return Number(value) ? Number(value) : 0;
};
exports.convertHexToDecimal = convertHexToDecimal;
/**
 * Determines whether a value is a "plain" object.
 *
 * @param value - A value to check
 * @returns True if the passed value is a plain object
 */
function isPlainObject(value) {
    return Boolean(value) && typeof value === 'object' && !Array.isArray(value);
}
exports.isPlainObject = isPlainObject;
/**
 * Type guard for {@link NonEmptyArray}.
 *
 * @template T - The non-empty array member type.
 * @param value - The value to check.
 * @returns Whether the value is a non-empty array.
 */
function isNonEmptyArray(value) {
    return Array.isArray(value) && value.length > 0;
}
exports.isNonEmptyArray = isNonEmptyArray;
/**
 * Type guard for {@link Json}.
 *
 * @param value - The value to check.
 * @returns Whether the value is valid JSON.
 */
function isValidJson(value) {
    try {
        return (0, fast_deep_equal_1.default)(value, JSON.parse(JSON.stringify(value)));
    }
    catch {
        return false;
    }
}
exports.isValidJson = isValidJson;
/**
 * Utility method to log if error is a common fetch error and otherwise rethrow it.
 *
 * @param error - Caught error that we should either rethrow or log to console
 * @param codesToCatch - array of error codes for errors we want to catch and log in a particular context
 */
function logOrRethrowError(error, codesToCatch = []) {
    if (!error) {
        return;
    }
    if (error instanceof Error) {
        const includesErrorCodeToCatch = codesToCatch.some((code) => error.message.includes(`Fetch failed with status '${code}'`));
        if (includesErrorCodeToCatch ||
            error.message.includes('Failed to fetch') ||
            error === TIMEOUT_ERROR) {
            console.error(error);
        }
        else {
            throw error;
        }
    }
    else {
        // eslint-disable-next-line @typescript-eslint/only-throw-error
        throw error;
    }
}
/**
 * Checks if two strings are equal, ignoring case.
 *
 * @param value1 - The first string to compare.
 * @param value2 - The second string to compare.
 * @returns `true` if the strings are equal, ignoring case; otherwise, `false`.
 */
function isEqualCaseInsensitive(value1, value2) {
    if (typeof value1 !== 'string' || typeof value2 !== 'string') {
        return false;
    }
    return value1.toLowerCase() === value2.toLowerCase();
}
exports.isEqualCaseInsensitive = isEqualCaseInsensitive;
