// @metamask/connect-multichain/dist/src/multichain/transports/multichainApiClientWrapper/index.js
/* eslint-disable no-restricted-syntax -- Private class properties use established patterns */
/* eslint-disable @typescript-eslint/explicit-function-return-type -- Inferred types are sufficient */
/* eslint-disable @typescript-eslint/parameter-properties -- Constructor shorthand is intentional */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _MultichainApiClientWrapperTransport_instances, _MultichainApiClientWrapperTransport_notificationCallbacks, _MultichainApiClientWrapperTransport_walletCreateSession, _MultichainApiClientWrapperTransport_walletGetSession, _MultichainApiClientWrapperTransport_walletRevokeSession, _MultichainApiClientWrapperTransport_walletInvokeMethod;
import { providerErrors } from '@metamask/rpc-errors';
import { getUniqueRequestId } from '../../utils';
export class MultichainApiClientWrapperTransport {
    constructor(metamaskConnectMultichain) {
        _MultichainApiClientWrapperTransport_instances.add(this);
        this.metamaskConnectMultichain = metamaskConnectMultichain;
        _MultichainApiClientWrapperTransport_notificationCallbacks.set(this, new Set());
    }
    isTransportDefined() {
        try {
            return Boolean(this.metamaskConnectMultichain.transport);
        }
        catch (_error) {
            return false;
        }
    }
    clearNotificationCallbacks() {
        __classPrivateFieldGet(this, _MultichainApiClientWrapperTransport_notificationCallbacks, "f").clear();
    }
    notifyCallbacks(data) {
        __classPrivateFieldGet(this, _MultichainApiClientWrapperTransport_notificationCallbacks, "f").forEach((callback) => {
            callback(data);
        });
    }
    clearTransportNotificationListener() {
        var _a;
        (_a = this.notificationListener) === null || _a === void 0 ? void 0 : _a.call(this);
        this.notificationListener = undefined;
    }
    setupTransportNotificationListener() {
        if (!this.isTransportDefined() || this.notificationListener) {
            return;
        }
        this.notificationListener =
            this.metamaskConnectMultichain.transport.onNotification(this.notifyCallbacks.bind(this));
    }
    connect() {
        return __awaiter(this, void 0, void 0, function* () {
            console.log('📚 connect');
            yield this.metamaskConnectMultichain.emitSessionChanged();
        });
    }
    disconnect() {
        return __awaiter(this, void 0, void 0, function* () {
            return Promise.resolve();
        });
    }
    isConnected() {
        return true;
    }
    request(params_1) {
        return __awaiter(this, arguments, void 0, function* (params, _options = {}) {
            const id = getUniqueRequestId();
            const requestPayload = Object.assign({ id, jsonrpc: '2.0' }, params);
            switch (requestPayload.method) {
                case 'wallet_createSession':
                    return __classPrivateFieldGet(this, _MultichainApiClientWrapperTransport_instances, "m", _MultichainApiClientWrapperTransport_walletCreateSession).call(this, requestPayload);
                case 'wallet_getSession':
                    return __classPrivateFieldGet(this, _MultichainApiClientWrapperTransport_instances, "m", _MultichainApiClientWrapperTransport_walletGetSession).call(this, requestPayload);
                case 'wallet_revokeSession':
                    return __classPrivateFieldGet(this, _MultichainApiClientWrapperTransport_instances, "m", _MultichainApiClientWrapperTransport_walletRevokeSession).call(this, requestPayload);
                case 'wallet_invokeMethod':
                    return __classPrivateFieldGet(this, _MultichainApiClientWrapperTransport_instances, "m", _MultichainApiClientWrapperTransport_walletInvokeMethod).call(this, requestPayload);
                default:
                    throw new Error(`Unsupported method: ${requestPayload.method}`);
            }
            throw new Error(`Unknown method: ${requestPayload.method}`);
        });
    }
    onNotification(callback) {
        this.setupTransportNotificationListener();
        __classPrivateFieldGet(this, _MultichainApiClientWrapperTransport_notificationCallbacks, "f").add(callback);
        return () => {
            __classPrivateFieldGet(this, _MultichainApiClientWrapperTransport_notificationCallbacks, "f").delete(callback);
        };
    }
}
_MultichainApiClientWrapperTransport_notificationCallbacks = new WeakMap(), _MultichainApiClientWrapperTransport_instances = new WeakSet(), _MultichainApiClientWrapperTransport_walletCreateSession = function _MultichainApiClientWrapperTransport_walletCreateSession(request) {
    return __awaiter(this, void 0, void 0, function* () {
        const createSessionParams = request.params;
        const scopes = Object.keys(Object.assign(Object.assign({}, createSessionParams.optionalScopes), createSessionParams.requiredScopes));
        const scopeAccounts = [];
        scopes.forEach((scope) => {
            var _a, _b, _c, _d;
            const requiredScope = (_a = createSessionParams.requiredScopes) === null || _a === void 0 ? void 0 : _a[scope];
            const optionalScope = (_b = createSessionParams.optionalScopes) === null || _b === void 0 ? void 0 : _b[scope];
            if (requiredScope) {
                scopeAccounts.push(...((_c = requiredScope.accounts) !== null && _c !== void 0 ? _c : []));
            }
            if (optionalScope) {
                scopeAccounts.push(...((_d = optionalScope.accounts) !== null && _d !== void 0 ? _d : []));
            }
        });
        const accounts = [...new Set(scopeAccounts)];
        console.log('📚 SDK connect');
        yield this.metamaskConnectMultichain.connect(scopes, accounts, createSessionParams.sessionProperties);
        console.log('📚 SDK connected');
        return this.metamaskConnectMultichain.transport.request({
            method: 'wallet_getSession',
        });
    });
}, _MultichainApiClientWrapperTransport_walletGetSession = function _MultichainApiClientWrapperTransport_walletGetSession(request) {
    return __awaiter(this, void 0, void 0, function* () {
        if (!this.isTransportDefined()) {
            return {
                jsonrpc: '2.0',
                id: request.id,
                result: {
                    sessionScopes: {},
                },
            };
        }
        return this.metamaskConnectMultichain.transport.request({
            method: 'wallet_getSession',
        });
    });
}, _MultichainApiClientWrapperTransport_walletRevokeSession = function _MultichainApiClientWrapperTransport_walletRevokeSession(request) {
    return __awaiter(this, void 0, void 0, function* () {
        var _a;
        if (!this.isTransportDefined()) {
            return { jsonrpc: '2.0', id: request.id, result: true };
        }
        const revokeSessionParams = request.params;
        const scopes = (_a = revokeSessionParams === null || revokeSessionParams === void 0 ? void 0 : revokeSessionParams.scopes) !== null && _a !== void 0 ? _a : [];
        try {
            yield this.metamaskConnectMultichain.disconnect(scopes);
            return { jsonrpc: '2.0', id: request.id, result: true };
        }
        catch (_error) {
            return { jsonrpc: '2.0', id: request.id, result: false };
        }
    });
}, _MultichainApiClientWrapperTransport_walletInvokeMethod = function _MultichainApiClientWrapperTransport_walletInvokeMethod(request) {
    return __awaiter(this, void 0, void 0, function* () {
        if (!this.isTransportDefined()) {
            return { error: providerErrors.unauthorized() };
        }
        const result = this.metamaskConnectMultichain.invokeMethod(request.params);
        return {
            result,
        };
    });
};
