// @metamask/connect-evm/dist/browser/es/connect-evm.mjs
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __typeError = (msg) => {
  throw TypeError(msg);
};
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __accessCheck = (obj, member, msg) => member.has(obj) || __typeError("Cannot " + msg);
var __privateGet = (obj, member, getter) => (__accessCheck(obj, member, "read from private field"), getter ? getter.call(obj) : member.get(obj));
var __privateAdd = (obj, member, value) => member.has(obj) ? __typeError("Cannot add the same private member more than once") : member instanceof WeakSet ? member.add(obj) : member.set(obj, value);
var __privateSet = (obj, member, value, setter) => (__accessCheck(obj, member, "write to private field"), setter ? setter.call(obj, value) : member.set(obj, value), value);
var __privateMethod = (obj, member, method) => (__accessCheck(obj, member, "access private method"), method);
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};

// src/utils/infura.ts
import { getInfuraRpcUrls as getInfuraRpcUrlsMultichain } from "@metamask/connect-multichain";
import {
  hexToNumber,
  KnownCaipNamespace,
  numberToHex,
  parseCaipChainId
} from "@metamask/utils";
var getInfuraRpcUrls = ({
  infuraApiKey,
  chainIds
}) => {
  const caipChainIds = chainIds == null ? void 0 : chainIds.map(
    (chainId) => `eip155:${hexToNumber(chainId)}`
  );
  const caipMap = getInfuraRpcUrlsMultichain({
    infuraApiKey,
    caipChainIds
  });
  return Object.entries(caipMap).reduce(
    (acc, [key, url]) => {
      const { namespace: namespace2, reference } = parseCaipChainId(key);
      if (namespace2 !== KnownCaipNamespace.Eip155) {
        return acc;
      }
      const chainId = numberToHex(parseInt(reference, 10));
      acc[chainId] = url;
      return acc;
    },
    {}
  );
};

// src/connect.ts
import { analytics } from "@metamask/analytics";
import { parseScopeString } from "@metamask/chain-agnostic-permission";
import {
  createMultichainClient,
  getWalletActionAnalyticsProperties,
  isRejectionError,
  TransportType
} from "@metamask/connect-multichain";
import { createDeferredPromise, hexToNumber as hexToNumber3 } from "@metamask/utils";

// src/constants.ts
var IGNORED_METHODS = [
  "metamask_getProviderState",
  "metamask_sendDomainMetadata",
  "metamask_logWeb3ShimUsage",
  "wallet_registerOnboarding",
  "net_version",
  "wallet_getPermissions"
];
var CONNECT_METHODS = [
  "wallet_requestPermissions",
  "eth_requestAccounts"
];
var ACCOUNTS_METHODS = ["eth_accounts", "eth_coinbase"];
var CHAIN_METHODS = ["eth_chainId"];
var INTERCEPTABLE_METHODS = [
  ...ACCOUNTS_METHODS,
  ...IGNORED_METHODS,
  ...CONNECT_METHODS,
  ...CHAIN_METHODS,
  // These have bespoke handlers
  "wallet_revokePermissions",
  "wallet_switchEthereumChain",
  "wallet_addEthereumChain"
];

// src/logger.ts
import {
  createLogger,
  enableDebug as debug
} from "@metamask/connect-multichain";
var namespace = "metamask-connect:evm";
var logger = createLogger(namespace, "63");
var enableDebug = () => {
  debug(namespace);
};

// src/provider.ts
import { EventEmitter, RPCInvokeMethodErr } from "@metamask/connect-multichain";
import { hexToNumber as hexToNumber2 } from "@metamask/utils";
var _core, _requestInterceptor, _accounts, _selectedChainId;
var EIP1193Provider = class extends EventEmitter {
  constructor(core, interceptor) {
    super();
    /** The core instance of the Multichain SDK */
    __privateAdd(this, _core);
    /** Interceptor function to handle specific methods */
    __privateAdd(this, _requestInterceptor);
    /** The currently permitted accounts */
    __privateAdd(this, _accounts, []);
    /** The currently selected chain ID on the wallet */
    __privateAdd(this, _selectedChainId);
    __privateSet(this, _core, core);
    __privateSet(this, _requestInterceptor, interceptor);
    this.request = this.request.bind(this);
    this.sendAsync = this.sendAsync.bind(this);
    this.send = this.send.bind(this);
    this.on = this.on.bind(this);
    this.off = this.off.bind(this);
    this.emit = this.emit.bind(this);
    this.once = this.once.bind(this);
    this.removeListener = this.removeListener.bind(this);
    this.listenerCount = this.listenerCount.bind(this);
  }
  /**
   * Performs a EIP-1193 request.
   *
   * @param request - The request object containing the method and params
   * @returns The result of the request
   */
  request(request) {
    return __async(this, null, function* () {
      var _a, _b, _c, _d;
      logger(
        `request: ${request.method} - chainId: ${this.selectedChainId}`,
        request.params
      );
      if (INTERCEPTABLE_METHODS.includes(request.method)) {
        return (_a = __privateGet(this, _requestInterceptor)) == null ? void 0 : _a.call(this, request);
      }
      if (!__privateGet(this, _selectedChainId)) {
        throw new Error("No chain ID selected");
      }
      const decimalChainId = hexToNumber2(__privateGet(this, _selectedChainId));
      const scope = `eip155:${decimalChainId}`;
      const coreOptions = __privateGet(this, _core).options;
      const supportedNetworks = (_c = (_b = coreOptions == null ? void 0 : coreOptions.api) == null ? void 0 : _b.supportedNetworks) != null ? _c : {};
      if (!supportedNetworks[scope]) {
        throw new Error(
          `Chain ${scope} is not configured in supportedNetworks. Requests cannot be made to chains not explicitly configured in supportedNetworks.`
        );
      }
      try {
        return yield __privateGet(this, _core).invokeMethod({
          scope,
          request: {
            method: request.method,
            params: request.params
          }
        });
      } catch (error) {
        if (error instanceof RPCInvokeMethodErr && error.rpcCode !== void 0) {
          const rpcError = new Error(
            (_d = error.rpcMessage) != null ? _d : error.reason
          );
          rpcError.code = error.rpcCode;
          throw rpcError;
        }
        throw error;
      }
    });
  }
  // Getters and setters
  get selectedAccount() {
    return this.accounts[0];
  }
  set accounts(accounts) {
    __privateSet(this, _accounts, accounts);
  }
  get accounts() {
    return __privateGet(this, _accounts);
  }
  get selectedChainId() {
    return __privateGet(this, _selectedChainId);
  }
  set selectedChainId(chainId) {
    __privateSet(this, _selectedChainId, chainId);
  }
  // ==========================================
  // Legacy compatibility methods
  // ==========================================
  /**
   * Alias for selectedChainId for legacy compatibility.
   * Many dApps expect a `chainId` property on the provider.
   */
  get chainId() {
    return this.selectedChainId;
  }
  /**
   * Legacy method for sending JSON-RPC requests.
   *
   * @deprecated Use `request` instead. This method is provided for backwards compatibility.
   * @param request - The JSON-RPC request object
   * @param callback - Optional callback function. If provided, the method returns void.
   * @returns A promise resolving to the JSON-RPC response, or void if a callback is provided.
   */
  sendAsync(request, callback) {
    return __async(this, null, function* () {
      var _a;
      const id = (_a = request.id) != null ? _a : 1;
      const promise = this.request({
        method: request.method,
        params: request.params
      }).then(
        (result) => ({
          id,
          jsonrpc: "2.0",
          result
        })
      ).catch(
        (error) => {
          var _a2, _b;
          return {
            id,
            jsonrpc: "2.0",
            error: {
              code: (_a2 = error.code) != null ? _a2 : -32603,
              message: (_b = error.message) != null ? _b : "Internal error",
              data: error.data
            }
          };
        }
      );
      if (callback) {
        promise.then((response) => {
          if (response.error) {
            callback(new Error(response.error.message), response);
          } else {
            callback(null, response);
          }
        }).catch((error) => {
          callback(error, null);
        });
        return;
      }
      return promise;
    });
  }
  /**
   * Legacy method for sending JSON-RPC requests synchronously (callback-based).
   *
   * @deprecated Use `request` instead. This method is provided for backwards compatibility.
   * @param request - The JSON-RPC request object
   * @param callback - The callback function to receive the response
   */
  send(request, callback) {
    this.sendAsync(request, callback);
  }
};
_core = new WeakMap();
_requestInterceptor = new WeakMap();
_accounts = new WeakMap();
_selectedChainId = new WeakMap();

// src/utils/caip.ts
import {
  getPermittedEthChainIds as _getPermittedEthChainIds,
  getEthAccounts as _getEthAccounts
} from "@metamask/chain-agnostic-permission";
var getEthAccounts = (sessionScopes) => {
  if (!sessionScopes) {
    return [];
  }
  return _getEthAccounts({
    requiredScopes: sessionScopes,
    optionalScopes: sessionScopes
  });
};
var getPermittedEthChainIds = (sessionScopes) => {
  if (!sessionScopes) {
    return [];
  }
  return _getPermittedEthChainIds({
    requiredScopes: sessionScopes,
    optionalScopes: sessionScopes
  });
};

// src/utils/type-guards.ts
function isConnectRequest(req) {
  return req.method === "wallet_requestPermissions" || req.method === "eth_requestAccounts";
}
function isSwitchChainRequest(req) {
  return req.method === "wallet_switchEthereumChain";
}
function isAddChainRequest(req) {
  return req.method === "wallet_addEthereumChain";
}
function isAccountsRequest(req) {
  return req.method === "eth_accounts" || req.method === "eth_coinbase";
}
function isChainIdRequest(req) {
  return req.method === "eth_chainId";
}
function validSupportedChainsUrls(record, recordName) {
  const invalidUrls = [];
  for (const [key, url] of Object.entries(record)) {
    try {
      new URL(url);
    } catch (e) {
      invalidUrls.push(`${key}: ${url}`);
    }
  }
  if (invalidUrls.length > 0) {
    throw new Error(
      `${recordName} contains invalid URLs:
${invalidUrls.join("\n")}`
    );
  }
}

// src/connect.ts
var DEFAULT_CHAIN_ID = "0x1";
var CHAIN_STORE_KEY = "cache_eth_chainId";
var _core2, _provider, _sessionScopes, _eventHandlers, _sessionChangedHandler, _displayUriHandler, _removeNotificationHandler, _status, _pendingPreferredChainId, _initPromise, _MetamaskConnectEVM_instances, getCoreOptions_fn, createInvokeOptions_fn, trackWalletActionRequested_fn, trackWalletActionSucceeded_fn, trackWalletActionFailed_fn, getSelectedChainId_fn, requestInterceptor_fn, clearConnectionState_fn, addEthereumChain_fn, request_fn, cacheChainId_fn, onSessionChanged_fn, onChainChanged_fn, onAccountsChanged_fn, onConnect_fn, onDisconnect_fn, onDisplayUri_fn;
var _MetamaskConnectEVM = class _MetamaskConnectEVM {
  /**
   * Creates a new MetamaskConnectEVM instance.
   * Use the static `create()` method instead to ensure proper async initialization.
   *
   * @param options - The options for the MetamaskConnectEVM instance
   * @param options.core - The core instance of the Multichain SDK
   * @param options.eventHandlers - Optional event handlers for EIP-1193 provider events
   */
  constructor({ core, eventHandlers }) {
    __privateAdd(this, _MetamaskConnectEVM_instances);
    /** The core instance of the Multichain SDK */
    __privateAdd(this, _core2);
    /** An instance of the EIP-1193 provider interface */
    __privateAdd(this, _provider);
    /** The session scopes currently permitted */
    __privateAdd(this, _sessionScopes, {});
    /** Optional event handlers for the EIP-1193 provider events. */
    __privateAdd(this, _eventHandlers);
    /** The handler for the wallet_sessionChanged event */
    __privateAdd(this, _sessionChangedHandler);
    /** The handler for the display_uri event */
    __privateAdd(this, _displayUriHandler);
    /** The clean-up function for the notification handler */
    __privateAdd(this, _removeNotificationHandler);
    /** The current connection status */
    __privateAdd(this, _status, "disconnected");
    /**
     * The preferred chain ID to use for the active connect() call.
     * Set to the first explicit chainId passed by the caller, cleared once the
     * connect event resolves. Takes priority over the cache in #getSelectedChainId
     * so that an explicit chainIds request always wins over a prior cached value.
     */
    __privateAdd(this, _pendingPreferredChainId);
    /** Deferred that resolves once #onSessionChanged has completed for the first time. */
    __privateAdd(this, _initPromise);
    __privateSet(this, _core2, core);
    __privateSet(this, _provider, new EIP1193Provider(
      core,
      __privateMethod(this, _MetamaskConnectEVM_instances, requestInterceptor_fn).bind(this)
    ));
    __privateSet(this, _eventHandlers, eventHandlers);
    __privateSet(this, _sessionChangedHandler, __privateMethod(this, _MetamaskConnectEVM_instances, onSessionChanged_fn).bind(this));
    __privateGet(this, _core2).on("wallet_sessionChanged", __privateGet(this, _sessionChangedHandler));
    __privateSet(this, _displayUriHandler, __privateMethod(this, _MetamaskConnectEVM_instances, onDisplayUri_fn).bind(this));
    __privateGet(this, _core2).on("display_uri", __privateGet(this, _displayUriHandler));
    __privateSet(this, _initPromise, createDeferredPromise());
    logger("Connect/EVM constructor completed");
  }
  /**
   * Creates a fully initialized MetamaskConnectEVM instance.
   * This is the recommended way to instantiate the class, as it ensures
   * all async initialization (like session recovery) completes before
   * the instance is returned.
   *
   * @param options - The options for the MetamaskConnectEVM instance
   * @param options.core - The core instance of the Multichain SDK
   * @param options.eventHandlers - Optional event handlers for EIP-1193 provider events
   * @returns A promise that resolves with a fully initialized MetamaskConnectEVM instance
   */
  static create(options) {
    return __async(this, null, function* () {
      const instance = new _MetamaskConnectEVM(options);
      yield __privateGet(instance, _core2).emitSessionChanged();
      yield __privateGet(instance, _initPromise).promise;
      return instance;
    });
  }
  /**
   * Connects to the wallet with the specified chain ID and optional account.
   *
   * @param options - The connection options
   * @param [options.account] - Optional param to specify an account to connect to
   * @param [options.forceRequest] - Optional param to force a connection request regardless of whether there is a pre-existing session
   * @param [options.chainIds] - Array of chain IDs to request permission for (defaults to ethereum mainnet). The first entry is used as the active chain returned by the call. Ethereum mainnet is always included in the permission request as a bootstrap fallback.
   * @returns A promise that resolves with the connected accounts and chain ID
   */
  connect() {
    return __async(this, arguments, function* ({
      account,
      forceRequest,
      chainIds = [DEFAULT_CHAIN_ID]
    } = {}) {
      var _a;
      logger("request: connect", { account });
      if (!chainIds || chainIds.length === 0) {
        throw new Error("chainIds must be an array of at least one chain ID");
      }
      __privateSet(this, _pendingPreferredChainId, chainIds[0]);
      const caipChainIds = Array.from(
        new Set((_a = chainIds.concat(DEFAULT_CHAIN_ID)) != null ? _a : [DEFAULT_CHAIN_ID])
      ).map((id) => `eip155:${hexToNumber3(id)}`);
      const caipAccountIds = account ? caipChainIds.map((caipChainId) => `${caipChainId}:${account}`) : [];
      __privateSet(this, _status, "connecting");
      try {
        const result = new Promise(
          (resolve) => {
            __privateGet(this, _provider).once("connect", ({ chainId, accounts }) => {
              logger("fulfilled-request: connect", {
                chainId,
                accounts
              });
              resolve({
                accounts,
                chainId
              });
            });
          }
        );
        yield __privateGet(this, _core2).connect(
          caipChainIds,
          caipAccountIds,
          void 0,
          forceRequest
        );
        return yield result;
      } catch (error) {
        __privateSet(this, _status, "disconnected");
        logger("Error connecting to wallet", error);
        throw error;
      } finally {
        __privateSet(this, _pendingPreferredChainId, void 0);
      }
    });
  }
  /**
   * Connects to the wallet and signs a message using personal_sign.
   *
   * @param options - The connection options
   * @param options.message - The message to sign after connecting
   * @param [options.chainIds] - Optional hex chain IDs to connect to (defaults to ethereum mainnet if not provided)
   * @returns A promise that resolves with the connected accounts, chainId, and signature
   * @throws Error if the selected account is not available after timeout
   */
  connectAndSign(_0) {
    return __async(this, arguments, function* ({
      message,
      chainIds
    }) {
      var _a, _b;
      const { accounts, chainId } = yield this.connect({
        chainIds: chainIds != null ? chainIds : [DEFAULT_CHAIN_ID]
      });
      const signature = yield __privateGet(this, _provider).request({
        method: "personal_sign",
        params: [accounts[0], message]
      });
      (_b = (_a = __privateGet(this, _eventHandlers)) == null ? void 0 : _a.connectAndSign) == null ? void 0 : _b.call(_a, {
        accounts,
        chainId,
        signature
      });
      return { accounts, chainId, signature };
    });
  }
  /**
   * Connects to the wallet and invokes a method with specified parameters.
   *
   * @param options - The options for connecting and invoking the method
   * @param options.method - The method name to invoke
   * @param options.params - The parameters to pass to the method, or a function that receives the account and returns params
   * @param [options.chainIds] - Optional hex chain IDs to connect to (defaults to ethereum mainnet if not provided)
   * @param [options.account] - Optional specific account to connect to
   * @param [options.forceRequest] - Whether to force a request regardless of an existing session
   * @returns A promise that resolves with the connected accounts, chainId, and the result of the method invocation
   * @throws Error if the selected account is not available after timeout (for methods that require an account)
   */
  connectWith(_0) {
    return __async(this, arguments, function* ({
      method,
      params,
      chainIds,
      account,
      forceRequest
    }) {
      var _a, _b;
      const { accounts: connectedAccounts, chainId: connectedChainId } = yield this.connect({
        chainIds: chainIds != null ? chainIds : [DEFAULT_CHAIN_ID],
        account,
        forceRequest
      });
      const resolvedParams = typeof params === "function" ? params(connectedAccounts[0]) : params;
      const result = yield __privateGet(this, _provider).request({
        method,
        params: resolvedParams
      });
      (_b = (_a = __privateGet(this, _eventHandlers)) == null ? void 0 : _a.connectWith) == null ? void 0 : _b.call(_a, {
        accounts: connectedAccounts,
        chainId: connectedChainId,
        result
      });
      return { accounts: connectedAccounts, chainId: connectedChainId, result };
    });
  }
  /**
   * Disconnects from the wallet by revoking the session and cleaning up event listeners.
   *
   * @returns A promise that resolves when disconnection is complete
   */
  disconnect() {
    return __async(this, null, function* () {
      var _a;
      logger("request: disconnect");
      const sessionScopes = __privateGet(this, _sessionScopes);
      const eip155Scopes = Object.keys(sessionScopes).filter((scope) => {
        const { namespace: namespace2 } = parseScopeString(scope);
        return namespace2 === "eip155";
      });
      yield __privateGet(this, _core2).disconnect(eip155Scopes);
      __privateMethod(this, _MetamaskConnectEVM_instances, onDisconnect_fn).call(this);
      __privateMethod(this, _MetamaskConnectEVM_instances, clearConnectionState_fn).call(this);
      (_a = __privateGet(this, _removeNotificationHandler)) == null ? void 0 : _a.call(this);
      __privateSet(this, _removeNotificationHandler, void 0);
      logger("fulfilled-request: disconnect");
    });
  }
  /**
   * Switches the Ethereum chain. Will track state internally whenever possible.
   *
   * @param options - The options for the switch chain request
   * @param options.chainId - The chain ID to switch to
   * @param [options.chainConfiguration] - The chain configuration to use in case the chain is not present by the wallet
   * @returns A promise that resolves when the chain has been switched
   */
  switchChain(_0) {
    return __async(this, arguments, function* ({
      chainId,
      chainConfiguration
    }) {
      const method = "wallet_switchEthereumChain";
      const scope = `eip155:${hexToNumber3(chainId)}`;
      const params = [{ chainId }];
      if (this.selectedChainId === chainId) {
        return Promise.resolve();
      }
      const permittedChainIds = getPermittedEthChainIds(__privateGet(this, _sessionScopes));
      if (permittedChainIds.includes(chainId) && __privateGet(this, _core2).transportType === TransportType.MWP) {
        yield __privateMethod(this, _MetamaskConnectEVM_instances, cacheChainId_fn).call(this, chainId);
        __privateMethod(this, _MetamaskConnectEVM_instances, onChainChanged_fn).call(this, chainId);
        return Promise.resolve();
      }
      yield __privateMethod(this, _MetamaskConnectEVM_instances, trackWalletActionRequested_fn).call(this, method, scope, params);
      try {
        const result = yield __privateMethod(this, _MetamaskConnectEVM_instances, request_fn).call(this, {
          method: "wallet_switchEthereumChain",
          params
        });
        const resultWithError = result;
        if (resultWithError == null ? void 0 : resultWithError.error) {
          throw new Error(resultWithError.error.message);
        }
        yield __privateMethod(this, _MetamaskConnectEVM_instances, trackWalletActionSucceeded_fn).call(this, method, scope, params);
        if (result.result === null) {
          yield __privateMethod(this, _MetamaskConnectEVM_instances, cacheChainId_fn).call(this, chainId);
          __privateMethod(this, _MetamaskConnectEVM_instances, onChainChanged_fn).call(this, chainId);
        }
        return Promise.resolve();
      } catch (error) {
        yield __privateMethod(this, _MetamaskConnectEVM_instances, trackWalletActionFailed_fn).call(this, method, scope, params, error);
        if (error.message.includes("Unrecognized chain ID")) {
          return __privateMethod(this, _MetamaskConnectEVM_instances, addEthereumChain_fn).call(this, chainConfiguration);
        }
        throw error;
      }
    });
  }
  /**
   * Gets the EIP-1193 provider instance
   *
   * @returns The EIP-1193 provider instance
   */
  getProvider() {
    return __privateGet(this, _provider);
  }
  /**
   * Gets the currently selected chain ID on the wallet
   *
   * @returns The currently selected chain ID or undefined if no chain is selected
   */
  getChainId() {
    return this.selectedChainId;
  }
  /**
   * Gets the currently selected account on the wallet
   *
   * @returns The currently selected account or undefined if no account is selected
   */
  getAccount() {
    return __privateGet(this, _provider).selectedAccount;
  }
  // Convenience getters for the EIP-1193 provider
  /**
   * Gets the currently permitted accounts
   *
   * @returns The currently permitted accounts
   */
  get accounts() {
    return __privateGet(this, _provider).accounts;
  }
  /**
   * Gets the currently selected account on the wallet
   *
   * @returns The currently selected account or undefined if no account is selected
   */
  get selectedAccount() {
    return __privateGet(this, _provider).selectedAccount;
  }
  /**
   * Gets the currently selected chain ID on the wallet
   *
   * @returns The currently selected chain ID or undefined if no chain is selected
   */
  get selectedChainId() {
    return __privateGet(this, _provider).selectedChainId;
  }
  /**
   * Gets the current connection status
   *
   * @returns The current connection status
   */
  get status() {
    return __privateGet(this, _status);
  }
};
_core2 = new WeakMap();
_provider = new WeakMap();
_sessionScopes = new WeakMap();
_eventHandlers = new WeakMap();
_sessionChangedHandler = new WeakMap();
_displayUriHandler = new WeakMap();
_removeNotificationHandler = new WeakMap();
_status = new WeakMap();
_pendingPreferredChainId = new WeakMap();
_initPromise = new WeakMap();
_MetamaskConnectEVM_instances = new WeakSet();
/**
 * Gets the core options for analytics checks.
 *
 * @returns The multichain options from the core instance
 */
getCoreOptions_fn = function() {
  return __privateGet(this, _core2).options;
};
/**
 * Creates invoke options for analytics tracking.
 *
 * @param method - The RPC method name
 * @param scope - The CAIP chain ID scope
 * @param params - The method parameters
 * @returns Invoke options object for analytics
 */
createInvokeOptions_fn = function(method, scope, params) {
  return {
    scope,
    request: { method, params }
  };
};
trackWalletActionRequested_fn = function(method, scope, params) {
  return __async(this, null, function* () {
    const coreOptions = __privateMethod(this, _MetamaskConnectEVM_instances, getCoreOptions_fn).call(this);
    try {
      const invokeOptions = __privateMethod(this, _MetamaskConnectEVM_instances, createInvokeOptions_fn).call(this, method, scope, params);
      const props = yield getWalletActionAnalyticsProperties(
        coreOptions,
        __privateGet(this, _core2).storage,
        invokeOptions,
        __privateGet(this, _core2).transportType
      );
      analytics.track("mmconnect_wallet_action_requested", props);
    } catch (error) {
      logger("Error tracking mmconnect_wallet_action_requested event", error);
    }
  });
};
trackWalletActionSucceeded_fn = function(method, scope, params) {
  return __async(this, null, function* () {
    const coreOptions = __privateMethod(this, _MetamaskConnectEVM_instances, getCoreOptions_fn).call(this);
    try {
      const invokeOptions = __privateMethod(this, _MetamaskConnectEVM_instances, createInvokeOptions_fn).call(this, method, scope, params);
      const props = yield getWalletActionAnalyticsProperties(
        coreOptions,
        __privateGet(this, _core2).storage,
        invokeOptions,
        __privateGet(this, _core2).transportType
      );
      analytics.track("mmconnect_wallet_action_succeeded", props);
    } catch (error) {
      logger("Error tracking mmconnect_wallet_action_succeeded event", error);
    }
  });
};
trackWalletActionFailed_fn = function(method, scope, params, error) {
  return __async(this, null, function* () {
    const coreOptions = __privateMethod(this, _MetamaskConnectEVM_instances, getCoreOptions_fn).call(this);
    try {
      const invokeOptions = __privateMethod(this, _MetamaskConnectEVM_instances, createInvokeOptions_fn).call(this, method, scope, params);
      const props = yield getWalletActionAnalyticsProperties(
        coreOptions,
        __privateGet(this, _core2).storage,
        invokeOptions,
        __privateGet(this, _core2).transportType
      );
      const isRejection = isRejectionError(error);
      if (isRejection) {
        analytics.track("mmconnect_wallet_action_rejected", props);
      } else {
        analytics.track("mmconnect_wallet_action_failed", props);
      }
    } catch (e) {
      logger("Error tracking wallet action rejected or failed event", error);
    }
  });
};
getSelectedChainId_fn = function(permittedChainIds) {
  return __async(this, null, function* () {
    if (__privateGet(this, _pendingPreferredChainId) && permittedChainIds.includes(__privateGet(this, _pendingPreferredChainId))) {
      return __privateGet(this, _pendingPreferredChainId);
    }
    try {
      const cachedChainId = yield __privateGet(this, _core2).storage.adapter.get(CHAIN_STORE_KEY);
      if (cachedChainId) {
        const chainId = JSON.parse(cachedChainId);
        if (permittedChainIds.includes(chainId)) {
          return chainId;
        }
      }
    } catch (error) {
      logger("Error retrieving cached chainId", error);
    }
    return permittedChainIds[0];
  });
};
requestInterceptor_fn = function(request) {
  return __async(this, null, function* () {
    logger(`Intercepting request for method: ${request.method}`);
    if (IGNORED_METHODS.includes(request.method)) {
      return Promise.reject(
        new Error(
          `Method: ${request.method} is not supported by Metamask Connect/EVM`
        )
      );
    }
    if (request.method === "wallet_revokePermissions") {
      return this.disconnect();
    }
    if (isConnectRequest(request)) {
      const shouldForceConnectionRequest = request.method === "wallet_requestPermissions";
      const { method, params } = request;
      const permitted = getPermittedEthChainIds(__privateGet(this, _sessionScopes));
      if (permitted.length === 0) {
        permitted.push(DEFAULT_CHAIN_ID);
      }
      const selected = __privateGet(this, _provider).selectedChainId;
      const preferred = selected && permitted.includes(selected) ? selected : permitted[0];
      const chainIds = [
        preferred,
        ...permitted.filter((id) => id !== preferred)
      ];
      const scope = `eip155:${hexToNumber3(preferred)}`;
      yield __privateMethod(this, _MetamaskConnectEVM_instances, trackWalletActionRequested_fn).call(this, method, scope, params);
      try {
        const result = yield this.connect({
          chainIds,
          forceRequest: shouldForceConnectionRequest
        });
        yield __privateMethod(this, _MetamaskConnectEVM_instances, trackWalletActionSucceeded_fn).call(this, method, scope, params);
        return result;
      } catch (error) {
        yield __privateMethod(this, _MetamaskConnectEVM_instances, trackWalletActionFailed_fn).call(this, method, scope, params, error);
        throw error;
      }
    }
    if (isSwitchChainRequest(request)) {
      return this.switchChain({
        chainId: request.params[0].chainId
      });
    }
    if (isAddChainRequest(request)) {
      return __privateMethod(this, _MetamaskConnectEVM_instances, addEthereumChain_fn).call(this, request.params[0]);
    }
    if (isAccountsRequest(request)) {
      return __privateGet(this, _provider).accounts;
    }
    if (isChainIdRequest(request)) {
      return __privateGet(this, _provider).selectedChainId;
    }
    logger("Request not intercepted, forwarding to default handler", request);
    return Promise.resolve();
  });
};
/**
 * Clears the internal connection state: accounts and chainId
 */
clearConnectionState_fn = function() {
  __privateGet(this, _provider).accounts = [];
  __privateGet(this, _provider).selectedChainId = void 0;
};
addEthereumChain_fn = function(chainConfiguration) {
  return __async(this, null, function* () {
    logger("addEthereumChain called", { chainConfiguration });
    const method = "wallet_addEthereumChain";
    if (!chainConfiguration) {
      throw new Error("No chain configuration found.");
    }
    const chainId = chainConfiguration.chainId || __privateGet(this, _provider).selectedChainId || "0x1";
    const decimalChainId = hexToNumber3(chainId);
    const scope = `eip155:${decimalChainId}`;
    const params = [chainConfiguration];
    yield __privateMethod(this, _MetamaskConnectEVM_instances, trackWalletActionRequested_fn).call(this, method, scope, params);
    try {
      const result = yield __privateMethod(this, _MetamaskConnectEVM_instances, request_fn).call(this, {
        method: "wallet_addEthereumChain",
        params
      });
      if (result.result === null) {
        yield __privateMethod(this, _MetamaskConnectEVM_instances, cacheChainId_fn).call(this, chainId);
        __privateMethod(this, _MetamaskConnectEVM_instances, onChainChanged_fn).call(this, chainId);
      }
      yield __privateMethod(this, _MetamaskConnectEVM_instances, trackWalletActionSucceeded_fn).call(this, method, scope, params);
    } catch (error) {
      yield __privateMethod(this, _MetamaskConnectEVM_instances, trackWalletActionFailed_fn).call(this, method, scope, params, error);
      throw error;
    }
  });
};
request_fn = function(request) {
  return __async(this, null, function* () {
    logger("direct request to metamask-provider called", request);
    const result = __privateGet(this, _core2).transport.sendEip1193Message(request);
    if (request.method === "wallet_addEthereumChain" || request.method === "wallet_switchEthereumChain") {
      __privateGet(this, _core2).openSimpleDeeplinkIfNeeded();
    }
    return result;
  });
};
cacheChainId_fn = function(chainId) {
  return __async(this, null, function* () {
    try {
      yield __privateGet(this, _core2).storage.adapter.set(
        CHAIN_STORE_KEY,
        JSON.stringify(chainId)
      );
    } catch (error) {
      logger("Error caching chainId", error);
    }
  });
};
onSessionChanged_fn = function(session) {
  return __async(this, null, function* () {
    var _a;
    logger("event: wallet_sessionChanged", session);
    __privateSet(this, _sessionScopes, (_a = session == null ? void 0 : session.sessionScopes) != null ? _a : {});
    const hexPermittedChainIds = getPermittedEthChainIds(__privateGet(this, _sessionScopes));
    if (hexPermittedChainIds.length === 0) {
      __privateMethod(this, _MetamaskConnectEVM_instances, onDisconnect_fn).call(this);
    } else {
      let initialAccounts = [];
      if (__privateGet(this, _core2).status === "connected") {
        const ethAccountsResponse = yield __privateGet(this, _core2).transport.sendEip1193Message({
          method: "eth_accounts",
          params: []
        });
        initialAccounts = ethAccountsResponse.result;
      } else {
        initialAccounts = getEthAccounts(__privateGet(this, _sessionScopes));
      }
      const chainId = yield __privateMethod(this, _MetamaskConnectEVM_instances, getSelectedChainId_fn).call(this, hexPermittedChainIds);
      __privateMethod(this, _MetamaskConnectEVM_instances, onConnect_fn).call(this, {
        chainId,
        accounts: initialAccounts
      });
    }
    __privateGet(this, _initPromise).resolve();
  });
};
/**
 * Handles chain change events and updates the provider's selected chain ID.
 *
 * @param chainId - The new hex chain ID
 */
onChainChanged_fn = function(chainId) {
  var _a, _b;
  if (chainId === __privateGet(this, _provider).selectedChainId) {
    return;
  }
  logger("handler: chainChanged", { chainId });
  __privateGet(this, _provider).selectedChainId = chainId;
  __privateGet(this, _provider).emit("chainChanged", chainId);
  (_b = (_a = __privateGet(this, _eventHandlers)) == null ? void 0 : _a.chainChanged) == null ? void 0 : _b.call(_a, chainId);
};
/**
 * Handles accounts change events and updates the provider's accounts list.
 *
 * @param accounts - The new list of permitted accounts
 */
onAccountsChanged_fn = function(accounts) {
  var _a, _b;
  const accountsUnchanged = accounts.length === __privateGet(this, _provider).accounts.length && accounts.every((acct, idx) => acct === __privateGet(this, _provider).accounts[idx]);
  if (accountsUnchanged) {
    return;
  }
  logger("handler: accountsChanged", accounts);
  __privateGet(this, _provider).accounts = accounts;
  __privateGet(this, _provider).emit("accountsChanged", accounts);
  (_b = (_a = __privateGet(this, _eventHandlers)) == null ? void 0 : _a.accountsChanged) == null ? void 0 : _b.call(_a, accounts);
};
/**
 * Handles connection events and emits the connect event to listeners.
 *
 * @param options - The connection options
 * @param options.chainId - The hex chain ID of the connection
 * @param options.accounts - The accounts of the connection
 */
onConnect_fn = function({
  chainId,
  accounts
}) {
  var _a, _b, _c, _d, _e, _f, _g;
  logger("handler: connect", { chainId, accounts });
  const data = {
    chainId,
    accounts
  };
  if (__privateGet(this, _status) === "connected") {
    __privateMethod(this, _MetamaskConnectEVM_instances, onChainChanged_fn).call(this, chainId);
    __privateMethod(this, _MetamaskConnectEVM_instances, onAccountsChanged_fn).call(this, accounts);
    return;
  }
  __privateGet(this, _provider).selectedChainId = chainId;
  __privateGet(this, _provider).accounts = accounts;
  __privateSet(this, _status, "connected");
  __privateGet(this, _provider).emit("connect", data);
  (_b = (_a = __privateGet(this, _eventHandlers)) == null ? void 0 : _a.connect) == null ? void 0 : _b.call(_a, data);
  __privateGet(this, _provider).emit("chainChanged", chainId);
  (_d = (_c = __privateGet(this, _eventHandlers)) == null ? void 0 : _c.chainChanged) == null ? void 0 : _d.call(_c, chainId);
  __privateGet(this, _provider).emit("accountsChanged", accounts);
  (_f = (_e = __privateGet(this, _eventHandlers)) == null ? void 0 : _e.accountsChanged) == null ? void 0 : _f.call(_e, accounts);
  (_g = __privateGet(this, _removeNotificationHandler)) == null ? void 0 : _g.call(this);
  const onAccountsChanged = (accs) => {
    logger("core-event: accountsChanged", accs);
    __privateMethod(this, _MetamaskConnectEVM_instances, onAccountsChanged_fn).call(this, accs);
  };
  const onChainChanged = (chainChanged) => {
    logger("core-event: chainChanged", chainChanged.chainId);
    __privateMethod(this, _MetamaskConnectEVM_instances, cacheChainId_fn).call(this, chainChanged.chainId).catch((error) => {
      logger("Error caching chainId in notification handler", error);
    });
    __privateMethod(this, _MetamaskConnectEVM_instances, onChainChanged_fn).call(this, chainChanged.chainId);
  };
  __privateGet(this, _core2).on("metamask_accountsChanged", onAccountsChanged);
  __privateGet(this, _core2).on("metamask_chainChanged", onChainChanged);
  __privateSet(this, _removeNotificationHandler, () => {
    __privateGet(this, _core2).off("metamask_accountsChanged", onAccountsChanged);
    __privateGet(this, _core2).off("metamask_chainChanged", onChainChanged);
  });
};
/**
 * Handles disconnection events and emits the disconnect event to listeners.
 * Also clears accounts by triggering an accountsChanged event with an empty array.
 */
onDisconnect_fn = function() {
  var _a, _b;
  if (__privateGet(this, _status) === "disconnected") {
    return;
  }
  __privateSet(this, _status, "disconnected");
  logger("handler: disconnect");
  __privateGet(this, _provider).emit("disconnect");
  (_b = (_a = __privateGet(this, _eventHandlers)) == null ? void 0 : _a.disconnect) == null ? void 0 : _b.call(_a);
  __privateMethod(this, _MetamaskConnectEVM_instances, onAccountsChanged_fn).call(this, []);
};
/**
 * Handles display_uri events and emits them to the provider.
 * This allows consumers to display their own custom QR code UI.
 *
 * @param uri - The deeplink URI to be displayed as a QR code
 */
onDisplayUri_fn = function(uri) {
  var _a, _b;
  if (__privateGet(this, _status) !== "connecting") {
    return;
  }
  logger("handler: display_uri", uri);
  __privateGet(this, _provider).emit("display_uri", uri);
  (_b = (_a = __privateGet(this, _eventHandlers)) == null ? void 0 : _a.displayUri) == null ? void 0 : _b.call(_a, uri);
};
var MetamaskConnectEVM = _MetamaskConnectEVM;
function createEVMClient(options) {
  return __async(this, null, function* () {
    var _a, _b;
    if (options.debug) {
      enableDebug();
    }
    logger("Creating Metamask Connect/EVM with options:", options);
    if (!((_a = options.api) == null ? void 0 : _a.supportedNetworks) || Object.keys(options.api.supportedNetworks).length === 0) {
      throw new Error(
        "supportedNetworks is required and must contain at least one chain configuration"
      );
    }
    validSupportedChainsUrls(options.api.supportedNetworks, "supportedNetworks");
    const supportedNetworksCaipChainId = Object.entries(
      options.api.supportedNetworks
    ).reduce((acc, [hexChainId, url]) => {
      const decimalChainId = parseInt(hexChainId, 16);
      const caip2ChainId = `eip155:${decimalChainId}`;
      acc[caip2ChainId] = url;
      return acc;
    }, {});
    try {
      const core = yield createMultichainClient(__spreadProps(__spreadValues({}, options), {
        api: {
          supportedNetworks: supportedNetworksCaipChainId
        },
        analytics: {
          // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
          integrationType: ((_b = options.analytics) == null ? void 0 : _b.integrationType) || "direct"
        },
        versions: {
          // typeof guard needed: Metro (React Native) bundles TS source directly,
          // bypassing the tsup build that substitutes __PACKAGE_VERSION__.
          "connect-evm": false ? "unknown" : "1.0.0"
        }
      }));
      return MetamaskConnectEVM.create({
        core,
        eventHandlers: options.eventHandlers,
        supportedNetworks: options.api.supportedNetworks
      });
    } catch (error) {
      console.error("Error creating Metamask Connect/EVM", error);
      throw error;
    }
  });
}
export {
  createEVMClient,
  getInfuraRpcUrls
};
