// @metamask/chain-agnostic-permission/dist/scope/assert.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.assertIsInternalScopesObject = exports.assertIsInternalScopeString = exports.assertIsExternalScopesObject = exports.assertScopesSupported = exports.assertScopeSupported = void 0;
const utils_1 = require("@metamask/utils");
const errors_1 = require("./errors.cjs");
const supported_1 = require("./supported.cjs");
/**
 * Asserts that a scope string and its associated scope object are supported.
 *
 * @param scopeString - The scope string against which to assert support.
 * @param scopeObject - The scope object against which to assert support.
 * @param hooks - An object containing the following properties:
 * @param hooks.isEvmChainIdSupported - A predicate that determines if an EVM chainID is supported.
 * @param hooks.isNonEvmScopeSupported - A predicate that determines if an non EVM scopeString is supported.
 * @param hooks.getNonEvmSupportedMethods - A function that returns the supported methods for a non EVM scope.
 */
const assertScopeSupported = (scopeString, scopeObject, { isEvmChainIdSupported, isNonEvmScopeSupported, getNonEvmSupportedMethods, }) => {
    const { methods, notifications } = scopeObject;
    if (!(0, supported_1.isSupportedScopeString)(scopeString, {
        isEvmChainIdSupported,
        isNonEvmScopeSupported,
    })) {
        throw errors_1.Caip25Errors.requestedChainsNotSupportedError();
    }
    const allMethodsSupported = methods.every((method) => (0, supported_1.isSupportedMethod)(scopeString, method, { getNonEvmSupportedMethods }));
    if (!allMethodsSupported) {
        throw errors_1.Caip25Errors.requestedMethodsNotSupportedError();
    }
    if (notifications &&
        !notifications.every((notification) => (0, supported_1.isSupportedNotification)(scopeString, notification))) {
        throw errors_1.Caip25Errors.requestedNotificationsNotSupportedError();
    }
};
exports.assertScopeSupported = assertScopeSupported;
/**
 * Asserts that all scope strings and their associated scope objects are supported.
 *
 * @param scopes - The scopes object against which to assert support.
 * @param hooks - An object containing the following properties:
 * @param hooks.isEvmChainIdSupported - A predicate that determines if an EVM chainID is supported.
 * @param hooks.isNonEvmScopeSupported - A predicate that determines if an non EVM scopeString is supported.
 * @param hooks.getNonEvmSupportedMethods - A function that returns the supported methods for a non EVM scope.
 */
const assertScopesSupported = (scopes, { isEvmChainIdSupported, isNonEvmScopeSupported, getNonEvmSupportedMethods, }) => {
    for (const [scopeString, scopeObject] of Object.entries(scopes)) {
        (0, exports.assertScopeSupported)(scopeString, scopeObject, {
            isEvmChainIdSupported,
            isNonEvmScopeSupported,
            getNonEvmSupportedMethods,
        });
    }
};
exports.assertScopesSupported = assertScopesSupported;
/**
 * Asserts that an object is a valid ExternalScopeObject.
 *
 * @param obj - The object to assert.
 */
function assertIsExternalScopeObject(obj) {
    if (typeof obj !== 'object' || obj === null) {
        throw new Error('ExternalScopeObject must be an object');
    }
    if ((0, utils_1.hasProperty)(obj, 'references')) {
        if (!Array.isArray(obj.references) ||
            !obj.references.every(utils_1.isCaipReference)) {
            throw new Error('ExternalScopeObject.references must be an array of CaipReference');
        }
    }
    if ((0, utils_1.hasProperty)(obj, 'accounts')) {
        if (!Array.isArray(obj.accounts) || !obj.accounts.every(utils_1.isCaipAccountId)) {
            throw new Error('ExternalScopeObject.accounts must be an array of CaipAccountId');
        }
    }
    if ((0, utils_1.hasProperty)(obj, 'methods')) {
        if (!Array.isArray(obj.methods) ||
            !obj.methods.every((method) => typeof method === 'string')) {
            throw new Error('ExternalScopeObject.methods must be an array of strings');
        }
    }
    if ((0, utils_1.hasProperty)(obj, 'notifications')) {
        if (!Array.isArray(obj.notifications) ||
            !obj.notifications.every((notification) => typeof notification === 'string')) {
            throw new Error('ExternalScopeObject.notifications must be an array of strings');
        }
    }
    if ((0, utils_1.hasProperty)(obj, 'rpcDocuments')) {
        if (!Array.isArray(obj.rpcDocuments) ||
            !obj.rpcDocuments.every((doc) => typeof doc === 'string')) {
            throw new Error('ExternalScopeObject.rpcDocuments must be an array of strings');
        }
    }
    if ((0, utils_1.hasProperty)(obj, 'rpcEndpoints')) {
        if (!Array.isArray(obj.rpcEndpoints) ||
            !obj.rpcEndpoints.every((endpoint) => typeof endpoint === 'string')) {
            throw new Error('ExternalScopeObject.rpcEndpoints must be an array of strings');
        }
    }
}
/**
 * Asserts that a scope string is a valid ExternalScopeString.
 *
 * @param scopeString - The scope string to assert.
 */
function assertIsExternalScopeString(scopeString) {
    if (typeof scopeString !== 'string' ||
        (!(0, utils_1.isCaipNamespace)(scopeString) && !(0, utils_1.isCaipChainId)(scopeString))) {
        throw new Error('scopeString is not a valid ExternalScopeString');
    }
}
/**
 * Asserts that an object is a valid ExternalScopesObject.
 *
 * @param obj - The object to assert.
 */
function assertIsExternalScopesObject(obj) {
    if (typeof obj !== 'object' || obj === null) {
        throw new Error('ExternalScopesObject must be an object');
    }
    for (const [scopeString, scopeObject] of Object.entries(obj)) {
        assertIsExternalScopeString(scopeString);
        assertIsExternalScopeObject(scopeObject);
    }
}
exports.assertIsExternalScopesObject = assertIsExternalScopesObject;
/**
 * Asserts that an object is a valid InternalScopeObject.
 *
 * @param obj - The object to assert.
 */
function assertIsInternalScopeObject(obj) {
    if (typeof obj !== 'object' || obj === null) {
        throw new Error('InternalScopeObject must be an object');
    }
    if (!(0, utils_1.hasProperty)(obj, 'accounts') ||
        !Array.isArray(obj.accounts) ||
        !obj.accounts.every(utils_1.isCaipAccountId)) {
        throw new Error('InternalScopeObject.accounts must be an array of CaipAccountId');
    }
}
/**
 * Asserts that a scope string is a valid InternalScopeString.
 *
 * @param scopeString - The scope string to assert.
 */
function assertIsInternalScopeString(scopeString) {
    if (typeof scopeString !== 'string' ||
        // `InternalScopeString` is defined as either `KnownCaipNamespace.Wallet` or
        // `CaipChainId`, so our conditions intentionally match the type.
        (scopeString !== utils_1.KnownCaipNamespace.Wallet && !(0, utils_1.isCaipChainId)(scopeString))) {
        throw new Error('scopeString is not a valid InternalScopeString');
    }
}
exports.assertIsInternalScopeString = assertIsInternalScopeString;
/**
 * Asserts that an object is a valid InternalScopesObject.
 *
 * @param obj - The object to assert.
 */
function assertIsInternalScopesObject(obj) {
    if (typeof obj !== 'object' || obj === null) {
        throw new Error('InternalScopesObject must be an object');
    }
    for (const [scopeString, scopeObject] of Object.entries(obj)) {
        assertIsInternalScopeString(scopeString);
        assertIsInternalScopeObject(scopeObject);
    }
}
exports.assertIsInternalScopesObject = assertIsInternalScopesObject;
