// @metamask/chain-agnostic-permission/dist/scope/constants.cjs
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isKnownSessionPropertyValue = exports.KnownSessionProperties = exports.KnownNotifications = exports.KnownWalletNamespaceRpcMethods = exports.KnownRpcMethods = exports.Eip1193OnlyMethods = exports.KnownWalletRpcMethods = exports.CaipReferenceRegexes = exports.KnownWalletScopeString = void 0;
const api_specs_1 = __importDefault(require("@metamask/api-specs"));
/**
 * ScopeStrings for offchain methods that are not specific to a chainId but are specific to a CAIP namespace.
 */
var KnownWalletScopeString;
(function (KnownWalletScopeString) {
    KnownWalletScopeString["Eip155"] = "wallet:eip155";
})(KnownWalletScopeString || (exports.KnownWalletScopeString = KnownWalletScopeString = {}));
/**
 * Regexes defining how references must be formed for non-wallet known CAIP namespaces
 */
exports.CaipReferenceRegexes = {
    eip155: /^(0|[1-9][0-9]*)$/u,
    bip122: /.*/u,
    solana: /.*/u,
    tron: /.*/u,
};
/**
 * Methods that do not belong exclusively to any CAIP namespace.
 */
exports.KnownWalletRpcMethods = [
    'wallet_registerOnboarding',
    'wallet_scanQRCode',
];
/**
 * Methods that belong to the `wallet:eip155` scope.
 */
const WalletEip155Methods = ['wallet_addEthereumChain'];
/**
 * Methods that are only supported via the EIP-1193 API.
 */
exports.Eip1193OnlyMethods = [
    'wallet_switchEthereumChain',
    'wallet_getPermissions',
    'wallet_requestPermissions',
    'wallet_revokePermissions',
    'eth_requestAccounts',
    'eth_accounts',
    'eth_coinbase',
    'net_version',
    'metamask_logWeb3ShimUsage',
    'metamask_getProviderState',
    'metamask_sendDomainMetadata',
    'wallet_registerOnboarding',
];
/**
 * All MetaMask methods, except for ones we have specified in the constants above.
 */
const Eip155Methods = api_specs_1.default.methods
    .map(({ name }) => name)
    .filter((method) => !WalletEip155Methods.includes(method))
    .filter((method) => !exports.KnownWalletRpcMethods.includes(method))
    .filter((method) => !exports.Eip1193OnlyMethods.includes(method));
/**
 * Methods by ecosystem that are chain specific.
 */
exports.KnownRpcMethods = {
    eip155: Eip155Methods,
    bip122: [],
    solana: [],
    tron: [],
};
/**
 * Methods for CAIP namespaces that aren't chain specific.
 */
exports.KnownWalletNamespaceRpcMethods = {
    eip155: WalletEip155Methods,
    bip122: [],
    solana: [],
    tron: [],
};
/**
 * Notifications for known CAIP namespaces.
 */
exports.KnownNotifications = {
    eip155: ['eth_subscription'],
    bip122: [],
    solana: [],
    tron: [],
};
/**
 * Session properties for known CAIP namespaces.
 */
var KnownSessionProperties;
(function (KnownSessionProperties) {
    KnownSessionProperties["SolanaAccountChangedNotifications"] = "solana_accountChanged_notifications";
    KnownSessionProperties["TronAccountChangedNotifications"] = "tron_accountChanged_notifications";
    KnownSessionProperties["Bip122AccountChangedNotifications"] = "bip122_accountChanged_notifications";
})(KnownSessionProperties || (exports.KnownSessionProperties = KnownSessionProperties = {}));
/**
 * Checks if a given value is a known session property.
 *
 * @param value - The value to check.
 * @returns `true` if the value is a known session property, otherwise `false`.
 */
function isKnownSessionPropertyValue(value) {
    return Object.values(KnownSessionProperties).includes(value);
}
exports.isKnownSessionPropertyValue = isKnownSessionPropertyValue;
