// @metamask/chain-agnostic-permission/dist/scope/supported.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isSupportedSessionProperty = exports.isSupportedNotification = exports.isSupportedMethod = exports.isSupportedAccount = exports.isSupportedScopeString = void 0;
const controller_utils_1 = require("@metamask/controller-utils");
const utils_1 = require("@metamask/utils");
const constants_1 = require("./constants.cjs");
const types_1 = require("./types.cjs");
/**
 * Determines if a scope string is supported.
 *
 * @param scopeString - The scope string to check.
 * @param hooks - An object containing the following properties:
 * @param hooks.isEvmChainIdSupported - A predicate that determines if an EVM chainID is supported.
 * @param hooks.isNonEvmScopeSupported - A predicate that determines if an non EVM scopeString is supported.
 * @returns A boolean indicating if the scope string is supported.
 */
const isSupportedScopeString = (scopeString, { isEvmChainIdSupported, isNonEvmScopeSupported, }) => {
    const { namespace, reference } = (0, types_1.parseScopeString)(scopeString);
    switch (namespace) {
        case utils_1.KnownCaipNamespace.Wallet:
            if ((0, utils_1.isCaipChainId)(scopeString) &&
                reference !== utils_1.KnownCaipNamespace.Eip155) {
                return isNonEvmScopeSupported(scopeString);
            }
            return true;
        case utils_1.KnownCaipNamespace.Eip155:
            return (!reference ||
                (constants_1.CaipReferenceRegexes.eip155.test(reference) &&
                    isEvmChainIdSupported((0, controller_utils_1.toHex)(reference))));
        default:
            return (0, utils_1.isCaipChainId)(scopeString)
                ? isNonEvmScopeSupported(scopeString)
                : false;
    }
};
exports.isSupportedScopeString = isSupportedScopeString;
/**
 * Determines if an account is supported by the wallet (i.e. on a keyring known to the wallet).
 *
 * @param account - The CAIP account ID to check.
 * @param hooks - An object containing the following properties:
 * @param hooks.getEvmInternalAccounts - A function that returns the EVM internal accounts.
 * @param hooks.getNonEvmAccountAddresses - A function that returns the supported CAIP-10 account addresses for a non EVM scope.
 * @returns A boolean indicating if the account is supported by the wallet.
 */
const isSupportedAccount = (account, { getEvmInternalAccounts, getNonEvmAccountAddresses, }) => {
    const { address, chainId, chain: { namespace, reference }, } = (0, utils_1.parseCaipAccountId)(account);
    const isSupportedEip155Account = () => getEvmInternalAccounts().some((internalAccount) => ['eip155:eoa', 'eip155:erc4337'].includes(internalAccount.type) &&
        (0, controller_utils_1.isEqualCaseInsensitive)(address, internalAccount.address));
    const isSupportedNonEvmAccount = () => getNonEvmAccountAddresses(chainId).includes(account);
    switch (namespace) {
        case utils_1.KnownCaipNamespace.Wallet:
            if (reference === utils_1.KnownCaipNamespace.Eip155) {
                return isSupportedEip155Account();
            }
            return isSupportedNonEvmAccount();
        case utils_1.KnownCaipNamespace.Eip155:
            return isSupportedEip155Account();
        default:
            return isSupportedNonEvmAccount();
    }
};
exports.isSupportedAccount = isSupportedAccount;
/**
 * Determines if a method is supported by the wallet.
 *
 * @param scopeString - The scope string to check.
 * @param method - The method to check.
 * @param hooks - An object containing the following properties:
 * @param hooks.getNonEvmSupportedMethods - A function that returns the supported methods for a non EVM scope.
 * @returns A boolean indicating if the method is supported by the wallet.
 */
const isSupportedMethod = (scopeString, method, { getNonEvmSupportedMethods, }) => {
    const { namespace, reference } = (0, types_1.parseScopeString)(scopeString);
    if (!namespace) {
        return false;
    }
    const isSupportedNonEvmMethod = () => (0, utils_1.isCaipChainId)(scopeString) &&
        getNonEvmSupportedMethods(scopeString).includes(method);
    if (namespace === utils_1.KnownCaipNamespace.Wallet) {
        if (!reference) {
            return constants_1.KnownWalletRpcMethods.includes(method);
        }
        if (reference === utils_1.KnownCaipNamespace.Eip155) {
            return constants_1.KnownWalletNamespaceRpcMethods[reference].includes(method);
        }
        return isSupportedNonEvmMethod();
    }
    if (namespace === utils_1.KnownCaipNamespace.Eip155) {
        return constants_1.KnownRpcMethods[namespace].includes(method);
    }
    return isSupportedNonEvmMethod();
};
exports.isSupportedMethod = isSupportedMethod;
/**
 * Determines if a notification is supported by the wallet.
 *
 * @param scopeString - The scope string to check.
 * @param notification - The notification to check.
 * @returns A boolean indicating if the notification is supported by the wallet.
 */
const isSupportedNotification = (scopeString, notification) => {
    const { namespace } = (0, types_1.parseScopeString)(scopeString);
    if (namespace === utils_1.KnownCaipNamespace.Eip155) {
        return constants_1.KnownNotifications[namespace].includes(notification);
    }
    return false;
};
exports.isSupportedNotification = isSupportedNotification;
/**
 * Determines if a session property is supported by the wallet.
 *
 * @param property - The property to check.
 * @returns A boolean indicating if the property is supported by the wallet.
 */
const isSupportedSessionProperty = (property) => {
    return Object.values(constants_1.KnownSessionProperties).includes(property);
};
exports.isSupportedSessionProperty = isSupportedSessionProperty;
