// @metamask/chain-agnostic-permission/dist/scope/transform.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.normalizeAndMergeScopes = exports.mergeInternalScopes = exports.mergeNormalizedScopes = exports.mergeScopeObject = exports.normalizeScope = exports.getUniqueArrayItems = void 0;
const lodash_1 = require("lodash");
const types_1 = require("./types.cjs");
/**
 * Returns a list of unique items
 *
 * @param list - The list of items to filter
 * @returns A list of unique items
 */
const getUniqueArrayItems = (list) => {
    return Array.from(new Set(list));
};
exports.getUniqueArrayItems = getUniqueArrayItems;
/**
 * Normalizes a ScopeString and ExternalScopeObject into a separate
 * InternalScopeString and NormalizedScopeObject for each reference in the `references`
 * value if defined and adds an empty `accounts` array if not defined.
 *
 * @param scopeString - The string representing the scope
 * @param externalScopeObject - The object that defines the scope
 * @returns a map of caipChainId to ScopeObjects
 */
const normalizeScope = (scopeString, externalScopeObject) => {
    const { references, ...scopeObject } = externalScopeObject;
    const { namespace, reference } = (0, types_1.parseScopeString)(scopeString);
    const normalizedScopeObject = {
        accounts: [],
        ...scopeObject,
    };
    const shouldFlatten = namespace &&
        !reference &&
        references !== undefined &&
        references.length > 0;
    if (shouldFlatten) {
        return Object.fromEntries(references.map((ref) => [
            `${namespace}:${ref}`,
            (0, lodash_1.cloneDeep)(normalizedScopeObject),
        ]));
    }
    return { [scopeString]: normalizedScopeObject };
};
exports.normalizeScope = normalizeScope;
/**
 * Merges two NormalizedScopeObjects
 *
 * @param scopeObjectA - The first scope object to merge.
 * @param scopeObjectB - The second scope object to merge.
 * @returns The merged scope object.
 */
const mergeScopeObject = (scopeObjectA, scopeObjectB) => {
    const mergedScopeObject = {
        methods: (0, exports.getUniqueArrayItems)([
            ...scopeObjectA.methods,
            ...scopeObjectB.methods,
        ]),
        notifications: (0, exports.getUniqueArrayItems)([
            ...scopeObjectA.notifications,
            ...scopeObjectB.notifications,
        ]),
        accounts: (0, exports.getUniqueArrayItems)([
            ...scopeObjectA.accounts,
            ...scopeObjectB.accounts,
        ]),
    };
    if (scopeObjectA.rpcDocuments || scopeObjectB.rpcDocuments) {
        mergedScopeObject.rpcDocuments = (0, exports.getUniqueArrayItems)([
            ...(scopeObjectA.rpcDocuments ?? []),
            ...(scopeObjectB.rpcDocuments ?? []),
        ]);
    }
    if (scopeObjectA.rpcEndpoints || scopeObjectB.rpcEndpoints) {
        mergedScopeObject.rpcEndpoints = (0, exports.getUniqueArrayItems)([
            ...(scopeObjectA.rpcEndpoints ?? []),
            ...(scopeObjectB.rpcEndpoints ?? []),
        ]);
    }
    return mergedScopeObject;
};
exports.mergeScopeObject = mergeScopeObject;
/**
 * Merges two NormalizedScopeObjects
 *
 * @param scopeA - The first normalized scope object to merge.
 * @param scopeB - The second normalized scope object to merge.
 * @returns The merged normalized scope object from the [CAIP-25](https://github.com/ChainAgnostic/CAIPs/blob/main/CAIPs/caip-25.md) request.
 */
const mergeNormalizedScopes = (scopeA, scopeB) => {
    const scope = {};
    Object.entries(scopeA).forEach(([_scopeString, scopeObjectA]) => {
        // Cast needed because index type is returned as `string` by `Object.entries`
        const scopeString = _scopeString;
        const scopeObjectB = scopeB[scopeString];
        scope[scopeString] = scopeObjectB
            ? (0, exports.mergeScopeObject)(scopeObjectA, scopeObjectB)
            : scopeObjectA;
    });
    Object.entries(scopeB).forEach(([_scopeString, scopeObjectB]) => {
        // Cast needed because index type is returned as `string` by `Object.entries`
        const scopeString = _scopeString;
        const scopeObjectA = scopeA[scopeString];
        if (!scopeObjectA) {
            scope[scopeString] = scopeObjectB;
        }
    });
    return scope;
};
exports.mergeNormalizedScopes = mergeNormalizedScopes;
/**
 * Merges two InternalScopeObjects
 *
 * @param scopeA - The first internal scope object to merge.
 * @param scopeB - The second internal scope object to merge.
 * @returns The merged internal scope object from the [CAIP-25](https://github.com/ChainAgnostic/CAIPs/blob/main/CAIPs/caip-25.md) request.
 */
const mergeInternalScopes = (scopeA, scopeB) => {
    const resultScope = (0, lodash_1.cloneDeep)(scopeA);
    Object.entries(scopeB).forEach(([scopeString, rightScopeObject]) => {
        const internalScopeString = scopeString;
        const leftRequiredScopeObject = resultScope[internalScopeString];
        if (!leftRequiredScopeObject) {
            resultScope[internalScopeString] = rightScopeObject;
        }
        else {
            resultScope[internalScopeString] = {
                accounts: (0, exports.getUniqueArrayItems)([
                    ...leftRequiredScopeObject.accounts,
                    ...rightScopeObject.accounts,
                ]),
            };
        }
    });
    return resultScope;
};
exports.mergeInternalScopes = mergeInternalScopes;
/**
 * Normalizes and merges a set of ExternalScopesObjects into a NormalizedScopesObject (i.e. a set of NormalizedScopeObjects where references are flattened).
 *
 * @param scopes - The external scopes to normalize and merge.
 * @returns The normalized and merged scopes.
 */
const normalizeAndMergeScopes = (scopes) => {
    let mergedScopes = {};
    Object.keys(scopes).forEach((scopeString) => {
        const normalizedScopes = (0, exports.normalizeScope)(scopeString, scopes[scopeString]);
        mergedScopes = (0, exports.mergeNormalizedScopes)(mergedScopes, normalizedScopes);
    });
    return mergedScopes;
};
exports.normalizeAndMergeScopes = normalizeAndMergeScopes;
