// @metamask/chain-agnostic-permission/dist/scope/transform.mjs
import $lodash from "lodash";
const { cloneDeep } = $lodash;
import { parseScopeString } from "./types.mjs";
/**
 * Returns a list of unique items
 *
 * @param list - The list of items to filter
 * @returns A list of unique items
 */
export const getUniqueArrayItems = (list) => {
    return Array.from(new Set(list));
};
/**
 * Normalizes a ScopeString and ExternalScopeObject into a separate
 * InternalScopeString and NormalizedScopeObject for each reference in the `references`
 * value if defined and adds an empty `accounts` array if not defined.
 *
 * @param scopeString - The string representing the scope
 * @param externalScopeObject - The object that defines the scope
 * @returns a map of caipChainId to ScopeObjects
 */
export const normalizeScope = (scopeString, externalScopeObject) => {
    const { references, ...scopeObject } = externalScopeObject;
    const { namespace, reference } = parseScopeString(scopeString);
    const normalizedScopeObject = {
        accounts: [],
        ...scopeObject,
    };
    const shouldFlatten = namespace &&
        !reference &&
        references !== undefined &&
        references.length > 0;
    if (shouldFlatten) {
        return Object.fromEntries(references.map((ref) => [
            `${namespace}:${ref}`,
            cloneDeep(normalizedScopeObject),
        ]));
    }
    return { [scopeString]: normalizedScopeObject };
};
/**
 * Merges two NormalizedScopeObjects
 *
 * @param scopeObjectA - The first scope object to merge.
 * @param scopeObjectB - The second scope object to merge.
 * @returns The merged scope object.
 */
export const mergeScopeObject = (scopeObjectA, scopeObjectB) => {
    const mergedScopeObject = {
        methods: getUniqueArrayItems([
            ...scopeObjectA.methods,
            ...scopeObjectB.methods,
        ]),
        notifications: getUniqueArrayItems([
            ...scopeObjectA.notifications,
            ...scopeObjectB.notifications,
        ]),
        accounts: getUniqueArrayItems([
            ...scopeObjectA.accounts,
            ...scopeObjectB.accounts,
        ]),
    };
    if (scopeObjectA.rpcDocuments || scopeObjectB.rpcDocuments) {
        mergedScopeObject.rpcDocuments = getUniqueArrayItems([
            ...(scopeObjectA.rpcDocuments ?? []),
            ...(scopeObjectB.rpcDocuments ?? []),
        ]);
    }
    if (scopeObjectA.rpcEndpoints || scopeObjectB.rpcEndpoints) {
        mergedScopeObject.rpcEndpoints = getUniqueArrayItems([
            ...(scopeObjectA.rpcEndpoints ?? []),
            ...(scopeObjectB.rpcEndpoints ?? []),
        ]);
    }
    return mergedScopeObject;
};
/**
 * Merges two NormalizedScopeObjects
 *
 * @param scopeA - The first normalized scope object to merge.
 * @param scopeB - The second normalized scope object to merge.
 * @returns The merged normalized scope object from the [CAIP-25](https://github.com/ChainAgnostic/CAIPs/blob/main/CAIPs/caip-25.md) request.
 */
export const mergeNormalizedScopes = (scopeA, scopeB) => {
    const scope = {};
    Object.entries(scopeA).forEach(([_scopeString, scopeObjectA]) => {
        // Cast needed because index type is returned as `string` by `Object.entries`
        const scopeString = _scopeString;
        const scopeObjectB = scopeB[scopeString];
        scope[scopeString] = scopeObjectB
            ? mergeScopeObject(scopeObjectA, scopeObjectB)
            : scopeObjectA;
    });
    Object.entries(scopeB).forEach(([_scopeString, scopeObjectB]) => {
        // Cast needed because index type is returned as `string` by `Object.entries`
        const scopeString = _scopeString;
        const scopeObjectA = scopeA[scopeString];
        if (!scopeObjectA) {
            scope[scopeString] = scopeObjectB;
        }
    });
    return scope;
};
/**
 * Merges two InternalScopeObjects
 *
 * @param scopeA - The first internal scope object to merge.
 * @param scopeB - The second internal scope object to merge.
 * @returns The merged internal scope object from the [CAIP-25](https://github.com/ChainAgnostic/CAIPs/blob/main/CAIPs/caip-25.md) request.
 */
export const mergeInternalScopes = (scopeA, scopeB) => {
    const resultScope = cloneDeep(scopeA);
    Object.entries(scopeB).forEach(([scopeString, rightScopeObject]) => {
        const internalScopeString = scopeString;
        const leftRequiredScopeObject = resultScope[internalScopeString];
        if (!leftRequiredScopeObject) {
            resultScope[internalScopeString] = rightScopeObject;
        }
        else {
            resultScope[internalScopeString] = {
                accounts: getUniqueArrayItems([
                    ...leftRequiredScopeObject.accounts,
                    ...rightScopeObject.accounts,
                ]),
            };
        }
    });
    return resultScope;
};
/**
 * Normalizes and merges a set of ExternalScopesObjects into a NormalizedScopesObject (i.e. a set of NormalizedScopeObjects where references are flattened).
 *
 * @param scopes - The external scopes to normalize and merge.
 * @returns The normalized and merged scopes.
 */
export const normalizeAndMergeScopes = (scopes) => {
    let mergedScopes = {};
    Object.keys(scopes).forEach((scopeString) => {
        const normalizedScopes = normalizeScope(scopeString, scopes[scopeString]);
        mergedScopes = mergeNormalizedScopes(mergedScopes, normalizedScopes);
    });
    return mergedScopes;
};
