// @metamask/connect-multichain/dist/src/multichain/index.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var _MetaMaskConnectMultichain_instances, _MetaMaskConnectMultichain_provider, _MetaMaskConnectMultichain_providerTransportWrapper, _MetaMaskConnectMultichain_transport, _MetaMaskConnectMultichain_dappClient, _MetaMaskConnectMultichain_beforeUnloadListener, _MetaMaskConnectMultichain_listener, _MetaMaskConnectMultichain_anonId, _MetaMaskConnectMultichain_sdkInfo, _MetaMaskConnectMultichain_setupAnalytics, _MetaMaskConnectMultichain_onTransportNotification, _MetaMaskConnectMultichain_getStoredTransport, _MetaMaskConnectMultichain_setupTransport, _MetaMaskConnectMultichain_buildConnectionMetadata, _MetaMaskConnectMultichain_init, _MetaMaskConnectMultichain_createDappClient, _MetaMaskConnectMultichain_setupMWP, _MetaMaskConnectMultichain_onBeforeUnload, _MetaMaskConnectMultichain_createBeforeUnloadListener, _MetaMaskConnectMultichain_renderInstallModalAsync, _MetaMaskConnectMultichain_showInstallModal, _MetaMaskConnectMultichain_headlessConnect, _MetaMaskConnectMultichain_setupDefaultTransport, _MetaMaskConnectMultichain_deeplinkConnect, _MetaMaskConnectMultichain_handleConnection, _MetaMaskConnectMultichain_getCaipSession, _MetaMaskConnectMultichain_openConnectDeeplinkIfNeeded;
/* eslint-disable @typescript-eslint/no-misused-promises */
/* eslint-disable @typescript-eslint/naming-convention */
/* eslint-disable no-restricted-globals */
/* eslint-disable promise/always-return -- Event handlers */
/* eslint-disable no-async-promise-executor -- Async promise executor needed for complex flow */
import { analytics } from '@metamask/analytics';
import { ErrorCode, ProtocolError, SessionStore, WebSocketTransport, } from '@metamask/mobile-wallet-protocol-core';
import { DappClient } from '@metamask/mobile-wallet-protocol-dapp-client';
import { getMultichainClient, } from '@metamask/multichain-api-client';
import { METAMASK_CONNECT_BASE_URL, METAMASK_DEEPLINK_BASE, MWP_RELAY_URL, } from '../config';
import { getVersion, TransportType, } from '../domain';
import { getBaseAnalyticsProperties, isRejectionError, } from './utils/analytics';
import { createLogger, enableDebug, isEnabled as isLoggerEnabled, } from '../domain/logger';
import { MultichainCore, } from '../domain/multichain';
import { getPlatformType, hasExtension, isSecure, PlatformType, } from '../domain/platform';
import { RpcClient } from './rpc/handlers/rpcClient';
import { RequestRouter } from './rpc/requestRouter';
import { DefaultTransport } from './transports/default';
import { MultichainApiClientWrapperTransport } from './transports/multichainApiClientWrapper';
import { MWPTransport } from './transports/mwp';
import { keymanager } from './transports/mwp/KeyManager';
import { getDappId, getGlobalObject, mergeRequestedSessionWithExisting, openDeeplink, setupDappMetadata, } from './utils';
export { getInfuraRpcUrls } from '../domain/multichain/api/infura';
// ENFORCE NAMESPACE THAT CAN BE DISABLED
const logger = createLogger('metamask-sdk:core');
const SINGLETON_KEY = '__METAMASK_CONNECT_MULTICHAIN_SINGLETON__';
export class MetaMaskConnectMultichain extends MultichainCore {
    get status() {
        return this._status;
    }
    set status(value) {
        if (this._status === value) {
            return;
        }
        this._status = value;
        this.emit('stateChanged', value);
    }
    get provider() {
        return __classPrivateFieldGet(this, _MetaMaskConnectMultichain_provider, "f");
    }
    get transport() {
        if (!__classPrivateFieldGet(this, _MetaMaskConnectMultichain_transport, "f")) {
            throw new Error('Transport not initialized, establish connection first');
        }
        return __classPrivateFieldGet(this, _MetaMaskConnectMultichain_transport, "f");
    }
    get dappClient() {
        if (!__classPrivateFieldGet(this, _MetaMaskConnectMultichain_dappClient, "f")) {
            throw new Error('DappClient not initialized, establish connection first');
        }
        return __classPrivateFieldGet(this, _MetaMaskConnectMultichain_dappClient, "f");
    }
    get storage() {
        return this.options.storage;
    }
    get transportType() {
        return __classPrivateFieldGet(this, _MetaMaskConnectMultichain_transport, "f") instanceof MWPTransport
            ? TransportType.MWP
            : TransportType.Browser;
    }
    constructor(options) {
        var _a, _b, _c, _d, _e, _f, _g;
        const withDappMetadata = setupDappMetadata(options);
        // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
        const integrationType = ((_a = options.analytics) === null || _a === void 0 ? void 0 : _a.integrationType) || 'direct';
        const allOptions = Object.assign(Object.assign({}, withDappMetadata), { ui: Object.assign(Object.assign({}, withDappMetadata.ui), { preferExtension: (_b = withDappMetadata.ui.preferExtension) !== null && _b !== void 0 ? _b : true, showInstallModal: (_c = withDappMetadata.ui.showInstallModal) !== null && _c !== void 0 ? _c : false, headless: (_d = withDappMetadata.ui.headless) !== null && _d !== void 0 ? _d : false }), analytics: Object.assign(Object.assign({}, ((_e = options.analytics) !== null && _e !== void 0 ? _e : {})), { integrationType }), versions: Object.assign({ 
                // typeof guard needed: Metro (React Native) bundles TS source directly,
                // bypassing the tsup build that substitutes __PACKAGE_VERSION__.
                'connect-multichain': typeof __PACKAGE_VERSION__ === 'undefined'
                    ? 'unknown'
                    : __PACKAGE_VERSION__ }, ((_f = options.versions) !== null && _f !== void 0 ? _f : {})) });
        super(allOptions);
        _MetaMaskConnectMultichain_instances.add(this);
        _MetaMaskConnectMultichain_provider.set(this, void 0);
        _MetaMaskConnectMultichain_providerTransportWrapper.set(this, void 0);
        _MetaMaskConnectMultichain_transport.set(this, undefined);
        _MetaMaskConnectMultichain_dappClient.set(this, undefined);
        _MetaMaskConnectMultichain_beforeUnloadListener.set(this, void 0);
        this._status = 'pending';
        _MetaMaskConnectMultichain_listener.set(this, void 0);
        _MetaMaskConnectMultichain_anonId.set(this, void 0);
        _MetaMaskConnectMultichain_sdkInfo.set(this, `Sdk/Javascript SdkVersion/${getVersion()} Platform/${getPlatformType()} dApp/${(_g = this.options.dapp.url) !== null && _g !== void 0 ? _g : this.options.dapp.name} dAppTitle/${this.options.dapp.name}`);
        __classPrivateFieldSet(this, _MetaMaskConnectMultichain_providerTransportWrapper, new MultichainApiClientWrapperTransport(this), "f");
        __classPrivateFieldSet(this, _MetaMaskConnectMultichain_provider, getMultichainClient({
            transport: __classPrivateFieldGet(this, _MetaMaskConnectMultichain_providerTransportWrapper, "f"),
        }), "f");
    }
    // Creates a singleton instance of MetaMaskConnectMultichain.
    // If the singleton already exists, it merges the incoming options with the
    // existing singleton options for the following keys: `api.supportedNetworks`,
    // `versions`, `ui.*`, `mobile.*`, `transport.extensionId`, `debug`. Take note
    // that the value for `dapp` is not merged as it does not make sense for
    // subsequent calls to `createMultichainClient` to have a different `dapp` value.
    static create(options) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b;
            const globalObject = getGlobalObject();
            const existing = globalObject[SINGLETON_KEY];
            if (existing) {
                const instance = yield existing;
                instance.mergeOptions(options);
                analytics.setGlobalProperty('mmconnect_versions', (_a = instance.options.versions) !== null && _a !== void 0 ? _a : {});
                if ((_b = options.analytics) === null || _b === void 0 ? void 0 : _b.integrationType) {
                    analytics.setGlobalProperty('integration_types', [
                        options.analytics.integrationType,
                    ]);
                }
                if (options.debug) {
                    enableDebug('metamask-sdk:*');
                }
                return instance;
            }
            const instancePromise = (() => __awaiter(this, void 0, void 0, function* () {
                const instance = new MetaMaskConnectMultichain(options);
                const isEnabled = yield isLoggerEnabled('metamask-sdk:core', instance.options.storage);
                if (isEnabled) {
                    enableDebug('metamask-sdk:core');
                }
                yield __classPrivateFieldGet(instance, _MetaMaskConnectMultichain_instances, "m", _MetaMaskConnectMultichain_init).call(instance);
                return instance;
            }))();
            globalObject[SINGLETON_KEY] = instancePromise;
            instancePromise.catch((error) => {
                globalObject[SINGLETON_KEY] = undefined;
                console.error('Error initializing MetaMaskConnectMultichain', error);
            });
            return instancePromise;
        });
    }
    // TODO: make this into param object
    connect(scopes, caipAccountIds, sessionProperties, forceRequest) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a;
            if (this.status === 'connecting' &&
                this.transportType === TransportType.MWP) {
                yield __classPrivateFieldGet(this, _MetaMaskConnectMultichain_instances, "m", _MetaMaskConnectMultichain_openConnectDeeplinkIfNeeded).call(this);
                throw new Error('Existing connection is pending. Please check your MetaMask Mobile app to continue.');
            }
            const { ui } = this.options;
            const platformType = getPlatformType();
            const isWeb = platformType === PlatformType.MetaMaskMobileWebview ||
                platformType === PlatformType.DesktopWeb;
            const { preferExtension = true, showInstallModal = false } = ui;
            const secure = isSecure();
            const hasExtensionInstalled = yield hasExtension();
            let transportType;
            if (platformType === PlatformType.MetaMaskMobileWebview ||
                (isWeb && hasExtensionInstalled && preferExtension)) {
                transportType = TransportType.Browser;
            }
            else {
                transportType = TransportType.MWP;
            }
            try {
                const baseProps = yield getBaseAnalyticsProperties(this.options, this.storage);
                const dappConfiguredChains = Object.keys(this.options.api.supportedNetworks);
                analytics.track('mmconnect_connection_initiated', Object.assign(Object.assign({}, baseProps), { transport_type: transportType, dapp_configured_chains: dappConfiguredChains, dapp_requested_chains: scopes }));
            }
            catch (error) {
                logger('Error tracking connection_initiated event', error);
            }
            const sessionData = yield __classPrivateFieldGet(this, _MetaMaskConnectMultichain_instances, "m", _MetaMaskConnectMultichain_getCaipSession).call(this);
            const { mergedScopes, mergedCaipAccountIds, mergedSessionProperties } = mergeRequestedSessionWithExisting(sessionData, scopes, caipAccountIds, sessionProperties);
            // Needed because empty object will cause wallet_createSession to return an error
            const nonEmptySessionProperties = Object.keys(mergedSessionProperties !== null && mergedSessionProperties !== void 0 ? mergedSessionProperties : {}).length > 0
                ? mergedSessionProperties
                : undefined;
            if (((_a = __classPrivateFieldGet(this, _MetaMaskConnectMultichain_transport, "f")) === null || _a === void 0 ? void 0 : _a.isConnected()) && !secure) {
                return __classPrivateFieldGet(this, _MetaMaskConnectMultichain_instances, "m", _MetaMaskConnectMultichain_handleConnection).call(this, __classPrivateFieldGet(this, _MetaMaskConnectMultichain_transport, "f")
                    .connect({
                    scopes: mergedScopes,
                    caipAccountIds: mergedCaipAccountIds,
                    sessionProperties: nonEmptySessionProperties,
                    forceRequest,
                })
                    .then(() => __awaiter(this, void 0, void 0, function* () {
                    if (__classPrivateFieldGet(this, _MetaMaskConnectMultichain_transport, "f") instanceof MWPTransport) {
                        return this.storage.setTransport(TransportType.MWP);
                    }
                    return this.storage.setTransport(TransportType.Browser);
                })), scopes, transportType);
            }
            // In MetaMask Mobile In App Browser, window.ethereum is available directly
            if (platformType === PlatformType.MetaMaskMobileWebview) {
                const defaultTransport = yield __classPrivateFieldGet(this, _MetaMaskConnectMultichain_instances, "m", _MetaMaskConnectMultichain_setupDefaultTransport).call(this);
                return __classPrivateFieldGet(this, _MetaMaskConnectMultichain_instances, "m", _MetaMaskConnectMultichain_handleConnection).call(this, defaultTransport.connect({
                    scopes: mergedScopes,
                    caipAccountIds: mergedCaipAccountIds,
                    sessionProperties: nonEmptySessionProperties,
                    forceRequest,
                }), scopes, transportType);
            }
            if (isWeb && hasExtensionInstalled && preferExtension) {
                // If metamask extension is available, connect to it
                const defaultTransport = yield __classPrivateFieldGet(this, _MetaMaskConnectMultichain_instances, "m", _MetaMaskConnectMultichain_setupDefaultTransport).call(this);
                // Web transport has no initial payload
                return __classPrivateFieldGet(this, _MetaMaskConnectMultichain_instances, "m", _MetaMaskConnectMultichain_handleConnection).call(this, defaultTransport.connect({
                    scopes: mergedScopes,
                    caipAccountIds: mergedCaipAccountIds,
                    sessionProperties: nonEmptySessionProperties,
                    forceRequest,
                }), scopes, transportType);
            }
            // Connection will now be InstallModal + QRCodes or Deeplinks, both require mwp
            yield __classPrivateFieldGet(this, _MetaMaskConnectMultichain_instances, "m", _MetaMaskConnectMultichain_setupMWP).call(this);
            // Determine preferred option for install modal
            const shouldShowInstallModal = hasExtensionInstalled
                ? showInstallModal
                : !preferExtension || showInstallModal;
            if (secure && !shouldShowInstallModal) {
                // Desktop is not preferred option, so we use deeplinks (mobile web)
                return __classPrivateFieldGet(this, _MetaMaskConnectMultichain_instances, "m", _MetaMaskConnectMultichain_handleConnection).call(this, __classPrivateFieldGet(this, _MetaMaskConnectMultichain_instances, "m", _MetaMaskConnectMultichain_deeplinkConnect).call(this, mergedScopes, mergedCaipAccountIds, nonEmptySessionProperties), scopes, transportType);
            }
            // Show install modal for RN, Web + Node
            return __classPrivateFieldGet(this, _MetaMaskConnectMultichain_instances, "m", _MetaMaskConnectMultichain_handleConnection).call(this, __classPrivateFieldGet(this, _MetaMaskConnectMultichain_instances, "m", _MetaMaskConnectMultichain_showInstallModal).call(this, shouldShowInstallModal, mergedScopes, mergedCaipAccountIds, nonEmptySessionProperties), scopes, transportType);
        });
    }
    emit(event, args) {
        var _a, _b;
        (_b = (_a = this.options.transport) === null || _a === void 0 ? void 0 : _a.onNotification) === null || _b === void 0 ? void 0 : _b.call(_a, { method: event, params: args });
        super.emit(event, args);
    }
    disconnect() {
        return __awaiter(this, arguments, void 0, function* (scopes = []) {
            var _a, _b, _c;
            const sessionData = yield __classPrivateFieldGet(this, _MetaMaskConnectMultichain_instances, "m", _MetaMaskConnectMultichain_getCaipSession).call(this);
            const remainingScopes = scopes.length === 0
                ? []
                : Object.keys(sessionData.sessionScopes).filter((scope) => !scopes.includes(scope));
            yield ((_a = __classPrivateFieldGet(this, _MetaMaskConnectMultichain_transport, "f")) === null || _a === void 0 ? void 0 : _a.disconnect(scopes));
            if (remainingScopes.length === 0) {
                yield this.storage.removeTransport();
                // We want to leave the DefaultTransport instance connected so that we can
                // still listen for wallet_sessionChanged events.
                if (this.transportType !== TransportType.Browser) {
                    yield ((_b = __classPrivateFieldGet(this, _MetaMaskConnectMultichain_listener, "f")) === null || _b === void 0 ? void 0 : _b.call(this));
                    (_c = __classPrivateFieldGet(this, _MetaMaskConnectMultichain_beforeUnloadListener, "f")) === null || _c === void 0 ? void 0 : _c.call(this);
                    __classPrivateFieldSet(this, _MetaMaskConnectMultichain_listener, undefined, "f");
                    __classPrivateFieldSet(this, _MetaMaskConnectMultichain_beforeUnloadListener, undefined, "f");
                    __classPrivateFieldSet(this, _MetaMaskConnectMultichain_transport, undefined, "f");
                    __classPrivateFieldGet(this, _MetaMaskConnectMultichain_providerTransportWrapper, "f").clearTransportNotificationListener();
                    __classPrivateFieldSet(this, _MetaMaskConnectMultichain_dappClient, undefined, "f");
                }
                this.status = 'disconnected';
            }
        });
    }
    invokeMethod(request) {
        return __awaiter(this, void 0, void 0, function* () {
            const { transport, options } = this;
            const rpcClient = new RpcClient(options, __classPrivateFieldGet(this, _MetaMaskConnectMultichain_sdkInfo, "f"));
            const requestRouter = new RequestRouter(transport, rpcClient, options, this.transportType);
            // TODO: need read only method support for solana
            return requestRouter.invokeMethod(request);
        });
    }
    // DRY THIS WITH REQUEST ROUTER
    openSimpleDeeplinkIfNeeded() {
        const { ui, mobile } = this.options;
        const { showInstallModal = false } = ui !== null && ui !== void 0 ? ui : {};
        const secure = isSecure();
        const shouldOpenDeeplink = secure && !showInstallModal;
        if (shouldOpenDeeplink) {
            setTimeout(() => __awaiter(this, void 0, void 0, function* () {
                const session = yield this.transport.getActiveSession();
                if (!session) {
                    throw new Error('No active session found');
                }
                const url = `${METAMASK_DEEPLINK_BASE}/mwp?id=${encodeURIComponent(session.id)}`;
                if (mobile === null || mobile === void 0 ? void 0 : mobile.preferredOpenLink) {
                    mobile.preferredOpenLink(url, '_self');
                }
                else {
                    openDeeplink(this.options, url, METAMASK_CONNECT_BASE_URL);
                }
            }), 10); // small delay to ensure the message encryption and dispatch completes
        }
    }
    // Provides a way for ecosystem clients (EVM, Solana, etc.) to get the current CAIP session data
    // when instantiating themselves (as they would have already missed any initial sessionChanged events emitted by ConnectMultichain)
    // without having to concern themselves with the current transport connection status.
    emitSessionChanged() {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b;
            const emptySession = { sessionScopes: {} };
            if (!((_a = __classPrivateFieldGet(this, _MetaMaskConnectMultichain_transport, "f")) === null || _a === void 0 ? void 0 : _a.isConnected())) {
                // If we aren't connected or connecting, there definitely is no active CAIP session
                // so we optimistically emit an empty session to signify that to the ecosystem client consumers (EVM, Solana, etc.)
                this.emit('wallet_sessionChanged', emptySession);
                return;
            }
            // Otherwise, we need to fetch the current CAIP session from the wallet
            const response = yield this.transport.request({
                method: 'wallet_getSession',
            });
            // And then simulate a sessionChanged event with the current CAIP session data
            this.emit('wallet_sessionChanged', (_b = response.result) !== null && _b !== void 0 ? _b : emptySession);
        });
    }
}
_MetaMaskConnectMultichain_provider = new WeakMap(), _MetaMaskConnectMultichain_providerTransportWrapper = new WeakMap(), _MetaMaskConnectMultichain_transport = new WeakMap(), _MetaMaskConnectMultichain_dappClient = new WeakMap(), _MetaMaskConnectMultichain_beforeUnloadListener = new WeakMap(), _MetaMaskConnectMultichain_listener = new WeakMap(), _MetaMaskConnectMultichain_anonId = new WeakMap(), _MetaMaskConnectMultichain_sdkInfo = new WeakMap(), _MetaMaskConnectMultichain_instances = new WeakSet(), _MetaMaskConnectMultichain_setupAnalytics = function _MetaMaskConnectMultichain_setupAnalytics() {
    return __awaiter(this, void 0, void 0, function* () {
        var _a, _b;
        const platform = getPlatformType();
        const isBrowser = platform === PlatformType.MetaMaskMobileWebview ||
            platform === PlatformType.DesktopWeb ||
            platform === PlatformType.MobileWeb;
        const isReactNative = platform === PlatformType.ReactNative;
        if (!isBrowser && !isReactNative) {
            return;
        }
        const dappId = getDappId(this.options.dapp);
        const anonId = yield this.storage.getAnonId();
        __classPrivateFieldSet(this, _MetaMaskConnectMultichain_anonId, anonId, "f");
        const { integrationType } = (_a = this.options.analytics) !== null && _a !== void 0 ? _a : {
            integrationType: '',
        };
        analytics.setGlobalProperty('mmconnect_versions', (_b = this.options.versions) !== null && _b !== void 0 ? _b : {});
        analytics.setGlobalProperty('dapp_id', dappId);
        analytics.setGlobalProperty('anon_id', anonId);
        analytics.setGlobalProperty('platform', platform);
        if (integrationType) {
            analytics.setGlobalProperty('integration_types', [integrationType]);
        }
        analytics.enable();
    });
}, _MetaMaskConnectMultichain_onTransportNotification = function _MetaMaskConnectMultichain_onTransportNotification(payload) {
    return __awaiter(this, void 0, void 0, function* () {
        var _a, _b, _c;
        if (typeof payload === 'object' &&
            payload !== null &&
            'method' in payload) {
            if (payload.method === 'wallet_sessionChanged') {
                const sessionScopes = (_b = (_a = payload.params) === null || _a === void 0 ? void 0 : _a.sessionScopes) !== null && _b !== void 0 ? _b : {};
                const hasScopes = Object.keys(sessionScopes).length > 0;
                // During passive init, status is already 'loaded' — don't downgrade to 'disconnected'
                // just because the extension has no active session yet.
                if (this.status === 'loaded' && !hasScopes) {
                    return;
                }
                this.status = hasScopes ? 'connected' : 'disconnected';
            }
            this.emit(payload.method, (_c = payload.params) !== null && _c !== void 0 ? _c : payload.result);
        }
    });
}, _MetaMaskConnectMultichain_getStoredTransport = function _MetaMaskConnectMultichain_getStoredTransport() {
    return __awaiter(this, void 0, void 0, function* () {
        const transportType = yield this.storage.getTransport();
        const hasExtensionInstalled = yield hasExtension();
        if (transportType) {
            if (transportType === TransportType.Browser) {
                if (hasExtensionInstalled) {
                    const apiTransport = new DefaultTransport();
                    __classPrivateFieldSet(this, _MetaMaskConnectMultichain_transport, apiTransport, "f");
                    __classPrivateFieldGet(this, _MetaMaskConnectMultichain_providerTransportWrapper, "f").setupTransportNotificationListener();
                    __classPrivateFieldSet(this, _MetaMaskConnectMultichain_listener, apiTransport.onNotification(__classPrivateFieldGet(this, _MetaMaskConnectMultichain_instances, "m", _MetaMaskConnectMultichain_onTransportNotification).bind(this)), "f");
                    return apiTransport;
                }
            }
            else if (transportType === TransportType.MWP) {
                const { adapter: kvstore } = this.options.storage;
                const dappClient = yield __classPrivateFieldGet(this, _MetaMaskConnectMultichain_instances, "m", _MetaMaskConnectMultichain_createDappClient).call(this);
                const apiTransport = new MWPTransport(dappClient, kvstore);
                __classPrivateFieldSet(this, _MetaMaskConnectMultichain_dappClient, dappClient, "f");
                __classPrivateFieldSet(this, _MetaMaskConnectMultichain_transport, apiTransport, "f");
                __classPrivateFieldGet(this, _MetaMaskConnectMultichain_providerTransportWrapper, "f").setupTransportNotificationListener();
                __classPrivateFieldSet(this, _MetaMaskConnectMultichain_listener, apiTransport.onNotification(__classPrivateFieldGet(this, _MetaMaskConnectMultichain_instances, "m", _MetaMaskConnectMultichain_onTransportNotification).bind(this)), "f");
                return apiTransport;
            }
            yield this.storage.removeTransport();
        }
        return undefined;
    });
}, _MetaMaskConnectMultichain_setupTransport = function _MetaMaskConnectMultichain_setupTransport() {
    return __awaiter(this, void 0, void 0, function* () {
        var _a;
        const transport = yield __classPrivateFieldGet(this, _MetaMaskConnectMultichain_instances, "m", _MetaMaskConnectMultichain_getStoredTransport).call(this);
        if (transport) {
            if (!this.transport.isConnected()) {
                this.status = 'connecting';
                yield this.transport.connect();
            }
            this.status = 'connected';
            if (this.transport instanceof MWPTransport) {
                yield this.storage.setTransport(TransportType.MWP);
            }
            else {
                yield this.storage.setTransport(TransportType.Browser);
            }
        }
        else {
            this.status = 'loaded';
            const hasExtensionInstalled = yield hasExtension();
            const preferExtension = (_a = this.options.ui.preferExtension) !== null && _a !== void 0 ? _a : true;
            // Setup passive listening for extension wallet_sessionChanged events
            if (hasExtensionInstalled && preferExtension) {
                yield __classPrivateFieldGet(this, _MetaMaskConnectMultichain_instances, "m", _MetaMaskConnectMultichain_setupDefaultTransport).call(this, { persist: false });
                // Normally calling DefaultTransport.connect() ensures that the transport is initialized
                // and that wallet_sessionChanged (faked) is emitted. But because we are not
                // calling transport.connect(), we need to initialize DefaultTransport manually.
                try {
                    yield this.transport.init();
                }
                catch (error) {
                    console.error('Passive init failed:', error);
                }
            }
        }
    });
}, _MetaMaskConnectMultichain_buildConnectionMetadata = function _MetaMaskConnectMultichain_buildConnectionMetadata() {
    const metadata = {
        dapp: this.options.dapp,
        sdk: { version: getVersion(), platform: getPlatformType() },
    };
    if (__classPrivateFieldGet(this, _MetaMaskConnectMultichain_anonId, "f")) {
        metadata.analytics = { remote_session_id: __classPrivateFieldGet(this, _MetaMaskConnectMultichain_anonId, "f") };
    }
    return metadata;
}, _MetaMaskConnectMultichain_init = function _MetaMaskConnectMultichain_init() {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            yield __classPrivateFieldGet(this, _MetaMaskConnectMultichain_instances, "m", _MetaMaskConnectMultichain_setupAnalytics).call(this);
            yield __classPrivateFieldGet(this, _MetaMaskConnectMultichain_instances, "m", _MetaMaskConnectMultichain_setupTransport).call(this);
        }
        catch (error) {
            yield this.storage.removeTransport();
            this.status = 'pending';
            logger('MetaMaskSDK error during initialization', error);
        }
    });
}, _MetaMaskConnectMultichain_createDappClient = function _MetaMaskConnectMultichain_createDappClient() {
    return __awaiter(this, void 0, void 0, function* () {
        const { adapter: kvstore } = this.options.storage;
        const sessionstore = yield SessionStore.create(kvstore);
        const websocket = 
        // eslint-disable-next-line no-negated-condition
        typeof window !== 'undefined'
            ? WebSocket
            : (yield import('ws')).WebSocket;
        const transport = yield WebSocketTransport.create({
            url: MWP_RELAY_URL,
            kvstore,
            websocket,
        });
        const dappClient = new DappClient({ transport, sessionstore, keymanager });
        return dappClient;
    });
}, _MetaMaskConnectMultichain_setupMWP = function _MetaMaskConnectMultichain_setupMWP() {
    return __awaiter(this, void 0, void 0, function* () {
        if (__classPrivateFieldGet(this, _MetaMaskConnectMultichain_transport, "f") instanceof MWPTransport) {
            return;
        }
        // Only setup MWP if it is not already mwp
        const { adapter: kvstore } = this.options.storage;
        const dappClient = yield __classPrivateFieldGet(this, _MetaMaskConnectMultichain_instances, "m", _MetaMaskConnectMultichain_createDappClient).call(this);
        __classPrivateFieldSet(this, _MetaMaskConnectMultichain_dappClient, dappClient, "f");
        const apiTransport = new MWPTransport(dappClient, kvstore);
        __classPrivateFieldSet(this, _MetaMaskConnectMultichain_transport, apiTransport, "f");
        __classPrivateFieldGet(this, _MetaMaskConnectMultichain_providerTransportWrapper, "f").setupTransportNotificationListener();
        __classPrivateFieldSet(this, _MetaMaskConnectMultichain_listener, this.transport.onNotification(__classPrivateFieldGet(this, _MetaMaskConnectMultichain_instances, "m", _MetaMaskConnectMultichain_onTransportNotification).bind(this)), "f");
        yield this.storage.setTransport(TransportType.MWP);
    });
}, _MetaMaskConnectMultichain_onBeforeUnload = function _MetaMaskConnectMultichain_onBeforeUnload() {
    return __awaiter(this, void 0, void 0, function* () {
        var _a;
        // Fixes glitch with "connecting" state when modal is still visible and we close screen or refresh
        if ((_a = this.options.ui.factory.modal) === null || _a === void 0 ? void 0 : _a.isMounted) {
            yield this.storage.removeTransport();
        }
    });
}, _MetaMaskConnectMultichain_createBeforeUnloadListener = function _MetaMaskConnectMultichain_createBeforeUnloadListener() {
    const handler = __classPrivateFieldGet(this, _MetaMaskConnectMultichain_instances, "m", _MetaMaskConnectMultichain_onBeforeUnload).bind(this);
    if (typeof window !== 'undefined' &&
        typeof window.addEventListener !== 'undefined') {
        window.addEventListener('beforeunload', handler);
    }
    return () => {
        if (typeof window !== 'undefined' &&
            typeof window.removeEventListener !== 'undefined') {
            window.removeEventListener('beforeunload', handler);
        }
    };
}, _MetaMaskConnectMultichain_renderInstallModalAsync = function _MetaMaskConnectMultichain_renderInstallModalAsync(desktopPreferred, scopes, caipAccountIds, sessionProperties) {
    return __awaiter(this, void 0, void 0, function* () {
        return new Promise((resolve, reject) => {
            // Use Connection Modal
            this.options.ui.factory
                .renderInstallModal(desktopPreferred, () => __awaiter(this, void 0, void 0, function* () {
                if (this.dappClient.state === 'CONNECTED' ||
                    this.dappClient.state === 'CONNECTING') {
                    yield this.dappClient.disconnect();
                }
                return new Promise((_resolve) => {
                    this.dappClient.on('session_request', (sessionRequest) => {
                        _resolve({
                            sessionRequest,
                            metadata: __classPrivateFieldGet(this, _MetaMaskConnectMultichain_instances, "m", _MetaMaskConnectMultichain_buildConnectionMetadata).call(this),
                        });
                    });
                    (() => __awaiter(this, void 0, void 0, function* () {
                        var _a;
                        try {
                            yield this.transport.connect({
                                scopes,
                                caipAccountIds,
                                sessionProperties,
                            });
                            yield this.options.ui.factory.unload();
                            (_a = this.options.ui.factory.modal) === null || _a === void 0 ? void 0 : _a.unmount();
                            this.status = 'connected';
                            yield this.storage.setTransport(TransportType.MWP);
                        }
                        catch (error) {
                            if (error instanceof ProtocolError) {
                                // Ignore Request expired errors to allow modal to regenerate expired qr codes
                                if (error.code !== ErrorCode.REQUEST_EXPIRED) {
                                    this.status = 'disconnected';
                                    // Close the modal on error
                                    yield this.options.ui.factory.unload(error);
                                    reject(error);
                                }
                                // If request is expires, the QRCode will automatically be regenerated we can ignore this case
                            }
                            else {
                                this.status = 'disconnected';
                                const normalizedError = error instanceof Error ? error : new Error(String(error));
                                // Close the modal on error
                                yield this.options.ui.factory.unload(normalizedError);
                                reject(normalizedError);
                            }
                        }
                    }))().catch(() => {
                        // Error already handled in the async function
                    });
                });
            }), (error) => __awaiter(this, void 0, void 0, function* () {
                if (error) {
                    yield this.storage.removeTransport();
                    reject(error);
                }
                else {
                    yield this.storage.setTransport(TransportType.MWP);
                    resolve();
                }
            }), (uri) => {
                this.emit('display_uri', uri);
            })
                .catch((error) => {
                reject(error instanceof Error ? error : new Error(String(error)));
            });
        });
    });
}, _MetaMaskConnectMultichain_showInstallModal = function _MetaMaskConnectMultichain_showInstallModal(desktopPreferred, scopes, caipAccountIds, sessionProperties) {
    return __awaiter(this, void 0, void 0, function* () {
        var _a;
        // create the listener only once to avoid memory leaks
        __classPrivateFieldSet(this, _MetaMaskConnectMultichain_beforeUnloadListener, (_a = __classPrivateFieldGet(this, _MetaMaskConnectMultichain_beforeUnloadListener, "f")) !== null && _a !== void 0 ? _a : __classPrivateFieldGet(this, _MetaMaskConnectMultichain_instances, "m", _MetaMaskConnectMultichain_createBeforeUnloadListener).call(this), "f");
        // In headless mode, don't render UI but still emit display_uri events
        if (this.options.ui.headless) {
            yield __classPrivateFieldGet(this, _MetaMaskConnectMultichain_instances, "m", _MetaMaskConnectMultichain_headlessConnect).call(this, scopes, caipAccountIds, sessionProperties);
        }
        else {
            yield __classPrivateFieldGet(this, _MetaMaskConnectMultichain_instances, "m", _MetaMaskConnectMultichain_renderInstallModalAsync).call(this, desktopPreferred, scopes, caipAccountIds, sessionProperties);
        }
    });
}, _MetaMaskConnectMultichain_headlessConnect = function _MetaMaskConnectMultichain_headlessConnect(scopes, caipAccountIds, sessionProperties) {
    return __awaiter(this, void 0, void 0, function* () {
        return new Promise((resolve, reject) => {
            if (this.dappClient.state === 'CONNECTED' ||
                this.dappClient.state === 'CONNECTING') {
                this.dappClient.disconnect().catch(() => {
                    // Ignore disconnect errors
                });
            }
            // Listen for session_request to generate and emit the QR code link
            this.dappClient.on('session_request', (sessionRequest) => {
                const connectionRequest = {
                    sessionRequest,
                    metadata: __classPrivateFieldGet(this, _MetaMaskConnectMultichain_instances, "m", _MetaMaskConnectMultichain_buildConnectionMetadata).call(this),
                };
                // Generate and emit the QR code link
                const deeplink = this.options.ui.factory.createConnectionDeeplink(connectionRequest);
                this.emit('display_uri', deeplink);
            });
            // Start the connection
            this.transport
                .connect({ scopes, caipAccountIds, sessionProperties })
                .then(() => __awaiter(this, void 0, void 0, function* () {
                this.status = 'connected';
                yield this.storage.setTransport(TransportType.MWP);
                resolve();
            }))
                .catch((error) => __awaiter(this, void 0, void 0, function* () {
                if (error instanceof ProtocolError) {
                    // In headless mode, we don't auto-regenerate QR codes
                    // since there's no modal to display them
                    this.status = 'disconnected';
                    yield this.storage.removeTransport();
                    reject(error);
                }
                else {
                    this.status = 'disconnected';
                    yield this.storage.removeTransport();
                    reject(error instanceof Error ? error : new Error(String(error)));
                }
            }));
        });
    });
}, _MetaMaskConnectMultichain_setupDefaultTransport = function _MetaMaskConnectMultichain_setupDefaultTransport() {
    return __awaiter(this, arguments, void 0, function* (options = { persist: true }) {
        if (__classPrivateFieldGet(this, _MetaMaskConnectMultichain_transport, "f") instanceof DefaultTransport) {
            return __classPrivateFieldGet(this, _MetaMaskConnectMultichain_transport, "f");
        }
        if (options === null || options === void 0 ? void 0 : options.persist) {
            yield this.storage.setTransport(TransportType.Browser);
        }
        const transport = new DefaultTransport();
        __classPrivateFieldSet(this, _MetaMaskConnectMultichain_listener, transport.onNotification(__classPrivateFieldGet(this, _MetaMaskConnectMultichain_instances, "m", _MetaMaskConnectMultichain_onTransportNotification).bind(this)), "f");
        __classPrivateFieldSet(this, _MetaMaskConnectMultichain_transport, transport, "f");
        __classPrivateFieldGet(this, _MetaMaskConnectMultichain_providerTransportWrapper, "f").setupTransportNotificationListener();
        return transport;
    });
}, _MetaMaskConnectMultichain_deeplinkConnect = function _MetaMaskConnectMultichain_deeplinkConnect(scopes, caipAccountIds, sessionProperties) {
    return __awaiter(this, void 0, void 0, function* () {
        return new Promise((resolve, reject) => __awaiter(this, void 0, void 0, function* () {
            // Handle the response to the initial wallet_createSession request
            const dappClientMessageHandler = (payload) => {
                var _a;
                if (typeof payload !== 'object' ||
                    payload === null ||
                    !('data' in payload)) {
                    return;
                }
                const data = payload.data;
                if (typeof data === 'object' && data !== null) {
                    // optimistically assume any error is due to the initial wallet_createSession request failure
                    if (data.error) {
                        this.dappClient.off('message', dappClientMessageHandler);
                        reject(data.error);
                    }
                    // if sessionScopes is set in the result, then this is a response to wallet_createSession
                    if ((_a = data === null || data === void 0 ? void 0 : data.result) === null || _a === void 0 ? void 0 : _a.sessionScopes) {
                        this.dappClient.off('message', dappClientMessageHandler);
                        // unsure if we need to call resolve here like we do above for reject()
                    }
                }
            };
            this.dappClient.on('message', dappClientMessageHandler);
            let timeout;
            if (this.transport.isConnected()) {
                timeout = setTimeout(() => {
                    this.openSimpleDeeplinkIfNeeded();
                }, 250);
            }
            else {
                this.dappClient.once('session_request', (sessionRequest) => {
                    var _a;
                    const connectionRequest = {
                        sessionRequest,
                        metadata: __classPrivateFieldGet(this, _MetaMaskConnectMultichain_instances, "m", _MetaMaskConnectMultichain_buildConnectionMetadata).call(this),
                    };
                    const deeplink = this.options.ui.factory.createConnectionDeeplink(connectionRequest);
                    const universalLink = this.options.ui.factory.createConnectionUniversalLink(connectionRequest);
                    // Emit display_uri event for deeplink connections
                    this.emit('display_uri', deeplink);
                    if ((_a = this.options.mobile) === null || _a === void 0 ? void 0 : _a.preferredOpenLink) {
                        this.options.mobile.preferredOpenLink(deeplink, '_self');
                    }
                    else {
                        openDeeplink(this.options, deeplink, universalLink);
                    }
                });
            }
            return this.transport
                .connect({ scopes, caipAccountIds, sessionProperties })
                .then(resolve)
                .catch((error) => __awaiter(this, void 0, void 0, function* () {
                yield this.storage.removeTransport();
                this.dappClient.off('message', dappClientMessageHandler);
                reject(error instanceof Error ? error : new Error(String(error)));
            }))
                .finally(() => {
                if (timeout) {
                    clearTimeout(timeout);
                }
            });
        }));
    });
}, _MetaMaskConnectMultichain_handleConnection = function _MetaMaskConnectMultichain_handleConnection(promise, scopes, transportType) {
    return __awaiter(this, void 0, void 0, function* () {
        this.status = 'connecting';
        return promise
            .then(() => __awaiter(this, void 0, void 0, function* () {
            this.status = 'connected';
            try {
                const baseProps = yield getBaseAnalyticsProperties(this.options, this.storage);
                analytics.track('mmconnect_connection_established', Object.assign(Object.assign({}, baseProps), { transport_type: transportType, user_permissioned_chains: scopes }));
            }
            catch (error) {
                logger('Error tracking connection_established event', error);
            }
            return undefined; // explicitly return `undefined` to avoid eslintpromise/always-return
        }))
            .catch((error) => __awaiter(this, void 0, void 0, function* () {
            this.status = 'disconnected';
            try {
                const baseProps = yield getBaseAnalyticsProperties(this.options, this.storage);
                const isRejection = isRejectionError(error);
                if (isRejection) {
                    analytics.track('mmconnect_connection_rejected', Object.assign(Object.assign({}, baseProps), { transport_type: transportType }));
                }
                else {
                    analytics.track('mmconnect_connection_failed', Object.assign(Object.assign({}, baseProps), { transport_type: transportType }));
                }
            }
            catch (_a) {
                logger('Error tracking connection failed/rejected event', error);
            }
            throw error;
        }));
    });
}, _MetaMaskConnectMultichain_getCaipSession = function _MetaMaskConnectMultichain_getCaipSession() {
    return __awaiter(this, void 0, void 0, function* () {
        var _a;
        let sessionData = {
            sessionScopes: {},
            sessionProperties: {},
        };
        if ((_a = __classPrivateFieldGet(this, _MetaMaskConnectMultichain_transport, "f")) === null || _a === void 0 ? void 0 : _a.isConnected()) {
            try {
                const response = yield this.transport.request({
                    method: 'wallet_getSession',
                });
                if (response.result) {
                    sessionData = response.result;
                }
            }
            catch (_b) {
                // If session retrieval fails, return empty session
            }
        }
        return sessionData;
    });
}, _MetaMaskConnectMultichain_openConnectDeeplinkIfNeeded = function _MetaMaskConnectMultichain_openConnectDeeplinkIfNeeded() {
    return __awaiter(this, void 0, void 0, function* () {
        var _a, _b;
        const { ui } = this.options;
        const { showInstallModal = false } = ui !== null && ui !== void 0 ? ui : {};
        const secure = isSecure();
        const shouldOpenDeeplink = secure && !showInstallModal;
        if (!shouldOpenDeeplink) {
            return;
        }
        const storedSessionRequest = yield ((_a = __classPrivateFieldGet(this, _MetaMaskConnectMultichain_transport, "f")) === null || _a === void 0 ? void 0 : _a.getStoredPendingSessionRequest());
        if (!storedSessionRequest) {
            return;
        }
        const connectionRequest = {
            sessionRequest: storedSessionRequest,
            metadata: __classPrivateFieldGet(this, _MetaMaskConnectMultichain_instances, "m", _MetaMaskConnectMultichain_buildConnectionMetadata).call(this),
        };
        const deeplink = this.options.ui.factory.createConnectionDeeplink(connectionRequest);
        const universalLink = this.options.ui.factory.createConnectionUniversalLink(connectionRequest);
        if ((_b = this.options.mobile) === null || _b === void 0 ? void 0 : _b.preferredOpenLink) {
            this.options.mobile.preferredOpenLink(deeplink, '_self');
        }
        else {
            openDeeplink(this.options, deeplink, universalLink);
        }
    });
};
