// @metamask/connect-evm/dist/src/connect.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _MetamaskConnectEVM_instances, _MetamaskConnectEVM_core, _MetamaskConnectEVM_provider, _MetamaskConnectEVM_sessionScopes, _MetamaskConnectEVM_eventHandlers, _MetamaskConnectEVM_sessionChangedHandler, _MetamaskConnectEVM_displayUriHandler, _MetamaskConnectEVM_removeNotificationHandler, _MetamaskConnectEVM_status, _MetamaskConnectEVM_pendingPreferredChainId, _MetamaskConnectEVM_initPromise, _MetamaskConnectEVM_getCoreOptions, _MetamaskConnectEVM_createInvokeOptions, _MetamaskConnectEVM_trackWalletActionRequested, _MetamaskConnectEVM_trackWalletActionSucceeded, _MetamaskConnectEVM_trackWalletActionFailed, _MetamaskConnectEVM_getSelectedChainId, _MetamaskConnectEVM_requestInterceptor, _MetamaskConnectEVM_clearConnectionState, _MetamaskConnectEVM_addEthereumChain, _MetamaskConnectEVM_request, _MetamaskConnectEVM_cacheChainId, _MetamaskConnectEVM_onSessionChanged, _MetamaskConnectEVM_onChainChanged, _MetamaskConnectEVM_onAccountsChanged, _MetamaskConnectEVM_onConnect, _MetamaskConnectEVM_onDisconnect, _MetamaskConnectEVM_onDisplayUri;
/* eslint-disable no-restricted-syntax -- Private class properties use established patterns */
/* eslint-disable @typescript-eslint/naming-convention -- __PACKAGE_VERSION__ is an esbuild define convention */
import { analytics } from '@metamask/analytics';
import { parseScopeString } from '@metamask/chain-agnostic-permission';
import { createMultichainClient, getWalletActionAnalyticsProperties, isRejectionError, TransportType, } from '@metamask/connect-multichain';
import { createDeferredPromise, hexToNumber } from '@metamask/utils';
import { IGNORED_METHODS } from './constants';
import { enableDebug, logger } from './logger';
import { EIP1193Provider } from './provider';
import { getEthAccounts, getPermittedEthChainIds } from './utils/caip';
import { isAccountsRequest, isAddChainRequest, isChainIdRequest, isConnectRequest, isSwitchChainRequest, validSupportedChainsUrls, } from './utils/type-guards';
const DEFAULT_CHAIN_ID = '0x1';
const CHAIN_STORE_KEY = 'cache_eth_chainId';
/**
 * The MetamaskConnectEVM class provides an EIP-1193 compatible interface for connecting
 * to MetaMask and interacting with Ethereum Virtual Machine (EVM) networks.
 *
 * This class serves as a modern replacement for MetaMask SDK V1, offering enhanced
 * functionality and cross-platform compatibility. It wraps the Multichain SDK to provide
 * a simplified, EIP-1193 compliant API for dapp developers.
 *
 * Key features:
 * - EIP-1193 provider interface for seamless integration with existing dapp code
 * - Automatic session recovery when reloading or opening in new tabs
 * - Chain switching with automatic chain addition if not configured
 * - Event-driven architecture with support for connect, disconnect, accountsChanged, and chainChanged events
 * - Cross-platform support for browser extensions and mobile applications
 * - Built-in handling of common Ethereum methods (eth_accounts, wallet_switchEthereumChain, etc.)
 *
 * @example
 * ```typescript
 * const client = await createEVMClient({
 *   dapp: { name: 'My DApp', url: 'https://mydapp.com' }
 * });
 *
 * const { accounts, chainId } = await client.connect({ chainIds: [1, 137] }); // Connect to Ethereum Mainnet, and Polygon
 *
 * const provider = client.getProvider();
 * const signedMessage = await provider.request({ method: 'personal_sign', params: ['0x0', accounts[0]] });
 * ```
 */
export class MetamaskConnectEVM {
    /**
     * Creates a new MetamaskConnectEVM instance.
     * Use the static `create()` method instead to ensure proper async initialization.
     *
     * @param options - The options for the MetamaskConnectEVM instance
     * @param options.core - The core instance of the Multichain SDK
     * @param options.eventHandlers - Optional event handlers for EIP-1193 provider events
     */
    constructor({ core, eventHandlers }) {
        _MetamaskConnectEVM_instances.add(this);
        /** The core instance of the Multichain SDK */
        _MetamaskConnectEVM_core.set(this, void 0);
        /** An instance of the EIP-1193 provider interface */
        _MetamaskConnectEVM_provider.set(this, void 0);
        /** The session scopes currently permitted */
        _MetamaskConnectEVM_sessionScopes.set(this, {});
        /** Optional event handlers for the EIP-1193 provider events. */
        _MetamaskConnectEVM_eventHandlers.set(this, void 0);
        /** The handler for the wallet_sessionChanged event */
        _MetamaskConnectEVM_sessionChangedHandler.set(this, void 0);
        /** The handler for the display_uri event */
        _MetamaskConnectEVM_displayUriHandler.set(this, void 0);
        /** The clean-up function for the notification handler */
        _MetamaskConnectEVM_removeNotificationHandler.set(this, void 0);
        /** The current connection status */
        _MetamaskConnectEVM_status.set(this, 'disconnected');
        /**
         * The preferred chain ID to use for the active connect() call.
         * Set to the first explicit chainId passed by the caller, cleared once the
         * connect event resolves. Takes priority over the cache in #getSelectedChainId
         * so that an explicit chainIds request always wins over a prior cached value.
         */
        _MetamaskConnectEVM_pendingPreferredChainId.set(this, void 0);
        /** Deferred that resolves once #onSessionChanged has completed for the first time. */
        _MetamaskConnectEVM_initPromise.set(this, void 0);
        __classPrivateFieldSet(this, _MetamaskConnectEVM_core, core, "f");
        __classPrivateFieldSet(this, _MetamaskConnectEVM_provider, new EIP1193Provider(core, __classPrivateFieldGet(this, _MetamaskConnectEVM_instances, "m", _MetamaskConnectEVM_requestInterceptor).bind(this)), "f");
        __classPrivateFieldSet(this, _MetamaskConnectEVM_eventHandlers, eventHandlers, "f");
        /**
         * Handles the wallet_sessionChanged event.
         * Updates the internal connection state with the new session data.
         *
         * @param session - The session data
         */
        // eslint-disable-next-line @typescript-eslint/no-misused-promises
        __classPrivateFieldSet(this, _MetamaskConnectEVM_sessionChangedHandler, __classPrivateFieldGet(this, _MetamaskConnectEVM_instances, "m", _MetamaskConnectEVM_onSessionChanged).bind(this), "f");
        __classPrivateFieldGet(this, _MetamaskConnectEVM_core, "f").on('wallet_sessionChanged', __classPrivateFieldGet(this, _MetamaskConnectEVM_sessionChangedHandler, "f"));
        /**
         * Handles the display_uri event.
         * Forwards the QR code URI to the provider for custom UI implementations.
         */
        __classPrivateFieldSet(this, _MetamaskConnectEVM_displayUriHandler, __classPrivateFieldGet(this, _MetamaskConnectEVM_instances, "m", _MetamaskConnectEVM_onDisplayUri).bind(this), "f");
        __classPrivateFieldGet(this, _MetamaskConnectEVM_core, "f").on('display_uri', __classPrivateFieldGet(this, _MetamaskConnectEVM_displayUriHandler, "f"));
        __classPrivateFieldSet(this, _MetamaskConnectEVM_initPromise, createDeferredPromise(), "f");
        logger('Connect/EVM constructor completed');
    }
    /**
     * Creates a fully initialized MetamaskConnectEVM instance.
     * This is the recommended way to instantiate the class, as it ensures
     * all async initialization (like session recovery) completes before
     * the instance is returned.
     *
     * @param options - The options for the MetamaskConnectEVM instance
     * @param options.core - The core instance of the Multichain SDK
     * @param options.eventHandlers - Optional event handlers for EIP-1193 provider events
     * @returns A promise that resolves with a fully initialized MetamaskConnectEVM instance
     */
    static create(options) {
        return __awaiter(this, void 0, void 0, function* () {
            const instance = new MetamaskConnectEVM(options);
            yield __classPrivateFieldGet(instance, _MetamaskConnectEVM_core, "f").emitSessionChanged();
            yield __classPrivateFieldGet(instance, _MetamaskConnectEVM_initPromise, "f").promise;
            return instance;
        });
    }
    /**
     * Connects to the wallet with the specified chain ID and optional account.
     *
     * @param options - The connection options
     * @param [options.account] - Optional param to specify an account to connect to
     * @param [options.forceRequest] - Optional param to force a connection request regardless of whether there is a pre-existing session
     * @param [options.chainIds] - Array of chain IDs to request permission for (defaults to ethereum mainnet). The first entry is used as the active chain returned by the call. Ethereum mainnet is always included in the permission request as a bootstrap fallback.
     * @returns A promise that resolves with the connected accounts and chain ID
     */
    connect() {
        return __awaiter(this, arguments, void 0, function* ({ account, forceRequest, chainIds = [DEFAULT_CHAIN_ID], } = {}) {
            var _a;
            logger('request: connect', { account });
            if (!chainIds || chainIds.length === 0) {
                throw new Error('chainIds must be an array of at least one chain ID');
            }
            // The first explicitly-requested chain is preferred for chain selection.
            // DEFAULT_CHAIN_ID is still included in the permission request as a bootstrap
            // fallback, but it must not win over an explicit caller request.
            __classPrivateFieldSet(this, _MetamaskConnectEVM_pendingPreferredChainId, chainIds[0], "f");
            const caipChainIds = Array.from(new Set((_a = chainIds.concat(DEFAULT_CHAIN_ID)) !== null && _a !== void 0 ? _a : [DEFAULT_CHAIN_ID])).map((id) => `eip155:${hexToNumber(id)}`);
            const caipAccountIds = account
                ? caipChainIds.map((caipChainId) => `${caipChainId}:${account}`)
                : [];
            __classPrivateFieldSet(this, _MetamaskConnectEVM_status, 'connecting', "f");
            try {
                // Wait for the wallet_sessionChanged event to fire and set the provider properties
                const result = new Promise((resolve) => {
                    __classPrivateFieldGet(this, _MetamaskConnectEVM_provider, "f").once('connect', ({ chainId, accounts }) => {
                        logger('fulfilled-request: connect', {
                            chainId,
                            accounts,
                        });
                        resolve({
                            accounts,
                            chainId,
                        });
                    });
                });
                yield __classPrivateFieldGet(this, _MetamaskConnectEVM_core, "f").connect(caipChainIds, caipAccountIds, undefined, forceRequest);
                // Await here so the finally block always runs after #onSessionChanged
                // has consumed #pendingPreferredChainId, not before.
                return yield result;
            }
            catch (error) {
                __classPrivateFieldSet(this, _MetamaskConnectEVM_status, 'disconnected', "f");
                logger('Error connecting to wallet', error);
                throw error;
            }
            finally {
                __classPrivateFieldSet(this, _MetamaskConnectEVM_pendingPreferredChainId, undefined, "f");
            }
        });
    }
    /**
     * Connects to the wallet and signs a message using personal_sign.
     *
     * @param options - The connection options
     * @param options.message - The message to sign after connecting
     * @param [options.chainIds] - Optional hex chain IDs to connect to (defaults to ethereum mainnet if not provided)
     * @returns A promise that resolves with the connected accounts, chainId, and signature
     * @throws Error if the selected account is not available after timeout
     */
    connectAndSign(_a) {
        return __awaiter(this, arguments, void 0, function* ({ message, chainIds, }) {
            var _b, _c;
            const { accounts, chainId } = yield this.connect({
                chainIds: chainIds !== null && chainIds !== void 0 ? chainIds : [DEFAULT_CHAIN_ID],
            });
            const signature = (yield __classPrivateFieldGet(this, _MetamaskConnectEVM_provider, "f").request({
                method: 'personal_sign',
                params: [accounts[0], message],
            }));
            (_c = (_b = __classPrivateFieldGet(this, _MetamaskConnectEVM_eventHandlers, "f")) === null || _b === void 0 ? void 0 : _b.connectAndSign) === null || _c === void 0 ? void 0 : _c.call(_b, {
                accounts,
                chainId,
                signature,
            });
            return { accounts, chainId, signature };
        });
    }
    /**
     * Connects to the wallet and invokes a method with specified parameters.
     *
     * @param options - The options for connecting and invoking the method
     * @param options.method - The method name to invoke
     * @param options.params - The parameters to pass to the method, or a function that receives the account and returns params
     * @param [options.chainIds] - Optional hex chain IDs to connect to (defaults to ethereum mainnet if not provided)
     * @param [options.account] - Optional specific account to connect to
     * @param [options.forceRequest] - Whether to force a request regardless of an existing session
     * @returns A promise that resolves with the connected accounts, chainId, and the result of the method invocation
     * @throws Error if the selected account is not available after timeout (for methods that require an account)
     */
    connectWith(_a) {
        return __awaiter(this, arguments, void 0, function* ({ method, params, chainIds, account, forceRequest, }) {
            var _b, _c;
            const { accounts: connectedAccounts, chainId: connectedChainId } = yield this.connect({
                chainIds: chainIds !== null && chainIds !== void 0 ? chainIds : [DEFAULT_CHAIN_ID],
                account,
                forceRequest,
            });
            const resolvedParams = typeof params === 'function' ? params(connectedAccounts[0]) : params;
            const result = yield __classPrivateFieldGet(this, _MetamaskConnectEVM_provider, "f").request({
                method,
                params: resolvedParams,
            });
            (_c = (_b = __classPrivateFieldGet(this, _MetamaskConnectEVM_eventHandlers, "f")) === null || _b === void 0 ? void 0 : _b.connectWith) === null || _c === void 0 ? void 0 : _c.call(_b, {
                accounts: connectedAccounts,
                chainId: connectedChainId,
                result,
            });
            return { accounts: connectedAccounts, chainId: connectedChainId, result };
        });
    }
    /**
     * Disconnects from the wallet by revoking the session and cleaning up event listeners.
     *
     * @returns A promise that resolves when disconnection is complete
     */
    disconnect() {
        return __awaiter(this, void 0, void 0, function* () {
            var _a;
            logger('request: disconnect');
            const sessionScopes = __classPrivateFieldGet(this, _MetamaskConnectEVM_sessionScopes, "f");
            const eip155Scopes = Object.keys(sessionScopes).filter((scope) => {
                const { namespace } = parseScopeString(scope);
                return namespace === 'eip155';
            });
            yield __classPrivateFieldGet(this, _MetamaskConnectEVM_core, "f").disconnect(eip155Scopes);
            __classPrivateFieldGet(this, _MetamaskConnectEVM_instances, "m", _MetamaskConnectEVM_onDisconnect).call(this);
            __classPrivateFieldGet(this, _MetamaskConnectEVM_instances, "m", _MetamaskConnectEVM_clearConnectionState).call(this);
            // Note: We intentionally do NOT remove the display_uri and wallet_sessionChanged
            // listeners here. These are instance-scoped listeners that should remain active
            // for the lifetime of the SDK instance, allowing reconnection to work properly.
            // Session-scoped listeners (like the notification handler below) are removed.
            (_a = __classPrivateFieldGet(this, _MetamaskConnectEVM_removeNotificationHandler, "f")) === null || _a === void 0 ? void 0 : _a.call(this);
            __classPrivateFieldSet(this, _MetamaskConnectEVM_removeNotificationHandler, undefined, "f");
            logger('fulfilled-request: disconnect');
        });
    }
    /**
     * Switches the Ethereum chain. Will track state internally whenever possible.
     *
     * @param options - The options for the switch chain request
     * @param options.chainId - The chain ID to switch to
     * @param [options.chainConfiguration] - The chain configuration to use in case the chain is not present by the wallet
     * @returns A promise that resolves when the chain has been switched
     */
    switchChain(_a) {
        return __awaiter(this, arguments, void 0, function* ({ chainId, chainConfiguration, }) {
            const method = 'wallet_switchEthereumChain';
            const scope = `eip155:${hexToNumber(chainId)}`;
            const params = [{ chainId }];
            // TODO (wenfix): better way to return here other than resolving.
            if (this.selectedChainId === chainId) {
                return Promise.resolve();
            }
            const permittedChainIds = getPermittedEthChainIds(__classPrivateFieldGet(this, _MetamaskConnectEVM_sessionScopes, "f"));
            if (permittedChainIds.includes(chainId) &&
                __classPrivateFieldGet(this, _MetamaskConnectEVM_core, "f").transportType === TransportType.MWP) {
                yield __classPrivateFieldGet(this, _MetamaskConnectEVM_instances, "m", _MetamaskConnectEVM_cacheChainId).call(this, chainId);
                __classPrivateFieldGet(this, _MetamaskConnectEVM_instances, "m", _MetamaskConnectEVM_onChainChanged).call(this, chainId);
                return Promise.resolve();
            }
            yield __classPrivateFieldGet(this, _MetamaskConnectEVM_instances, "m", _MetamaskConnectEVM_trackWalletActionRequested).call(this, method, scope, params);
            try {
                const result = yield __classPrivateFieldGet(this, _MetamaskConnectEVM_instances, "m", _MetamaskConnectEVM_request).call(this, {
                    method: 'wallet_switchEthereumChain',
                    params,
                });
                // When using the MWP transport, the error is returned instead of thrown,
                // so we force it into the catch block here.
                const resultWithError = result;
                if (resultWithError === null || resultWithError === void 0 ? void 0 : resultWithError.error) {
                    throw new Error(resultWithError.error.message);
                }
                yield __classPrivateFieldGet(this, _MetamaskConnectEVM_instances, "m", _MetamaskConnectEVM_trackWalletActionSucceeded).call(this, method, scope, params);
                if (result.result === null) {
                    // result is successful we eagerly call onChainChanged to update the provider's selected chain ID.
                    yield __classPrivateFieldGet(this, _MetamaskConnectEVM_instances, "m", _MetamaskConnectEVM_cacheChainId).call(this, chainId);
                    __classPrivateFieldGet(this, _MetamaskConnectEVM_instances, "m", _MetamaskConnectEVM_onChainChanged).call(this, chainId);
                }
                return Promise.resolve();
            }
            catch (error) {
                yield __classPrivateFieldGet(this, _MetamaskConnectEVM_instances, "m", _MetamaskConnectEVM_trackWalletActionFailed).call(this, method, scope, params, error);
                // Fallback to add the chain if its not configured in the wallet.
                if (error.message.includes('Unrecognized chain ID')) {
                    return __classPrivateFieldGet(this, _MetamaskConnectEVM_instances, "m", _MetamaskConnectEVM_addEthereumChain).call(this, chainConfiguration);
                }
                throw error;
            }
        });
    }
    /**
     * Gets the EIP-1193 provider instance
     *
     * @returns The EIP-1193 provider instance
     */
    getProvider() {
        return __classPrivateFieldGet(this, _MetamaskConnectEVM_provider, "f");
    }
    /**
     * Gets the currently selected chain ID on the wallet
     *
     * @returns The currently selected chain ID or undefined if no chain is selected
     */
    getChainId() {
        return this.selectedChainId;
    }
    /**
     * Gets the currently selected account on the wallet
     *
     * @returns The currently selected account or undefined if no account is selected
     */
    getAccount() {
        return __classPrivateFieldGet(this, _MetamaskConnectEVM_provider, "f").selectedAccount;
    }
    // Convenience getters for the EIP-1193 provider
    /**
     * Gets the currently permitted accounts
     *
     * @returns The currently permitted accounts
     */
    get accounts() {
        return __classPrivateFieldGet(this, _MetamaskConnectEVM_provider, "f").accounts;
    }
    /**
     * Gets the currently selected account on the wallet
     *
     * @returns The currently selected account or undefined if no account is selected
     */
    get selectedAccount() {
        return __classPrivateFieldGet(this, _MetamaskConnectEVM_provider, "f").selectedAccount;
    }
    /**
     * Gets the currently selected chain ID on the wallet
     *
     * @returns The currently selected chain ID or undefined if no chain is selected
     */
    get selectedChainId() {
        return __classPrivateFieldGet(this, _MetamaskConnectEVM_provider, "f").selectedChainId;
    }
    /**
     * Gets the current connection status
     *
     * @returns The current connection status
     */
    get status() {
        return __classPrivateFieldGet(this, _MetamaskConnectEVM_status, "f");
    }
}
_MetamaskConnectEVM_core = new WeakMap(), _MetamaskConnectEVM_provider = new WeakMap(), _MetamaskConnectEVM_sessionScopes = new WeakMap(), _MetamaskConnectEVM_eventHandlers = new WeakMap(), _MetamaskConnectEVM_sessionChangedHandler = new WeakMap(), _MetamaskConnectEVM_displayUriHandler = new WeakMap(), _MetamaskConnectEVM_removeNotificationHandler = new WeakMap(), _MetamaskConnectEVM_status = new WeakMap(), _MetamaskConnectEVM_pendingPreferredChainId = new WeakMap(), _MetamaskConnectEVM_initPromise = new WeakMap(), _MetamaskConnectEVM_instances = new WeakSet(), _MetamaskConnectEVM_getCoreOptions = function _MetamaskConnectEVM_getCoreOptions() {
    return __classPrivateFieldGet(this, _MetamaskConnectEVM_core, "f").options;
}, _MetamaskConnectEVM_createInvokeOptions = function _MetamaskConnectEVM_createInvokeOptions(method, scope, params) {
    return {
        scope,
        request: { method, params },
    };
}, _MetamaskConnectEVM_trackWalletActionRequested = function _MetamaskConnectEVM_trackWalletActionRequested(method, scope, params) {
    return __awaiter(this, void 0, void 0, function* () {
        const coreOptions = __classPrivateFieldGet(this, _MetamaskConnectEVM_instances, "m", _MetamaskConnectEVM_getCoreOptions).call(this);
        try {
            const invokeOptions = __classPrivateFieldGet(this, _MetamaskConnectEVM_instances, "m", _MetamaskConnectEVM_createInvokeOptions).call(this, method, scope, params);
            const props = yield getWalletActionAnalyticsProperties(coreOptions, __classPrivateFieldGet(this, _MetamaskConnectEVM_core, "f").storage, invokeOptions, __classPrivateFieldGet(this, _MetamaskConnectEVM_core, "f").transportType);
            analytics.track('mmconnect_wallet_action_requested', props);
        }
        catch (error) {
            logger('Error tracking mmconnect_wallet_action_requested event', error);
        }
    });
}, _MetamaskConnectEVM_trackWalletActionSucceeded = function _MetamaskConnectEVM_trackWalletActionSucceeded(method, scope, params) {
    return __awaiter(this, void 0, void 0, function* () {
        const coreOptions = __classPrivateFieldGet(this, _MetamaskConnectEVM_instances, "m", _MetamaskConnectEVM_getCoreOptions).call(this);
        try {
            const invokeOptions = __classPrivateFieldGet(this, _MetamaskConnectEVM_instances, "m", _MetamaskConnectEVM_createInvokeOptions).call(this, method, scope, params);
            const props = yield getWalletActionAnalyticsProperties(coreOptions, __classPrivateFieldGet(this, _MetamaskConnectEVM_core, "f").storage, invokeOptions, __classPrivateFieldGet(this, _MetamaskConnectEVM_core, "f").transportType);
            analytics.track('mmconnect_wallet_action_succeeded', props);
        }
        catch (error) {
            logger('Error tracking mmconnect_wallet_action_succeeded event', error);
        }
    });
}, _MetamaskConnectEVM_trackWalletActionFailed = function _MetamaskConnectEVM_trackWalletActionFailed(method, scope, params, error) {
    return __awaiter(this, void 0, void 0, function* () {
        const coreOptions = __classPrivateFieldGet(this, _MetamaskConnectEVM_instances, "m", _MetamaskConnectEVM_getCoreOptions).call(this);
        try {
            const invokeOptions = __classPrivateFieldGet(this, _MetamaskConnectEVM_instances, "m", _MetamaskConnectEVM_createInvokeOptions).call(this, method, scope, params);
            const props = yield getWalletActionAnalyticsProperties(coreOptions, __classPrivateFieldGet(this, _MetamaskConnectEVM_core, "f").storage, invokeOptions, __classPrivateFieldGet(this, _MetamaskConnectEVM_core, "f").transportType);
            const isRejection = isRejectionError(error);
            if (isRejection) {
                analytics.track('mmconnect_wallet_action_rejected', props);
            }
            else {
                analytics.track('mmconnect_wallet_action_failed', props);
            }
        }
        catch (_a) {
            logger('Error tracking wallet action rejected or failed event', error);
        }
    });
}, _MetamaskConnectEVM_getSelectedChainId = function _MetamaskConnectEVM_getSelectedChainId(permittedChainIds) {
    return __awaiter(this, void 0, void 0, function* () {
        // Honour an explicit caller preference set during an active connect() call
        if (__classPrivateFieldGet(this, _MetamaskConnectEVM_pendingPreferredChainId, "f") &&
            permittedChainIds.includes(__classPrivateFieldGet(this, _MetamaskConnectEVM_pendingPreferredChainId, "f"))) {
            return __classPrivateFieldGet(this, _MetamaskConnectEVM_pendingPreferredChainId, "f");
        }
        try {
            const cachedChainId = yield __classPrivateFieldGet(this, _MetamaskConnectEVM_core, "f").storage.adapter.get(CHAIN_STORE_KEY);
            if (cachedChainId) {
                const chainId = JSON.parse(cachedChainId);
                // Validate that the cached chainId is in the permitted chains list
                if (permittedChainIds.includes(chainId)) {
                    return chainId;
                }
            }
        }
        catch (error) {
            logger('Error retrieving cached chainId', error);
        }
        // Fallback to the first permitted chain if cache retrieval failed or returned an invalid chain
        return permittedChainIds[0];
    });
}, _MetamaskConnectEVM_requestInterceptor = function _MetamaskConnectEVM_requestInterceptor(request) {
    return __awaiter(this, void 0, void 0, function* () {
        logger(`Intercepting request for method: ${request.method}`);
        if (IGNORED_METHODS.includes(request.method)) {
            // TODO: replace with correct method unsupported provider error
            return Promise.reject(new Error(`Method: ${request.method} is not supported by Metamask Connect/EVM`));
        }
        if (request.method === 'wallet_revokePermissions') {
            return this.disconnect();
        }
        if (isConnectRequest(request)) {
            // When calling wallet_requestPermissions, we need to force a new session request to prompt
            // the user for accounts, because internally the Multichain SDK will check if
            // the user is already connected and skip the request if so, unless we
            // explicitly request a specific account. This is needed to workaround
            // wallet_requestPermissions not requesting specific accounts.
            const shouldForceConnectionRequest = request.method === 'wallet_requestPermissions';
            const { method, params } = request;
            const permitted = getPermittedEthChainIds(__classPrivateFieldGet(this, _MetamaskConnectEVM_sessionScopes, "f"));
            if (permitted.length === 0) {
                permitted.push(DEFAULT_CHAIN_ID);
            }
            const selected = __classPrivateFieldGet(this, _MetamaskConnectEVM_provider, "f").selectedChainId;
            const preferred = selected && permitted.includes(selected) ? selected : permitted[0];
            const chainIds = [
                preferred,
                ...permitted.filter((id) => id !== preferred),
            ];
            const scope = `eip155:${hexToNumber(preferred)}`;
            yield __classPrivateFieldGet(this, _MetamaskConnectEVM_instances, "m", _MetamaskConnectEVM_trackWalletActionRequested).call(this, method, scope, params);
            try {
                const result = yield this.connect({
                    chainIds,
                    forceRequest: shouldForceConnectionRequest,
                });
                yield __classPrivateFieldGet(this, _MetamaskConnectEVM_instances, "m", _MetamaskConnectEVM_trackWalletActionSucceeded).call(this, method, scope, params);
                return result;
            }
            catch (error) {
                yield __classPrivateFieldGet(this, _MetamaskConnectEVM_instances, "m", _MetamaskConnectEVM_trackWalletActionFailed).call(this, method, scope, params, error);
                throw error;
            }
        }
        if (isSwitchChainRequest(request)) {
            return this.switchChain({
                chainId: request.params[0].chainId,
            });
        }
        if (isAddChainRequest(request)) {
            return __classPrivateFieldGet(this, _MetamaskConnectEVM_instances, "m", _MetamaskConnectEVM_addEthereumChain).call(this, request.params[0]);
        }
        if (isAccountsRequest(request)) {
            return __classPrivateFieldGet(this, _MetamaskConnectEVM_provider, "f").accounts;
        }
        if (isChainIdRequest(request)) {
            return __classPrivateFieldGet(this, _MetamaskConnectEVM_provider, "f").selectedChainId;
        }
        logger('Request not intercepted, forwarding to default handler', request);
        return Promise.resolve();
    });
}, _MetamaskConnectEVM_clearConnectionState = function _MetamaskConnectEVM_clearConnectionState() {
    __classPrivateFieldGet(this, _MetamaskConnectEVM_provider, "f").accounts = [];
    __classPrivateFieldGet(this, _MetamaskConnectEVM_provider, "f").selectedChainId = undefined;
}, _MetamaskConnectEVM_addEthereumChain = function _MetamaskConnectEVM_addEthereumChain(chainConfiguration) {
    return __awaiter(this, void 0, void 0, function* () {
        logger('addEthereumChain called', { chainConfiguration });
        const method = 'wallet_addEthereumChain';
        if (!chainConfiguration) {
            throw new Error('No chain configuration found.');
        }
        // Get chain ID from config or use current chain
        const chainId = chainConfiguration.chainId ||
            __classPrivateFieldGet(this, _MetamaskConnectEVM_provider, "f").selectedChainId ||
            '0x1';
        const decimalChainId = hexToNumber(chainId);
        const scope = `eip155:${decimalChainId}`;
        const params = [chainConfiguration];
        yield __classPrivateFieldGet(this, _MetamaskConnectEVM_instances, "m", _MetamaskConnectEVM_trackWalletActionRequested).call(this, method, scope, params);
        try {
            const result = yield __classPrivateFieldGet(this, _MetamaskConnectEVM_instances, "m", _MetamaskConnectEVM_request).call(this, {
                method: 'wallet_addEthereumChain',
                params,
            });
            if (result.result === null) {
                // if result is successful we eagerly call onChainChanged to update the provider's selected chain ID.
                yield __classPrivateFieldGet(this, _MetamaskConnectEVM_instances, "m", _MetamaskConnectEVM_cacheChainId).call(this, chainId);
                __classPrivateFieldGet(this, _MetamaskConnectEVM_instances, "m", _MetamaskConnectEVM_onChainChanged).call(this, chainId);
            }
            yield __classPrivateFieldGet(this, _MetamaskConnectEVM_instances, "m", _MetamaskConnectEVM_trackWalletActionSucceeded).call(this, method, scope, params);
        }
        catch (error) {
            yield __classPrivateFieldGet(this, _MetamaskConnectEVM_instances, "m", _MetamaskConnectEVM_trackWalletActionFailed).call(this, method, scope, params, error);
            throw error;
        }
    });
}, _MetamaskConnectEVM_request = function _MetamaskConnectEVM_request(request) {
    return __awaiter(this, void 0, void 0, function* () {
        logger('direct request to metamask-provider called', request);
        const result = __classPrivateFieldGet(this, _MetamaskConnectEVM_core, "f").transport.sendEip1193Message(request);
        if (request.method === 'wallet_addEthereumChain' ||
            request.method === 'wallet_switchEthereumChain') {
            __classPrivateFieldGet(this, _MetamaskConnectEVM_core, "f").openSimpleDeeplinkIfNeeded();
        }
        return result;
    });
}, _MetamaskConnectEVM_cacheChainId = function _MetamaskConnectEVM_cacheChainId(chainId) {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            yield __classPrivateFieldGet(this, _MetamaskConnectEVM_core, "f").storage.adapter.set(CHAIN_STORE_KEY, JSON.stringify(chainId));
        }
        catch (error) {
            logger('Error caching chainId', error);
        }
    });
}, _MetamaskConnectEVM_onSessionChanged = function _MetamaskConnectEVM_onSessionChanged(session) {
    return __awaiter(this, void 0, void 0, function* () {
        var _a;
        logger('event: wallet_sessionChanged', session);
        __classPrivateFieldSet(this, _MetamaskConnectEVM_sessionScopes, (_a = session === null || session === void 0 ? void 0 : session.sessionScopes) !== null && _a !== void 0 ? _a : {}, "f");
        const hexPermittedChainIds = getPermittedEthChainIds(__classPrivateFieldGet(this, _MetamaskConnectEVM_sessionScopes, "f"));
        if (hexPermittedChainIds.length === 0) {
            __classPrivateFieldGet(this, _MetamaskConnectEVM_instances, "m", _MetamaskConnectEVM_onDisconnect).call(this);
        }
        else {
            let initialAccounts = [];
            if (__classPrivateFieldGet(this, _MetamaskConnectEVM_core, "f").status === 'connected') {
                const ethAccountsResponse = yield __classPrivateFieldGet(this, _MetamaskConnectEVM_core, "f").transport.sendEip1193Message({
                    method: 'eth_accounts',
                    params: [],
                });
                initialAccounts = ethAccountsResponse.result;
            }
            else {
                initialAccounts = getEthAccounts(__classPrivateFieldGet(this, _MetamaskConnectEVM_sessionScopes, "f"));
            }
            const chainId = yield __classPrivateFieldGet(this, _MetamaskConnectEVM_instances, "m", _MetamaskConnectEVM_getSelectedChainId).call(this, hexPermittedChainIds);
            __classPrivateFieldGet(this, _MetamaskConnectEVM_instances, "m", _MetamaskConnectEVM_onConnect).call(this, {
                chainId,
                accounts: initialAccounts,
            });
        }
        __classPrivateFieldGet(this, _MetamaskConnectEVM_initPromise, "f").resolve();
    });
}, _MetamaskConnectEVM_onChainChanged = function _MetamaskConnectEVM_onChainChanged(chainId) {
    var _a, _b;
    if (chainId === __classPrivateFieldGet(this, _MetamaskConnectEVM_provider, "f").selectedChainId) {
        return;
    }
    logger('handler: chainChanged', { chainId });
    __classPrivateFieldGet(this, _MetamaskConnectEVM_provider, "f").selectedChainId = chainId;
    __classPrivateFieldGet(this, _MetamaskConnectEVM_provider, "f").emit('chainChanged', chainId);
    (_b = (_a = __classPrivateFieldGet(this, _MetamaskConnectEVM_eventHandlers, "f")) === null || _a === void 0 ? void 0 : _a.chainChanged) === null || _b === void 0 ? void 0 : _b.call(_a, chainId);
}, _MetamaskConnectEVM_onAccountsChanged = function _MetamaskConnectEVM_onAccountsChanged(accounts) {
    var _a, _b;
    const accountsUnchanged = accounts.length === __classPrivateFieldGet(this, _MetamaskConnectEVM_provider, "f").accounts.length &&
        accounts.every((acct, idx) => acct === __classPrivateFieldGet(this, _MetamaskConnectEVM_provider, "f").accounts[idx]);
    if (accountsUnchanged) {
        return;
    }
    logger('handler: accountsChanged', accounts);
    __classPrivateFieldGet(this, _MetamaskConnectEVM_provider, "f").accounts = accounts;
    __classPrivateFieldGet(this, _MetamaskConnectEVM_provider, "f").emit('accountsChanged', accounts);
    (_b = (_a = __classPrivateFieldGet(this, _MetamaskConnectEVM_eventHandlers, "f")) === null || _a === void 0 ? void 0 : _a.accountsChanged) === null || _b === void 0 ? void 0 : _b.call(_a, accounts);
}, _MetamaskConnectEVM_onConnect = function _MetamaskConnectEVM_onConnect({ chainId, accounts, }) {
    var _a, _b, _c, _d, _e, _f, _g;
    logger('handler: connect', { chainId, accounts });
    const data = {
        chainId,
        accounts,
    };
    if (__classPrivateFieldGet(this, _MetamaskConnectEVM_status, "f") === 'connected') {
        __classPrivateFieldGet(this, _MetamaskConnectEVM_instances, "m", _MetamaskConnectEVM_onChainChanged).call(this, chainId);
        __classPrivateFieldGet(this, _MetamaskConnectEVM_instances, "m", _MetamaskConnectEVM_onAccountsChanged).call(this, accounts);
        return;
    }
    __classPrivateFieldGet(this, _MetamaskConnectEVM_provider, "f").selectedChainId = chainId;
    __classPrivateFieldGet(this, _MetamaskConnectEVM_provider, "f").accounts = accounts;
    __classPrivateFieldSet(this, _MetamaskConnectEVM_status, 'connected', "f");
    __classPrivateFieldGet(this, _MetamaskConnectEVM_provider, "f").emit('connect', data);
    (_b = (_a = __classPrivateFieldGet(this, _MetamaskConnectEVM_eventHandlers, "f")) === null || _a === void 0 ? void 0 : _a.connect) === null || _b === void 0 ? void 0 : _b.call(_a, data);
    __classPrivateFieldGet(this, _MetamaskConnectEVM_provider, "f").emit('chainChanged', chainId);
    (_d = (_c = __classPrivateFieldGet(this, _MetamaskConnectEVM_eventHandlers, "f")) === null || _c === void 0 ? void 0 : _c.chainChanged) === null || _d === void 0 ? void 0 : _d.call(_c, chainId);
    __classPrivateFieldGet(this, _MetamaskConnectEVM_provider, "f").emit('accountsChanged', accounts);
    (_f = (_e = __classPrivateFieldGet(this, _MetamaskConnectEVM_eventHandlers, "f")) === null || _e === void 0 ? void 0 : _e.accountsChanged) === null || _f === void 0 ? void 0 : _f.call(_e, accounts);
    (_g = __classPrivateFieldGet(this, _MetamaskConnectEVM_removeNotificationHandler, "f")) === null || _g === void 0 ? void 0 : _g.call(this);
    const onAccountsChanged = (accs) => {
        logger('core-event: accountsChanged', accs);
        __classPrivateFieldGet(this, _MetamaskConnectEVM_instances, "m", _MetamaskConnectEVM_onAccountsChanged).call(this, accs);
    };
    const onChainChanged = (chainChanged) => {
        logger('core-event: chainChanged', chainChanged.chainId);
        __classPrivateFieldGet(this, _MetamaskConnectEVM_instances, "m", _MetamaskConnectEVM_cacheChainId).call(this, chainChanged.chainId).catch((error) => {
            logger('Error caching chainId in notification handler', error);
        });
        __classPrivateFieldGet(this, _MetamaskConnectEVM_instances, "m", _MetamaskConnectEVM_onChainChanged).call(this, chainChanged.chainId);
    };
    __classPrivateFieldGet(this, _MetamaskConnectEVM_core, "f").on('metamask_accountsChanged', onAccountsChanged);
    __classPrivateFieldGet(this, _MetamaskConnectEVM_core, "f").on('metamask_chainChanged', onChainChanged);
    __classPrivateFieldSet(this, _MetamaskConnectEVM_removeNotificationHandler, () => {
        __classPrivateFieldGet(this, _MetamaskConnectEVM_core, "f").off('metamask_accountsChanged', onAccountsChanged);
        __classPrivateFieldGet(this, _MetamaskConnectEVM_core, "f").off('metamask_chainChanged', onChainChanged);
    }, "f");
}, _MetamaskConnectEVM_onDisconnect = function _MetamaskConnectEVM_onDisconnect() {
    var _a, _b;
    if (__classPrivateFieldGet(this, _MetamaskConnectEVM_status, "f") === 'disconnected') {
        return;
    }
    __classPrivateFieldSet(this, _MetamaskConnectEVM_status, 'disconnected', "f");
    logger('handler: disconnect');
    __classPrivateFieldGet(this, _MetamaskConnectEVM_provider, "f").emit('disconnect');
    (_b = (_a = __classPrivateFieldGet(this, _MetamaskConnectEVM_eventHandlers, "f")) === null || _a === void 0 ? void 0 : _a.disconnect) === null || _b === void 0 ? void 0 : _b.call(_a);
    __classPrivateFieldGet(this, _MetamaskConnectEVM_instances, "m", _MetamaskConnectEVM_onAccountsChanged).call(this, []);
}, _MetamaskConnectEVM_onDisplayUri = function _MetamaskConnectEVM_onDisplayUri(uri) {
    var _a, _b;
    if (__classPrivateFieldGet(this, _MetamaskConnectEVM_status, "f") !== 'connecting') {
        return;
    }
    logger('handler: display_uri', uri);
    __classPrivateFieldGet(this, _MetamaskConnectEVM_provider, "f").emit('display_uri', uri);
    (_b = (_a = __classPrivateFieldGet(this, _MetamaskConnectEVM_eventHandlers, "f")) === null || _a === void 0 ? void 0 : _a.displayUri) === null || _b === void 0 ? void 0 : _b.call(_a, uri);
};
/**
 * Creates a new Metamask Connect/EVM instance
 *
 * @param options - The options for the Metamask Connect/EVM layer
 * @param options.dapp - Dapp identification and branding settings
 * @param options.api - API configuration including read-only RPC map
 * @param options.api.supportedNetworks - A map of hex chain IDs to RPC URLs for read-only requests
 * @param [options.analytics.integrationType] - Integration type for analytics
 * @param [options.ui] - UI configuration options
 * @param [options.ui.headless] - Whether to run without UI
 * @param [options.ui.preferExtension] - Whether to prefer browser extension
 * @param [options.ui.showInstallModal] - Whether to render installation modal for desktop extension
 * @param [options.mobile] - Mobile configuration options
 * @param [options.mobile.preferredOpenLink] - Custom handler for opening deeplinks (useful for React Native, etc.)
 * @param [options.mobile.useDeeplink] - Whether to use native deeplinks instead of universal links
 * @param [options.transport] - Transport configuration (e.g., extensionId, notification handler)
 * @param [options.transport.extensionId] - Extension ID for browser extension transport
 * @param [options.transport.onNotification] - Callback for receiving transport notifications
 * @param [options.eventHandlers] - Event handlers for the Metamask Connect/EVM layer
 * @param [options.debug] - Enable debug logging
 * @returns The Metamask-Connect EVM client instance
 */
export function createEVMClient(options) {
    return __awaiter(this, void 0, void 0, function* () {
        var _a, _b;
        if (options.debug) {
            enableDebug();
        }
        logger('Creating Metamask Connect/EVM with options:', options);
        // Validate that supportedNetworks is provided and not empty
        if (!((_a = options.api) === null || _a === void 0 ? void 0 : _a.supportedNetworks) ||
            Object.keys(options.api.supportedNetworks).length === 0) {
            throw new Error('supportedNetworks is required and must contain at least one chain configuration');
        }
        validSupportedChainsUrls(options.api.supportedNetworks, 'supportedNetworks');
        const supportedNetworksCaipChainId = Object.entries(options.api.supportedNetworks).reduce((acc, [hexChainId, url]) => {
            const decimalChainId = parseInt(hexChainId, 16);
            const caip2ChainId = `eip155:${decimalChainId}`;
            acc[caip2ChainId] = url;
            return acc;
        }, {});
        try {
            const core = yield createMultichainClient(Object.assign(Object.assign({}, options), { api: {
                    supportedNetworks: supportedNetworksCaipChainId,
                }, analytics: {
                    // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
                    integrationType: ((_b = options.analytics) === null || _b === void 0 ? void 0 : _b.integrationType) || 'direct',
                }, versions: {
                    // typeof guard needed: Metro (React Native) bundles TS source directly,
                    // bypassing the tsup build that substitutes __PACKAGE_VERSION__.
                    'connect-evm': typeof __PACKAGE_VERSION__ === 'undefined'
                        ? 'unknown'
                        : __PACKAGE_VERSION__,
                } }));
            return MetamaskConnectEVM.create({
                core,
                eventHandlers: options.eventHandlers,
                supportedNetworks: options.api.supportedNetworks,
            });
        }
        catch (error) {
            console.error('Error creating Metamask Connect/EVM', error);
            throw error;
        }
    });
}
