// @metamask/connect-evm/dist/src/provider.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _EIP1193Provider_core, _EIP1193Provider_requestInterceptor, _EIP1193Provider_accounts, _EIP1193Provider_selectedChainId;
import { EventEmitter, RPCInvokeMethodErr } from '@metamask/connect-multichain';
import { hexToNumber } from '@metamask/utils';
import { INTERCEPTABLE_METHODS } from './constants';
import { logger } from './logger';
/**
 * EIP-1193 Provider wrapper around the Multichain SDK.
 */
export class EIP1193Provider extends EventEmitter {
    constructor(core, interceptor) {
        super();
        /** The core instance of the Multichain SDK */
        _EIP1193Provider_core.set(this, void 0);
        /** Interceptor function to handle specific methods */
        _EIP1193Provider_requestInterceptor.set(this, void 0);
        /** The currently permitted accounts */
        _EIP1193Provider_accounts.set(this, []);
        /** The currently selected chain ID on the wallet */
        _EIP1193Provider_selectedChainId.set(this, void 0);
        __classPrivateFieldSet(this, _EIP1193Provider_core, core, "f");
        __classPrivateFieldSet(this, _EIP1193Provider_requestInterceptor, interceptor, "f");
        // Bind all public methods to ensure `this` context is preserved
        // when methods are extracted or passed as callbacks.
        // This eliminates the need for Proxy wrappers in consumers.
        this.request = this.request.bind(this);
        this.sendAsync = this.sendAsync.bind(this);
        this.send = this.send.bind(this);
        // Bind inherited EventEmitter methods
        this.on = this.on.bind(this);
        this.off = this.off.bind(this);
        this.emit = this.emit.bind(this);
        this.once = this.once.bind(this);
        this.removeListener = this.removeListener.bind(this);
        this.listenerCount = this.listenerCount.bind(this);
    }
    /**
     * Performs a EIP-1193 request.
     *
     * @param request - The request object containing the method and params
     * @returns The result of the request
     */
    request(request) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b, _c, _d;
            logger(`request: ${request.method} - chainId: ${this.selectedChainId}`, request.params);
            /* Some methods require special handling, so we intercept them here
             * and handle them in MetamaskConnectEVM.requestInterceptor method.  */
            if (INTERCEPTABLE_METHODS.includes(request.method)) {
                return (_a = __classPrivateFieldGet(this, _EIP1193Provider_requestInterceptor, "f")) === null || _a === void 0 ? void 0 : _a.call(this, request);
            }
            if (!__classPrivateFieldGet(this, _EIP1193Provider_selectedChainId, "f")) {
                // TODO: replace with a better error
                throw new Error('No chain ID selected');
            }
            const decimalChainId = hexToNumber(__classPrivateFieldGet(this, _EIP1193Provider_selectedChainId, "f"));
            const scope = `eip155:${decimalChainId}`;
            // Validate that the chain is configured in supportedNetworks
            // This check is performed here to provide better error messages
            // The RpcClient will also validate, but this gives us a chance to provide
            // a clearer error message before the request is routed
            const coreOptions = __classPrivateFieldGet(this, _EIP1193Provider_core, "f").options; // TODO: options is `protected readonly` property, this needs to be refactored so `any` type assertion is not necessary
            const supportedNetworks = (_c = (_b = coreOptions === null || coreOptions === void 0 ? void 0 : coreOptions.api) === null || _b === void 0 ? void 0 : _b.supportedNetworks) !== null && _c !== void 0 ? _c : {};
            if (!supportedNetworks[scope]) {
                throw new Error(`Chain ${scope} is not configured in supportedNetworks. Requests cannot be made to chains not explicitly configured in supportedNetworks.`);
            }
            try {
                return yield __classPrivateFieldGet(this, _EIP1193Provider_core, "f").invokeMethod({
                    scope,
                    request: {
                        method: request.method,
                        params: request.params,
                    },
                });
            }
            catch (error) {
                if (error instanceof RPCInvokeMethodErr && error.rpcCode !== undefined) {
                    const rpcError = new Error((_d = error.rpcMessage) !== null && _d !== void 0 ? _d : error.reason);
                    rpcError.code = error.rpcCode;
                    throw rpcError;
                }
                throw error;
            }
        });
    }
    // Getters and setters
    get selectedAccount() {
        return this.accounts[0];
    }
    set accounts(accounts) {
        __classPrivateFieldSet(this, _EIP1193Provider_accounts, accounts, "f");
    }
    get accounts() {
        return __classPrivateFieldGet(this, _EIP1193Provider_accounts, "f");
    }
    get selectedChainId() {
        return __classPrivateFieldGet(this, _EIP1193Provider_selectedChainId, "f");
    }
    set selectedChainId(chainId) {
        __classPrivateFieldSet(this, _EIP1193Provider_selectedChainId, chainId, "f");
    }
    // ==========================================
    // Legacy compatibility methods
    // ==========================================
    /**
     * Alias for selectedChainId for legacy compatibility.
     * Many dApps expect a `chainId` property on the provider.
     */
    get chainId() {
        return this.selectedChainId;
    }
    /**
     * Legacy method for sending JSON-RPC requests.
     *
     * @deprecated Use `request` instead. This method is provided for backwards compatibility.
     * @param request - The JSON-RPC request object
     * @param callback - Optional callback function. If provided, the method returns void.
     * @returns A promise resolving to the JSON-RPC response, or void if a callback is provided.
     */
    sendAsync(request, callback) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a;
            const id = (_a = request.id) !== null && _a !== void 0 ? _a : 1;
            const promise = this.request({
                method: request.method,
                params: request.params,
            })
                .then((result) => ({
                id,
                jsonrpc: '2.0',
                result: result,
            }))
                .catch((error) => {
                var _a, _b;
                return ({
                    id,
                    jsonrpc: '2.0',
                    error: {
                        code: (_a = error.code) !== null && _a !== void 0 ? _a : -32603,
                        message: (_b = error.message) !== null && _b !== void 0 ? _b : 'Internal error',
                        data: error.data,
                    },
                });
            });
            if (callback) {
                promise
                    .then((response) => {
                    if (response.error) {
                        callback(new Error(response.error.message), response);
                    }
                    else {
                        callback(null, response);
                    }
                })
                    .catch((error) => {
                    callback(error, null);
                });
                return;
            }
            return promise;
        });
    }
    /**
     * Legacy method for sending JSON-RPC requests synchronously (callback-based).
     *
     * @deprecated Use `request` instead. This method is provided for backwards compatibility.
     * @param request - The JSON-RPC request object
     * @param callback - The callback function to receive the response
     */
    send(request, callback) {
        this.sendAsync(request, callback);
    }
}
_EIP1193Provider_core = new WeakMap(), _EIP1193Provider_requestInterceptor = new WeakMap(), _EIP1193Provider_accounts = new WeakMap(), _EIP1193Provider_selectedChainId = new WeakMap();
