// @metamask/connect-evm/dist/src/utils/type-guards.js
/**
 * Type guard for connect-like requests:
 * - wallet_requestPermissions
 * - eth_requestAccounts
 *
 * @param req - The request object to check
 * @returns True if the request is a connect-like request, false otherwise
 */
export function isConnectRequest(req) {
    return (req.method === 'wallet_requestPermissions' ||
        req.method === 'eth_requestAccounts');
}
/**
 * Type guard for wallet_switchEthereumChain request.
 *
 * @param req - The request object to check
 * @returns True if the request is a wallet_switchEthereumChain request, false otherwise
 */
export function isSwitchChainRequest(req) {
    return req.method === 'wallet_switchEthereumChain';
}
/**
 * Type guard for wallet_addEthereumChain request.
 *
 * @param req - The request object to check
 * @returns True if the request is a wallet_addEthereumChain request, false otherwise
 */
export function isAddChainRequest(req) {
    return req.method === 'wallet_addEthereumChain';
}
/**
 * Type guard for generic accounts request:
 * - eth_accounts
 * - eth_coinbase
 *
 * @param req - The request object to check
 * @returns True if the request is a generic accounts request, false otherwise
 */
export function isAccountsRequest(req) {
    return req.method === 'eth_accounts' || req.method === 'eth_coinbase';
}
/**
 * Type guard for generic eth_chainId request.
 *
 * @param req - The request object to check
 * @returns True if the request is a eth_chainId request, false otherwise
 */
export function isChainIdRequest(req) {
    return req.method === 'eth_chainId';
}
/**
 * Validates that all values in a Record are valid URLs.
 *
 * @param record - The record to validate (e.g., supportedNetworks)
 * @param recordName - The name of the record for error messages
 * @throws Error if any values are invalid URLs
 */
export function validSupportedChainsUrls(record, recordName) {
    const invalidUrls = [];
    for (const [key, url] of Object.entries(record)) {
        try {
            // eslint-disable-next-line no-new
            new URL(url);
        }
        catch (_a) {
            invalidUrls.push(`${key}: ${url}`);
        }
    }
    if (invalidUrls.length > 0) {
        throw new Error(`${recordName} contains invalid URLs:\n${invalidUrls.join('\n')}`);
    }
}
