// @metamask/chain-agnostic-permission/dist/caip25Permission.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.requestPermittedChainsPermissionIncremental = exports.getCaip25PermissionFromLegacyPermissions = exports.getCaip25CaveatFromPermission = exports.generateCaip25Caveat = exports.Caip25CaveatMutators = exports.caip25EndowmentBuilder = exports.caip25CaveatBuilder = exports.diffScopesForCaip25CaveatValue = exports.createCaip25Caveat = exports.Caip25EndowmentPermissionName = exports.Caip25CaveatType = void 0;
const permission_controller_1 = require("@metamask/permission-controller");
const utils_1 = require("@metamask/utils");
const lodash_1 = require("lodash");
const constants_1 = require("./constants.cjs");
const caip_permission_operator_accounts_1 = require("./operators/caip-permission-operator-accounts.cjs");
const caip_permission_operator_permittedChains_1 = require("./operators/caip-permission-operator-permittedChains.cjs");
const assert_1 = require("./scope/assert.cjs");
const supported_1 = require("./scope/supported.cjs");
const transform_1 = require("./scope/transform.cjs");
const types_1 = require("./scope/types.cjs");
/**
 * The name of the CAIP-25 permission caveat.
 */
exports.Caip25CaveatType = 'authorizedScopes';
/**
 * The target name of the CAIP-25 endowment permission.
 */
exports.Caip25EndowmentPermissionName = 'endowment:caip25';
/**
 * Creates a CAIP-25 permission caveat.
 *
 * @param value - The CAIP-25 permission caveat value.
 * @returns The CAIP-25 permission caveat (now including the type).
 */
const createCaip25Caveat = (value) => {
    return {
        type: exports.Caip25CaveatType,
        value,
    };
};
exports.createCaip25Caveat = createCaip25Caveat;
/**
 * Calculates the difference between two provided CAIP-25 permission caveat values, but only considering a single scope property at a time.
 *
 * @param originalValue - The existing CAIP-25 permission caveat value.
 * @param mergedValue - The result from merging existing and incoming CAIP-25 permission caveat values.
 * @param scopeToDiff - The required or optional scopes from the [CAIP-25](https://github.com/ChainAgnostic/CAIPs/blob/main/CAIPs/caip-25.md) request.
 * @returns The difference between original and merged CAIP-25 permission caveat values.
 */
function diffScopesForCaip25CaveatValue(originalValue, mergedValue, scopeToDiff) {
    const diff = (0, lodash_1.cloneDeep)(originalValue);
    const mergedScopeToDiff = mergedValue[scopeToDiff];
    for (const [scopeString, mergedScopeObject] of Object.entries(mergedScopeToDiff)) {
        const internalScopeString = scopeString;
        const originalScopeObject = diff[scopeToDiff][internalScopeString];
        if (originalScopeObject) {
            const newAccounts = mergedScopeObject.accounts.filter((account) => !originalScopeObject?.accounts.includes(account));
            if (newAccounts.length > 0) {
                diff[scopeToDiff][internalScopeString] = {
                    accounts: newAccounts,
                };
                continue;
            }
            delete diff[scopeToDiff][internalScopeString];
        }
        else {
            diff[scopeToDiff][internalScopeString] = mergedScopeObject;
        }
    }
    return diff;
}
exports.diffScopesForCaip25CaveatValue = diffScopesForCaip25CaveatValue;
/**
 * Checks if every account in the given scopes object is supported.
 *
 * @param scopesObject - The scopes object to iterate over.
 * @param listAccounts - The hook for getting internalAccount objects for all evm accounts.
 * @param getNonEvmAccountAddresses - The hook that returns the supported CAIP-10 account addresses for a non EVM scope.
 * addresses.
 * @returns True if every account in the scopes object is supported, false otherwise.
 */
function isEveryAccountInScopesObjectSupported(scopesObject, listAccounts, getNonEvmAccountAddresses) {
    return Object.values(scopesObject).every((scopeObject) => scopeObject.accounts.every((account) => (0, supported_1.isSupportedAccount)(account, {
        getEvmInternalAccounts: listAccounts,
        getNonEvmAccountAddresses,
    })));
}
/**
 * Helper that returns a `authorizedScopes` CAIP-25 caveat specification
 * that can be passed into the PermissionController constructor.
 *
 * @param options - The specification builder options.
 * @param options.findNetworkClientIdByChainId - The hook for getting the networkClientId that serves a chainId.
 * @param options.listAccounts - The hook for getting internalAccount objects for all evm accounts.
 * @param options.isNonEvmScopeSupported - The hook that determines if an non EVM scopeString is supported.
 * @param options.getNonEvmAccountAddresses - The hook that returns the supported CAIP-10 account addresses for a non EVM scope.
 * @returns The specification for the `caip25` caveat.
 */
const caip25CaveatBuilder = ({ findNetworkClientIdByChainId, listAccounts, isNonEvmScopeSupported, getNonEvmAccountAddresses, }) => {
    return {
        type: exports.Caip25CaveatType,
        validator: (caveat, _origin, _target) => {
            if (!caveat.value ||
                !(0, utils_1.hasProperty)(caveat.value, 'requiredScopes') ||
                !(0, utils_1.hasProperty)(caveat.value, 'optionalScopes') ||
                !(0, utils_1.hasProperty)(caveat.value, 'isMultichainOrigin') ||
                !(0, utils_1.hasProperty)(caveat.value, 'sessionProperties') ||
                typeof caveat.value.isMultichainOrigin !== 'boolean' ||
                !(0, utils_1.isObject)(caveat.value.sessionProperties)) {
                throw new Error(`${exports.Caip25EndowmentPermissionName} error: Received invalid value for caveat of type "${exports.Caip25CaveatType}".`);
            }
            const { requiredScopes, optionalScopes, sessionProperties } = caveat.value;
            const allSessionPropertiesSupported = Object.keys(sessionProperties).every((sessionProperty) => (0, supported_1.isSupportedSessionProperty)(sessionProperty));
            if (!allSessionPropertiesSupported) {
                throw new Error(`${exports.Caip25EndowmentPermissionName} error: Received unknown session property(s) for caveat of type "${exports.Caip25CaveatType}".`);
            }
            (0, assert_1.assertIsInternalScopesObject)(requiredScopes);
            (0, assert_1.assertIsInternalScopesObject)(optionalScopes);
            if (Object.keys(requiredScopes).length === 0 &&
                Object.keys(optionalScopes).length === 0) {
                throw new Error(`${exports.Caip25EndowmentPermissionName} error: Received no scopes for caveat of type "${exports.Caip25CaveatType}".`);
            }
            const isEvmChainIdSupported = (chainId) => {
                try {
                    findNetworkClientIdByChainId(chainId);
                    return true;
                }
                catch {
                    return false;
                }
            };
            const unsupportedScopes = Object.keys({
                ...requiredScopes,
                ...optionalScopes,
            }).filter((scopeString) => !(0, supported_1.isSupportedScopeString)(scopeString, {
                isEvmChainIdSupported,
                isNonEvmScopeSupported,
            }));
            if (unsupportedScopes.length > 0) {
                throw new Error(`${exports.Caip25EndowmentPermissionName} error: Received scopeString value(s): ${unsupportedScopes.join(', ')} for caveat of type "${exports.Caip25CaveatType}" that are not supported by the wallet.`);
            }
            const allRequiredAccountsSupported = isEveryAccountInScopesObjectSupported(requiredScopes, listAccounts, getNonEvmAccountAddresses);
            const allOptionalAccountsSupported = isEveryAccountInScopesObjectSupported(optionalScopes, listAccounts, getNonEvmAccountAddresses);
            if (!allRequiredAccountsSupported || !allOptionalAccountsSupported) {
                throw new Error(`${exports.Caip25EndowmentPermissionName} error: Received account value(s) for caveat of type "${exports.Caip25CaveatType}" that are not supported by the wallet.`);
            }
        },
        merger: (leftValue, rightValue) => {
            const mergedRequiredScopes = (0, transform_1.mergeInternalScopes)(leftValue.requiredScopes, rightValue.requiredScopes);
            const mergedOptionalScopes = (0, transform_1.mergeInternalScopes)(leftValue.optionalScopes, rightValue.optionalScopes);
            const mergedSessionProperties = {
                ...leftValue.sessionProperties,
                ...rightValue.sessionProperties,
            };
            const mergedValue = {
                requiredScopes: mergedRequiredScopes,
                optionalScopes: mergedOptionalScopes,
                sessionProperties: mergedSessionProperties,
                isMultichainOrigin: leftValue.isMultichainOrigin,
            };
            const partialDiff = diffScopesForCaip25CaveatValue(leftValue, mergedValue, 'requiredScopes');
            const diff = diffScopesForCaip25CaveatValue(partialDiff, mergedValue, 'optionalScopes');
            return [mergedValue, diff];
        },
    };
};
exports.caip25CaveatBuilder = caip25CaveatBuilder;
/**
 * Helper that returns a `endowment:caip25` specification that
 * can be passed into the PermissionController constructor.
 *
 * @returns The specification for the `caip25` endowment.
 */
const specificationBuilder = () => {
    return {
        permissionType: permission_controller_1.PermissionType.Endowment,
        targetName: exports.Caip25EndowmentPermissionName,
        allowedCaveats: [exports.Caip25CaveatType],
        endowmentGetter: (_getterOptions) => null,
        validator: (permission) => {
            if (permission.caveats?.length !== 1 ||
                permission.caveats?.[0]?.type !== exports.Caip25CaveatType) {
                throw new Error(`${exports.Caip25EndowmentPermissionName} error: Invalid caveats. There must be a single caveat of type "${exports.Caip25CaveatType}".`);
            }
        },
    };
};
/**
 * The `caip25` endowment specification builder. Passed to the
 * `PermissionController` for constructing and validating the
 * `endowment:caip25` permission.
 */
exports.caip25EndowmentBuilder = Object.freeze({
    targetName: exports.Caip25EndowmentPermissionName,
    specificationBuilder,
});
/**
 * Factories that construct caveat mutator functions that are passed to
 * PermissionController.updatePermissionsByCaveat.
 */
exports.Caip25CaveatMutators = {
    [exports.Caip25CaveatType]: {
        removeScope,
        removeAccount,
    },
};
/**
 * Removes the account from the scope object.
 *
 * @param targetAddress - The address to remove from the scope object.
 * @returns A function that removes the account from the scope object.
 */
function removeAccountFilterFn(targetAddress) {
    return (account) => {
        const parsed = (0, utils_1.parseCaipAccountId)(account);
        return parsed.address !== targetAddress;
    };
}
/**
 * Removes the account from the scope object.
 *
 * @param scopeObject - The scope object to remove the account from.
 * @param targetAddress - The address to remove from the scope object.
 */
function removeAccountFromScopeObject(scopeObject, targetAddress) {
    if (scopeObject.accounts) {
        scopeObject.accounts = scopeObject.accounts.filter(removeAccountFilterFn(targetAddress));
    }
}
/**
 * Removes the target account from the scope object.
 *
 * @param caip25CaveatValue - The CAIP-25 permission caveat value from which to remove the account (across all chain scopes).
 * @param targetAddress - The address to remove from the scope object. Not a CAIP-10 formatted address because it will be removed across each chain scope.
 * @returns The updated scope object.
 */
function removeAccount(caip25CaveatValue, targetAddress) {
    const updatedCaveatValue = (0, lodash_1.cloneDeep)(caip25CaveatValue);
    [
        updatedCaveatValue.requiredScopes,
        updatedCaveatValue.optionalScopes,
    ].forEach((scopes) => {
        Object.entries(scopes).forEach(([, scopeObject]) => {
            removeAccountFromScopeObject(scopeObject, targetAddress);
        });
    });
    const noChange = (0, lodash_1.isEqual)(updatedCaveatValue, caip25CaveatValue);
    if (noChange) {
        return {
            operation: permission_controller_1.CaveatMutatorOperation.Noop,
        };
    }
    const hasAccounts = [
        ...Object.values(updatedCaveatValue.requiredScopes),
        ...Object.values(updatedCaveatValue.optionalScopes),
    ].some(({ accounts }) => accounts.length > 0);
    if (hasAccounts) {
        return {
            operation: permission_controller_1.CaveatMutatorOperation.UpdateValue,
            value: updatedCaveatValue,
        };
    }
    return {
        operation: permission_controller_1.CaveatMutatorOperation.RevokePermission,
    };
}
/**
 * Removes the target scope from the value arrays of the given
 * `endowment:caip25` caveat. No-ops if the target scopeString is not in
 * the existing scopes.
 *
 * @param caip25CaveatValue - The CAIP-25 permission caveat value to remove the scope from.
 * @param targetScopeString - The scope that is being removed.
 * @returns The updated CAIP-25 permission caveat value.
 */
function removeScope(caip25CaveatValue, targetScopeString) {
    const newRequiredScopes = Object.entries(caip25CaveatValue.requiredScopes).filter(([scope]) => scope !== targetScopeString);
    const newOptionalScopes = Object.entries(caip25CaveatValue.optionalScopes).filter(([scope]) => {
        return scope !== targetScopeString;
    });
    const requiredScopesRemoved = newRequiredScopes.length !==
        Object.keys(caip25CaveatValue.requiredScopes).length;
    const optionalScopesRemoved = newOptionalScopes.length !==
        Object.keys(caip25CaveatValue.optionalScopes).length;
    if (!requiredScopesRemoved && !optionalScopesRemoved) {
        return {
            operation: permission_controller_1.CaveatMutatorOperation.Noop,
        };
    }
    const updatedCaveatValue = {
        ...caip25CaveatValue,
        requiredScopes: Object.fromEntries(newRequiredScopes),
        optionalScopes: Object.fromEntries(newOptionalScopes),
    };
    const hasNonWalletScopes = [...newRequiredScopes, ...newOptionalScopes].some(([scopeString]) => {
        const { namespace } = (0, types_1.parseScopeString)(scopeString);
        return namespace !== utils_1.KnownCaipNamespace.Wallet;
    });
    if (hasNonWalletScopes) {
        return {
            operation: permission_controller_1.CaveatMutatorOperation.UpdateValue,
            value: updatedCaveatValue,
        };
    }
    return {
        operation: permission_controller_1.CaveatMutatorOperation.RevokePermission,
    };
}
/**
 * Modifies the requested CAIP-25 permissions object after UI confirmation.
 *
 * @param caip25CaveatValue - The requested CAIP-25 caveat value to modify.
 * @param accountAddresses - The list of permitted eth addresses.
 * @param chainIds - The list of permitted eth chainIds.
 * @returns The updated CAIP-25 caveat value with the permitted accounts and chainIds set.
 */
const generateCaip25Caveat = (caip25CaveatValue, accountAddresses, chainIds) => {
    const caveatValueWithChains = (0, caip_permission_operator_permittedChains_1.setChainIdsInCaip25CaveatValue)(caip25CaveatValue, chainIds);
    const caveatValueWithAccounts = (0, caip_permission_operator_accounts_1.setNonSCACaipAccountIdsInCaip25CaveatValue)(caveatValueWithChains, accountAddresses);
    return {
        [exports.Caip25EndowmentPermissionName]: {
            caveats: [
                {
                    type: exports.Caip25CaveatType,
                    value: caveatValueWithAccounts,
                },
            ],
        },
    };
};
exports.generateCaip25Caveat = generateCaip25Caveat;
/**
 * Helper to get the CAIP-25 caveat from a permission
 *
 * @param [caip25Permission] - The CAIP-25 permission object
 * @param caip25Permission.caveats - The caveats of the CAIP-25 permission
 * @returns The CAIP-25 caveat or undefined if not found
 */
function getCaip25CaveatFromPermission(caip25Permission) {
    return caip25Permission?.caveats.find((caveat) => caveat.type === exports.Caip25CaveatType);
}
exports.getCaip25CaveatFromPermission = getCaip25CaveatFromPermission;
/**
 * Requests user approval for the CAIP-25 permission
 * and returns a granted permissions object.
 *
 * @param requestedPermissions - The legacy permissions to request approval for.
 * @param requestedPermissions.caveats - The legacy caveats processed by the function.
 * - `restrictReturnedAccounts`: Restricts which Ethereum accounts can be accessed
 * - `restrictNetworkSwitching`: Restricts which blockchain networks can be used
 * @returns The converted CAIP-25 permission object.
 */
const getCaip25PermissionFromLegacyPermissions = (requestedPermissions) => {
    const permissions = (0, lodash_1.pick)(requestedPermissions, [
        constants_1.PermissionKeys.eth_accounts,
        constants_1.PermissionKeys.permittedChains,
    ]);
    if (!permissions[constants_1.PermissionKeys.eth_accounts]) {
        permissions[constants_1.PermissionKeys.eth_accounts] = {};
    }
    if (!permissions[constants_1.PermissionKeys.permittedChains]) {
        permissions[constants_1.PermissionKeys.permittedChains] = {};
    }
    const requestedAccounts = permissions[constants_1.PermissionKeys.eth_accounts]?.caveats?.find((caveat) => caveat.type === constants_1.CaveatTypes.restrictReturnedAccounts)?.value ?? [];
    const requestedChains = permissions[constants_1.PermissionKeys.permittedChains]?.caveats?.find((caveat) => caveat.type === constants_1.CaveatTypes.restrictNetworkSwitching)?.value ?? [];
    const newCaveatValue = {
        requiredScopes: {},
        optionalScopes: {
            'wallet:eip155': {
                accounts: [],
            },
        },
        sessionProperties: {},
        isMultichainOrigin: false,
    };
    const caveatValueWithChains = (0, caip_permission_operator_permittedChains_1.setPermittedEthChainIds)(newCaveatValue, requestedChains);
    const caveatValueWithAccountsAndChains = (0, caip_permission_operator_accounts_1.setEthAccounts)(caveatValueWithChains, requestedAccounts);
    return {
        [exports.Caip25EndowmentPermissionName]: {
            caveats: [
                {
                    type: exports.Caip25CaveatType,
                    value: caveatValueWithAccountsAndChains,
                },
            ],
        },
    };
};
exports.getCaip25PermissionFromLegacyPermissions = getCaip25PermissionFromLegacyPermissions;
/**
 * Requests incremental permittedChains permission for the specified origin.
 * and updates the existing CAIP-25 permission.
 * Allows for granting without prompting for user approval which
 * would be used as part of flows like `wallet_addEthereumChain`
 * requests where the addition of the network and the permitting
 * of the chain are combined into one approval.
 *
 * @param options - The options object
 * @param options.origin - The origin to request approval for.
 * @param options.chainId - The chainId to add to the existing permittedChains.
 * @param options.autoApprove - If the chain should be granted without prompting for user approval.
 * @param options.metadata - Request data for the approval.
 * @param options.metadata.options - Additional metadata about the permission request.
 * @param options.hooks - Permission controller hooks for incremental operations.
 * @param options.hooks.requestPermissionsIncremental - Initiates an incremental permission request that prompts for user approval.
 * Incremental permission requests allow the caller to replace existing and/or add brand new permissions and caveats for the specified subject.
 * @param options.hooks.grantPermissionsIncremental - Incrementally grants approved permissions to the specified subject without prompting for user approval.
 * Every permission and caveat is stringently validated and an error is thrown if validation fails.
 */
const requestPermittedChainsPermissionIncremental = async ({ origin, chainId, autoApprove, hooks, metadata, }) => {
    const caveatValueWithChains = (0, caip_permission_operator_permittedChains_1.setPermittedEthChainIds)({
        requiredScopes: {},
        optionalScopes: {},
        sessionProperties: {},
        isMultichainOrigin: false,
    }, [chainId]);
    if (!autoApprove) {
        let options;
        if (metadata) {
            options = { metadata };
        }
        await hooks.requestPermissionsIncremental({ origin }, {
            [exports.Caip25EndowmentPermissionName]: {
                caveats: [
                    {
                        type: exports.Caip25CaveatType,
                        value: caveatValueWithChains,
                    },
                ],
            },
        }, options);
        return;
    }
    hooks.grantPermissionsIncremental({
        subject: { origin },
        approvedPermissions: {
            [exports.Caip25EndowmentPermissionName]: {
                caveats: [
                    {
                        type: exports.Caip25CaveatType,
                        value: caveatValueWithChains,
                    },
                ],
            },
        },
    });
};
exports.requestPermittedChainsPermissionIncremental = requestPermittedChainsPermissionIncremental;
