// @metamask/messenger/dist/Messenger.cjs
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _Messenger_instances, _Messenger_namespace, _Messenger_parent, _Messenger_actions, _Messenger_events, _Messenger_subscriptionDelegationTargets, _Messenger_actionDelegationTargets, _Messenger_initialEventPayloadGetters, _Messenger_eventPayloadCache, _Messenger_registerActionHandler, _Messenger_unregisterActionHandler, _Messenger_registerInitialEventPayload, _Messenger_publish, _Messenger_subscribe, _Messenger_isInCurrentNamespace;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Messenger = exports.MOCK_ANY_NAMESPACE = void 0;
/**
 * Messenger namespace checks can be disabled by using this as the `namespace` constructor
 * parameter, and using `MockAnyNamespace` as the Namespace type parameter.
 *
 * This is useful for mocking a variety of different actions/events in unit tests. Please do not
 * use this in production code.
 */
exports.MOCK_ANY_NAMESPACE = 'MOCK_ANY_NAMESPACE';
/**
 * A message broker for "actions" and "events".
 *
 * The messenger allows registering functions as 'actions' that can be called elsewhere,
 * and it allows publishing and subscribing to events. Both actions and events are identified by
 * unique strings prefixed by a namespace (which is delimited by a colon, e.g.
 * `Namespace:actionName`).
 *
 * @template Action - A type union of all Action types.
 * @template Event - A type union of all Event types.
 * @template Namespace - The namespace for the messenger.
 */
class Messenger {
    /**
     * Construct a messenger.
     *
     * If a parent messenger is given, all actions and events under this messenger's namespace will
     * be delegated to the parent automatically.
     *
     * @param args - Constructor arguments
     * @param args.captureException - Reports an error to an error monitoring service.
     * @param args.namespace - The messenger namespace.
     * @param args.parent - The parent messenger.
     */
    constructor({ captureException, namespace, parent, }) {
        _Messenger_instances.add(this);
        _Messenger_namespace.set(this, void 0);
        /**
         * The parent messenger. All actions/events under this namespace are automatically delegated to
         * the parent messenger.
         */
        _Messenger_parent.set(this, void 0);
        _Messenger_actions.set(this, new Map());
        _Messenger_events.set(this, new Map());
        /**
         * The set of messengers we've delegated events to and their event handlers, by event type.
         */
        _Messenger_subscriptionDelegationTargets.set(this, new Map());
        /**
         * The set of messengers we've delegated actions to, by action type.
         */
        _Messenger_actionDelegationTargets.set(this, new Map());
        /**
         * A map of functions for getting the initial event payload.
         *
         * Used only for events that represent state changes.
         */
        _Messenger_initialEventPayloadGetters.set(this, new Map());
        /**
         * A cache of selector return values for their respective handlers.
         */
        _Messenger_eventPayloadCache.set(this, new Map());
        __classPrivateFieldSet(this, _Messenger_namespace, namespace, "f");
        __classPrivateFieldSet(this, _Messenger_parent, parent, "f");
        this.captureException = captureException ?? __classPrivateFieldGet(this, _Messenger_parent, "f")?.captureException;
    }
    /**
     * Register an action handler.
     *
     * This will make the registered function available to call via the `call` method.
     *
     * The action being registered must be under the same namespace as the messenger.
     *
     * @param actionType - The action type. This is a unique identifier for this action.
     * @param handler - The action handler. This function gets called when the `call` method is
     * invoked with the given action type.
     * @throws Will throw when a handler has been registered for this action type already.
     * @template ActionType - A type union of Action type strings under this messenger's namespace.
     */
    registerActionHandler(actionType, handler) {
        if (!__classPrivateFieldGet(this, _Messenger_instances, "m", _Messenger_isInCurrentNamespace).call(this, actionType)) {
            throw new Error(`Only allowed registering action handlers prefixed by '${__classPrivateFieldGet(this, _Messenger_namespace, "f")}:'`);
        }
        __classPrivateFieldGet(this, _Messenger_instances, "m", _Messenger_registerActionHandler).call(this, actionType, handler);
        if (__classPrivateFieldGet(this, _Messenger_parent, "f")) {
            // @ts-expect-error The parent type isn't constructed in a way that proves it supports this
            // action, but this is OK because it's validated in the constructor.
            this.delegate({ actions: [actionType], messenger: __classPrivateFieldGet(this, _Messenger_parent, "f") });
        }
    }
    /**
     * Registers action handlers for a list of methods on a messenger client
     *
     * @param messengerClient - The object that is expected to make use of the messenger.
     * @param methodNames - The names of the methods on the messenger client to register as action
     * handlers.
     * @template MessengerClient - The type expected to make use of the messenger.
     * @template MethodNames - The type union of method names to register as action handlers.
     */
    registerMethodActionHandlers(messengerClient, methodNames) {
        for (const methodName of methodNames) {
            const method = messengerClient[methodName];
            if (typeof method === 'function') {
                const actionType = `${messengerClient.name}:${methodName}`;
                this.registerActionHandler(actionType, method.bind(messengerClient));
            }
        }
    }
    /**
     * Unregister an action handler.
     *
     * This will prevent this action from being called.
     *
     * The action being unregistered must be under the same namespace as the messenger.
     *
     * @param actionType - The action type. This is a unique identifier for this action.
     * @template ActionType - A type union of Action type strings under this messenger's namespace.
     */
    unregisterActionHandler(actionType) {
        if (!__classPrivateFieldGet(this, _Messenger_instances, "m", _Messenger_isInCurrentNamespace).call(this, actionType)) {
            throw new Error(`Only allowed unregistering action handlers prefixed by '${__classPrivateFieldGet(this, _Messenger_namespace, "f")}:'`);
        }
        __classPrivateFieldGet(this, _Messenger_instances, "m", _Messenger_unregisterActionHandler).call(this, actionType);
    }
    /**
     * Unregister all action handlers.
     *
     * This prevents all actions from being called.
     */
    clearActions() {
        for (const actionType of __classPrivateFieldGet(this, _Messenger_actions, "f").keys()) {
            __classPrivateFieldGet(this, _Messenger_instances, "m", _Messenger_unregisterActionHandler).call(this, actionType);
        }
    }
    /**
     * Call an action.
     *
     * This function will call the action handler corresponding to the given action type, passing
     * along any parameters given.
     *
     * @param actionType - The action type. This is a unique identifier for this action.
     * @param params - The action parameters. These must match the type of the parameters of the
     * registered action handler.
     * @throws Will throw when no handler has been registered for the given type.
     * @template ActionType - A type union of Action type strings.
     * @returns The action return value.
     */
    call(actionType, ...params) {
        const handler = __classPrivateFieldGet(this, _Messenger_actions, "f").get(actionType);
        if (!handler) {
            throw new Error(`A handler for ${actionType} has not been registered`);
        }
        return handler(...params);
    }
    /**
     * Register a function for getting the initial payload for an event.
     *
     * This is used for events that represent a state change, where the payload is the state.
     * Registering a function for getting the payload allows event selectors to have a point of
     * comparison the first time state changes.
     *
     * The event type must be under the same namespace as the messenger.
     *
     * @param args - The arguments to this function
     * @param args.eventType - The event type to register a payload for.
     * @param args.getPayload - A function for retrieving the event payload.
     * @template EventType - A type union of Event type strings under this messenger's namespace.
     */
    registerInitialEventPayload({ eventType, getPayload, }) {
        if (!__classPrivateFieldGet(this, _Messenger_instances, "m", _Messenger_isInCurrentNamespace).call(this, eventType)) {
            throw new Error(`Only allowed registering initial payloads for events prefixed by '${__classPrivateFieldGet(this, _Messenger_namespace, "f")}:'`);
        }
        if (__classPrivateFieldGet(this, _Messenger_parent, "f") &&
            !__classPrivateFieldGet(this, _Messenger_subscriptionDelegationTargets, "f").get(eventType)?.has(__classPrivateFieldGet(this, _Messenger_parent, "f"))) {
            // @ts-expect-error The parent type isn't constructed in a way that proves it supports this
            // event, but this is OK because it's validated in the constructor.
            this.delegate({ events: [eventType], messenger: __classPrivateFieldGet(this, _Messenger_parent, "f") });
        }
        __classPrivateFieldGet(this, _Messenger_instances, "m", _Messenger_registerInitialEventPayload).call(this, { eventType, getPayload });
    }
    /**
     * Publish an event.
     *
     * Publishes the given payload to all subscribers of the given event type.
     *
     * Note that this method should never throw directly. Any errors from
     * subscribers are captured and re-thrown in a timeout handler.
     *
     * The event being published must be under the same namespace as the messenger.
     *
     * @param eventType - The event type. This is a unique identifier for this event.
     * @param payload - The event payload. The type of the parameters for each event handler must
     * match the type of this payload.
     * @template EventType - A type union of Event type strings under this messenger's namespace.
     */
    publish(eventType, ...payload) {
        if (!__classPrivateFieldGet(this, _Messenger_instances, "m", _Messenger_isInCurrentNamespace).call(this, eventType)) {
            throw new Error(`Only allowed publishing events prefixed by '${__classPrivateFieldGet(this, _Messenger_namespace, "f")}:'`);
        }
        if (__classPrivateFieldGet(this, _Messenger_parent, "f") &&
            !__classPrivateFieldGet(this, _Messenger_subscriptionDelegationTargets, "f").get(eventType)?.has(__classPrivateFieldGet(this, _Messenger_parent, "f"))) {
            // @ts-expect-error The parent type isn't constructed in a way that proves it supports this
            // event, but this is OK because it's validated in the constructor.
            this.delegate({ events: [eventType], messenger: __classPrivateFieldGet(this, _Messenger_parent, "f") });
        }
        __classPrivateFieldGet(this, _Messenger_instances, "m", _Messenger_publish).call(this, eventType, ...payload);
    }
    subscribe(eventType, handler, selector) {
        // Widen type of event handler by dropping ReturnType parameter.
        //
        // We need to drop it here because it's used as the parameter to the event handler, and
        // functions in general are contravariant over the parameter type. This means the type is no
        // longer valid once it's added to a broader type union with other handlers (because as far
        // as TypeScript knows, we might call the handler with output from a different selector).
        //
        // This cast means the type system is not guaranteeing the handler is called with the matching
        // input selector return value. The parameter types do ensure they match when `subscribe` is
        // called, but past that point we need to make sure of that with manual review and tests
        // instead.
        const widenedHandler = handler;
        __classPrivateFieldGet(this, _Messenger_instances, "m", _Messenger_subscribe).call(this, eventType, widenedHandler, { delegation: false, selector });
        if (selector) {
            const getPayload = __classPrivateFieldGet(this, _Messenger_initialEventPayloadGetters, "f").get(eventType);
            if (getPayload) {
                const initialValue = selector(...getPayload());
                __classPrivateFieldGet(this, _Messenger_eventPayloadCache, "f").set(widenedHandler, initialValue);
            }
        }
    }
    /**
     * Unsubscribe from an event.
     *
     * Unregisters the given function as an event handler for the given event.
     *
     * @param eventType - The event type. This is a unique identifier for this event.
     * @param handler - The event handler to unregister.
     * @throws Will throw when the given event handler is not registered for this event.
     * @template EventType - A type union of Event type strings.
     * @template SelectorReturnValue - The selector return value.
     */
    unsubscribe(eventType, handler) {
        const subscribers = __classPrivateFieldGet(this, _Messenger_events, "f").get(eventType);
        // Widen type of event handler by dropping ReturnType parameter.
        //
        // We need to drop it here because it's used as the parameter to the event handler, and
        // functions in general are contravariant over the parameter type. This means the type is no
        // longer valid once it's added to a broader type union with other handlers (because as far
        // as TypeScript knows, we might call the handler with output from a different selector).
        //
        // This poses no risk in this case, since we never call the handler past this point.
        const widenedHandler = handler;
        if (!subscribers) {
            throw new Error(`Subscription not found for event: ${eventType}`);
        }
        const metadata = subscribers.get(widenedHandler);
        if (!metadata) {
            throw new Error(`Subscription not found for event: ${eventType}`);
        }
        if (metadata.selector) {
            __classPrivateFieldGet(this, _Messenger_eventPayloadCache, "f").delete(widenedHandler);
        }
        subscribers.delete(widenedHandler);
    }
    /**
     * Clear subscriptions for a specific event.
     *
     * This will remove all subscribed handlers for this event registered from this messenger. The
     * event may still have subscribers if it has been delegated to another messenger.
     *
     * @param eventType - The event type. This is a unique identifier for this event.
     * @template EventType - A type union of Event type strings.
     */
    clearEventSubscriptions(eventType) {
        const subscriptions = __classPrivateFieldGet(this, _Messenger_events, "f").get(eventType);
        if (!subscriptions) {
            return;
        }
        for (const [handler, metadata] of subscriptions.entries()) {
            if (metadata.delegation) {
                continue;
            }
            subscriptions.delete(handler);
        }
        if (subscriptions.size === 0) {
            __classPrivateFieldGet(this, _Messenger_events, "f").delete(eventType);
        }
    }
    /**
     * Clear all subscriptions.
     *
     * This will remove all subscribed handlers for all events registered from this messenger. Events
     * may still have subscribers if they are delegated to another messenger.
     */
    clearSubscriptions() {
        for (const eventType of __classPrivateFieldGet(this, _Messenger_events, "f").keys()) {
            this.clearEventSubscriptions(eventType);
        }
    }
    /**
     * Delegate actions and/or events to another messenger.
     *
     * The messenger these actions/events are delegated to will be able to call these actions and
     * subscribe to these events.
     *
     * Note that the messenger these actions/events are delegated to must still have these
     * actions/events included in its type definition (as part of the Action and Event type
     * parameters). Actions and events are statically type checked, they cannot be delegated
     * dynamically at runtime.
     *
     * @param args - Arguments.
     * @param args.actions - The action types to delegate.
     * @param args.events - The event types to delegate.
     * @param args.messenger - The messenger to delegate to.
     * @template Delegatee - The messenger the actions/events are delegated to.
     * @template DelegatedActions - An array of delegated action types.
     * @template DelegatedEvents - An array of delegated event types.
     */
    delegate({ actions, events, messenger, }) {
        for (const actionType of actions ?? []) {
            const delegatedActionHandler = (...args) => {
                // Cast to get more specific type, for this specific action
                // The types get collapsed by `this.#actions`
                const actionHandler = __classPrivateFieldGet(this, _Messenger_actions, "f").get(actionType);
                if (!actionHandler) {
                    throw new Error(`A handler for ${actionType} has not been registered`);
                }
                return actionHandler(...args);
            };
            let delegationTargets = __classPrivateFieldGet(this, _Messenger_actionDelegationTargets, "f").get(actionType);
            if (!delegationTargets) {
                delegationTargets = new Set();
                __classPrivateFieldGet(this, _Messenger_actionDelegationTargets, "f").set(actionType, delegationTargets);
            }
            if (delegationTargets.has(messenger)) {
                throw new Error(`The action '${actionType}' has already been delegated to this messenger`);
            }
            delegationTargets.add(messenger);
            messenger._internalRegisterDelegatedActionHandler(actionType, delegatedActionHandler);
        }
        for (const eventType of events ?? []) {
            const untypedSubscriber = (...payload) => {
                messenger._internalPublishDelegated(eventType, ...payload);
            };
            // Cast to get more specific subscriber type for this specific event.
            // The types get collapsed here to the type union of all delegated
            // events, rather than the single subscriber type corresponding to this
            // event.
            const subscriber = untypedSubscriber;
            let delegatedEventSubscriptions = __classPrivateFieldGet(this, _Messenger_subscriptionDelegationTargets, "f").get(eventType);
            if (!delegatedEventSubscriptions) {
                delegatedEventSubscriptions = new Map();
                __classPrivateFieldGet(this, _Messenger_subscriptionDelegationTargets, "f").set(eventType, delegatedEventSubscriptions);
            }
            if (delegatedEventSubscriptions.has(messenger)) {
                throw new Error(`The event '${eventType}' has already been delegated to this messenger`);
            }
            delegatedEventSubscriptions.set(messenger, subscriber);
            const getPayload = __classPrivateFieldGet(this, _Messenger_initialEventPayloadGetters, "f").get(eventType);
            if (getPayload) {
                messenger._internalRegisterDelegatedInitialEventPayload({
                    eventType,
                    getPayload,
                });
            }
            __classPrivateFieldGet(this, _Messenger_instances, "m", _Messenger_subscribe).call(this, eventType, subscriber, { delegation: true });
        }
    }
    /**
     * Revoke delegated actions and/or events from another messenger.
     *
     * The messenger these actions/events are delegated to will no longer be able to call these
     * actions or subscribe to these events.
     *
     * @param args - Arguments.
     * @param args.actions - The action types to revoke.
     * @param args.events - The event types to revoke.
     * @param args.messenger - The messenger these actions/events were delegated to.
     * @template Delegatee - The messenger the actions/events are being revoked from.
     * @template DelegatedActions - An array of delegated action types.
     * @template DelegatedEvents - An array of delegated event types.
     */
    revoke({ actions, events, messenger, }) {
        if (messenger === __classPrivateFieldGet(this, _Messenger_parent, "f")) {
            throw new Error('Cannot revoke from parent');
        }
        for (const actionType of actions ?? []) {
            const delegationTargets = __classPrivateFieldGet(this, _Messenger_actionDelegationTargets, "f").get(actionType);
            if (!delegationTargets?.has(messenger)) {
                // Nothing to revoke
                continue;
            }
            messenger._internalUnregisterDelegatedActionHandler(actionType);
            delegationTargets.delete(messenger);
            if (delegationTargets.size === 0) {
                __classPrivateFieldGet(this, _Messenger_actionDelegationTargets, "f").delete(actionType);
            }
        }
        for (const eventType of events ?? []) {
            const delegationTargets = __classPrivateFieldGet(this, _Messenger_subscriptionDelegationTargets, "f").get(eventType);
            if (!delegationTargets) {
                // Nothing to revoke
                continue;
            }
            const delegatedSubscriber = delegationTargets.get(messenger);
            if (!delegatedSubscriber) {
                // Nothing to revoke
                continue;
            }
            this.unsubscribe(eventType, delegatedSubscriber);
            delegationTargets.delete(messenger);
            if (delegationTargets.size === 0) {
                __classPrivateFieldGet(this, _Messenger_subscriptionDelegationTargets, "f").delete(eventType);
            }
        }
    }
    /**
     * Register an action handler for an action delegated from another messenger.
     *
     * This will make the registered function available to call via the `call` method.
     *
     * Note: This is an internal method. Never access this property from another module. This must be
     * exposed as a public property so that these methods can be called internally on other messenger
     * instances.
     *
     * @deprecated Internal use only. Use the `delegate` method for delegation.
     * @param actionType - The action type. This is a unique identifier for this action.
     * @param handler - The action handler. This function gets called when the `call` method is
     * invoked with the given action type.
     * @throws Will throw when a handler has been registered for this action type already.
     * @template ActionType - A type union of Action type strings.
     */
    _internalRegisterDelegatedActionHandler(actionType, 
    // Using wider `ActionConstraint` type here rather than `Action` because the `Action` type is
    // contravariant over the handler parameter type. Using `Action` would lead to a type error
    // here because the messenger we've delegated to supports _additional_ actions.
    handler) {
        __classPrivateFieldGet(this, _Messenger_instances, "m", _Messenger_registerActionHandler).call(this, actionType, handler);
    }
    /**
     * Unregister an action handler for an action delegated from another messenger.
     *
     * This will prevent this action from being called.
     *
     * Note: This is an internal method. Never access this property from another module. This must be
     * exposed as a public property so that these methods can be called internally on other messenger
     * instances.
     *
     * @deprecated Internal use only. Use the `delegate` method for delegation.
     * @param actionType - The action type. This is a unqiue identifier for this action.
     * @template ActionType - A type union of Action type strings.
     */
    _internalUnregisterDelegatedActionHandler(actionType) {
        __classPrivateFieldGet(this, _Messenger_instances, "m", _Messenger_unregisterActionHandler).call(this, actionType);
    }
    /**
     * Register a function for getting the initial payload for an event that has been delegated from
     * another messenger.
     *
     * This is used for events that represent a state change, where the payload is the state.
     * Registering a function for getting the payload allows event selectors to have a point of
     * comparison the first time state changes.
     *
     * Note: This is an internal method. Never access this property from another module. This must be
     * exposed as a public property so that these methods can be called internally on other messenger
     * instances.
     *
     * @deprecated Internal use only. Use the `delegate` method for delegation.
     * @param args - The arguments to this function
     * @param args.eventType - The event type to register a payload for.
     * @param args.getPayload - A function for retrieving the event payload.
     */
    _internalRegisterDelegatedInitialEventPayload({ eventType, getPayload, }) {
        __classPrivateFieldGet(this, _Messenger_instances, "m", _Messenger_registerInitialEventPayload).call(this, { eventType, getPayload });
    }
    /**
     * Publish an event that was delegated from another messenger.
     *
     * Publishes the given payload to all subscribers of the given event type.
     *
     * Note that this method should never throw directly. Any errors from
     * subscribers are captured and re-thrown in a timeout handler.
     *
     * Note: This is an internal method. Never access this property from another module. This must be
     * exposed as a public property so that these methods can be called internally on other messenger
     * instances.
     *
     * @deprecated Internal use only. Use the `delegate` method for delegation.
     * @param eventType - The event type. This is a unique identifier for this event.
     * @param payload - The event payload. The type of the parameters for each event handler must
     * match the type of this payload.
     * @template EventType - A type union of Event type strings.
     */
    _internalPublishDelegated(eventType, ...payload) {
        __classPrivateFieldGet(this, _Messenger_instances, "m", _Messenger_publish).call(this, eventType, ...payload);
    }
}
exports.Messenger = Messenger;
_Messenger_namespace = new WeakMap(), _Messenger_parent = new WeakMap(), _Messenger_actions = new WeakMap(), _Messenger_events = new WeakMap(), _Messenger_subscriptionDelegationTargets = new WeakMap(), _Messenger_actionDelegationTargets = new WeakMap(), _Messenger_initialEventPayloadGetters = new WeakMap(), _Messenger_eventPayloadCache = new WeakMap(), _Messenger_instances = new WeakSet(), _Messenger_registerActionHandler = function _Messenger_registerActionHandler(actionType, handler) {
    if (__classPrivateFieldGet(this, _Messenger_actions, "f").has(actionType)) {
        throw new Error(`A handler for ${actionType} has already been registered`);
    }
    __classPrivateFieldGet(this, _Messenger_actions, "f").set(actionType, handler);
}, _Messenger_unregisterActionHandler = function _Messenger_unregisterActionHandler(actionType) {
    __classPrivateFieldGet(this, _Messenger_actions, "f").delete(actionType);
}, _Messenger_registerInitialEventPayload = function _Messenger_registerInitialEventPayload({ eventType, getPayload, }) {
    __classPrivateFieldGet(this, _Messenger_initialEventPayloadGetters, "f").set(eventType, getPayload);
    const delegationTargets = __classPrivateFieldGet(this, _Messenger_subscriptionDelegationTargets, "f").get(eventType);
    if (!delegationTargets) {
        return;
    }
    for (const messenger of delegationTargets.keys()) {
        messenger._internalRegisterDelegatedInitialEventPayload({
            eventType,
            getPayload,
        });
    }
}, _Messenger_publish = function _Messenger_publish(eventType, ...payload) {
    const subscribers = __classPrivateFieldGet(this, _Messenger_events, "f").get(eventType);
    if (subscribers) {
        for (const [handler, { selector }] of subscribers.entries()) {
            try {
                if (selector) {
                    const previousValue = __classPrivateFieldGet(this, _Messenger_eventPayloadCache, "f").get(handler);
                    const newValue = selector(...payload);
                    if (newValue !== previousValue) {
                        __classPrivateFieldGet(this, _Messenger_eventPayloadCache, "f").set(handler, newValue);
                        handler(newValue, previousValue);
                    }
                }
                else {
                    handler(...payload);
                }
            }
            catch (error) {
                // Capture error without interrupting the event publishing.
                if (this.captureException) {
                    this.captureException(error instanceof Error ? error : new Error(String(error)));
                }
                else {
                    console.error(error);
                }
            }
        }
    }
}, _Messenger_subscribe = function _Messenger_subscribe(eventType, handler, metadata) {
    let subscribers = __classPrivateFieldGet(this, _Messenger_events, "f").get(eventType);
    if (!subscribers) {
        subscribers = new Map();
        __classPrivateFieldGet(this, _Messenger_events, "f").set(eventType, subscribers);
    }
    subscribers.set(handler, metadata);
}, _Messenger_isInCurrentNamespace = function _Messenger_isInCurrentNamespace(name) {
    return (__classPrivateFieldGet(this, _Messenger_namespace, "f") === exports.MOCK_ANY_NAMESPACE ||
        name.startsWith(`${__classPrivateFieldGet(this, _Messenger_namespace, "f")}:`));
};
