// @metamask/permission-controller/dist/PermissionController.mjs
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var _PermissionController_instances, _PermissionController_caveatSpecifications, _PermissionController_permissionSpecifications, _PermissionController_unrestrictedMethods, _PermissionController_getPermissionSpecification, _PermissionController_getCaveatSpecification, _PermissionController_expectGetCaveatMerger, _PermissionController_validatePermissionSpecifications, _PermissionController_getTypedPermissionSpecification, _PermissionController_deletePermission, _PermissionController_setCaveat, _PermissionController_deleteCaveat, _PermissionController_validateModifiedPermission, _PermissionController_targetExists, _PermissionController_applyGrantedPermissions, _PermissionController_validatePermission, _PermissionController_setValidatedPermissions, _PermissionController_constructCaveats, _PermissionController_validateCaveat, _PermissionController_validateRequestedPermissions, _PermissionController_mergeIncrementalPermissions, _PermissionController_mergePermission, _PermissionController_mergeCaveat, _PermissionController_requestUserApproval, _PermissionController_handleApprovedPermissions, _PermissionController_getSideEffects, _PermissionController_executeSideEffects, _PermissionController_validateApprovedPermissions, _PermissionController_hasApprovalRequest, _PermissionController_rejectPermissionsRequest, _PermissionController_executeRestrictedMethod;
function $importDefault(module) {
    if (module?.__esModule) {
        return module.default;
    }
    return module;
}
import { BaseController } from "@metamask/base-controller";
import { isNonEmptyArray, isPlainObject, isValidJson } from "@metamask/controller-utils";
import { JsonRpcError } from "@metamask/rpc-errors";
import { hasProperty } from "@metamask/utils";
import $deepFreeze from "deep-freeze-strict";
const deepFreeze = $importDefault($deepFreeze);
import { castDraft, produce as immerProduce } from "immer";
import { nanoid } from "nanoid";
import { decorateWithCaveats, isRestrictedMethodCaveatSpecification } from "./Caveat.mjs";
import { CaveatAlreadyExistsError, CaveatDoesNotExistError, CaveatInvalidJsonError, CaveatMergerDoesNotExistError, CaveatMergeTypeMismatchError, CaveatMissingValueError, CaveatSpecificationMismatchError, DuplicateCaveatError, EndowmentPermissionDoesNotExistError, ForbiddenCaveatError, internalError, InvalidApprovedPermissionError, InvalidCaveatError, InvalidCaveatFieldsError, InvalidCaveatsPropertyError, InvalidCaveatTypeError, InvalidMergedPermissionsError, invalidParams, InvalidSubjectIdentifierError, methodNotFound, PermissionDoesNotExistError, PermissionsRequestNotFoundError, unauthorized, UnrecognizedCaveatTypeError, UnrecognizedSubjectError, userRejectedRequest } from "./errors.mjs";
import { constructPermission, findCaveat, hasSpecificationType, PermissionType } from "./Permission.mjs";
import { getPermissionMiddlewareFactory } from "./permission-middleware.mjs";
import { collectUniqueAndPairedCaveats, MethodNames } from "./utils.mjs";
/**
 * The name of the {@link PermissionController}.
 */
const controllerName = 'PermissionController';
const MESSENGER_EXPOSED_METHODS = [
    'clearState',
    'getEndowments',
    'getSubjectNames',
    'getPermissions',
    'hasPermission',
    'hasPermissions',
    'grantPermissions',
    'grantPermissionsIncremental',
    'requestPermissions',
    'requestPermissionsIncremental',
    'revokeAllPermissions',
    'revokePermissionForAllSubjects',
    'revokePermissions',
    'updateCaveat',
    'getCaveat',
];
/**
 * Get the state metadata of the {@link PermissionController}.
 *
 * @template Permission - The controller's permission type union.
 * @returns The state metadata
 */
function getStateMetadata() {
    return {
        subjects: {
            includeInStateLogs: true,
            includeInDebugSnapshot: true,
            persist: true,
            usedInUi: true,
        },
    };
}
/**
 * Get the default state of the {@link PermissionController}.
 *
 * @template Permission - The controller's permission type union.
 * @returns The default state of the controller
 */
function getDefaultState() {
    return { subjects: {} };
}
/**
 * Describes the possible results of a {@link CaveatMutator} function.
 */
export var CaveatMutatorOperation;
(function (CaveatMutatorOperation) {
    CaveatMutatorOperation[CaveatMutatorOperation["Noop"] = 0] = "Noop";
    CaveatMutatorOperation[CaveatMutatorOperation["UpdateValue"] = 1] = "UpdateValue";
    CaveatMutatorOperation[CaveatMutatorOperation["DeleteCaveat"] = 2] = "DeleteCaveat";
    CaveatMutatorOperation[CaveatMutatorOperation["RevokePermission"] = 3] = "RevokePermission";
})(CaveatMutatorOperation || (CaveatMutatorOperation = {}));
/**
 * The permission controller. See the [Architecture](../ARCHITECTURE.md)
 * document for details.
 *
 * Assumes the existence of an {@link ApprovalController} reachable via the
 * {@link Messenger}.
 *
 * @template ControllerPermissionSpecification - A union of the types of all
 * permission specifications available to the controller. Any referenced caveats
 * must be included in the controller's caveat specifications.
 * @template ControllerCaveatSpecification - A union of the types of all
 * caveat specifications available to the controller.
 */
export class PermissionController extends BaseController {
    /**
     * The names of all JSON-RPC methods that will be ignored by the controller.
     *
     * @returns The names of all unrestricted JSON-RPC methods
     */
    get unrestrictedMethods() {
        return __classPrivateFieldGet(this, _PermissionController_unrestrictedMethods, "f");
    }
    /**
     * Constructs the PermissionController.
     *
     * @param options - Permission controller options.
     * @param options.caveatSpecifications - The specifications of all caveats
     * available to the controller. See {@link CaveatSpecificationMap} and the
     * documentation for more details.
     * @param options.permissionSpecifications - The specifications of all
     * permissions available to the controller. See
     * {@link PermissionSpecificationMap} and the README for more details.
     * @param options.unrestrictedMethods - The callable names of all JSON-RPC
     * methods ignored by the new controller.
     * @param options.messenger - The messenger. See
     * {@link BaseController} for more information.
     * @param options.state - Existing state to hydrate the controller with at
     * initialization.
     */
    constructor(options) {
        const { caveatSpecifications, permissionSpecifications, unrestrictedMethods, messenger, state = {}, } = options;
        super({
            name: controllerName,
            metadata: getStateMetadata(),
            messenger,
            state: {
                ...getDefaultState(),
                ...state,
            },
        });
        _PermissionController_instances.add(this);
        _PermissionController_caveatSpecifications.set(this, void 0);
        _PermissionController_permissionSpecifications.set(this, void 0);
        _PermissionController_unrestrictedMethods.set(this, void 0);
        __classPrivateFieldSet(this, _PermissionController_unrestrictedMethods, new Set(unrestrictedMethods), "f");
        __classPrivateFieldSet(this, _PermissionController_caveatSpecifications, deepFreeze({ ...caveatSpecifications }), "f");
        __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_validatePermissionSpecifications).call(this, permissionSpecifications, __classPrivateFieldGet(this, _PermissionController_caveatSpecifications, "f"));
        __classPrivateFieldSet(this, _PermissionController_permissionSpecifications, deepFreeze({
            ...permissionSpecifications,
        }), "f");
        this.messenger.registerMethodActionHandlers(this, MESSENGER_EXPOSED_METHODS);
        this.createPermissionMiddleware = getPermissionMiddlewareFactory({
            executeRestrictedMethod: __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_executeRestrictedMethod).bind(this),
            getRestrictedMethod: this.getRestrictedMethod.bind(this),
            isUnrestrictedMethod: this.unrestrictedMethods.has.bind(this.unrestrictedMethods),
        });
    }
    /**
     * Clears the state of the controller.
     */
    clearState() {
        this.update((_draftState) => {
            return {
                ...getDefaultState(),
            };
        });
    }
    /**
     * Gets the implementation of the specified restricted method.
     *
     * A JSON-RPC error is thrown if the method does not exist.
     *
     * @see {@link PermissionController.executeRestrictedMethod} and
     * {@link PermissionController.createPermissionMiddleware} for internal usage.
     * @param method - The name of the restricted method.
     * @param origin - The origin associated with the request for the restricted
     * method, if any.
     * @returns The restricted method implementation.
     */
    getRestrictedMethod(method, origin) {
        return __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_getTypedPermissionSpecification).call(this, PermissionType.RestrictedMethod, method, origin).methodImplementation;
    }
    /**
     * Gets a list of all origins of subjects.
     *
     * @returns The origins (i.e. IDs) of all subjects.
     */
    getSubjectNames() {
        return Object.keys(this.state.subjects);
    }
    /**
     * Gets the permission for the specified target of the subject corresponding
     * to the specified origin.
     *
     * @param origin - The origin of the subject.
     * @param targetName - The method name as invoked by a third party (i.e., not
     * a method key).
     * @returns The permission if it exists, or undefined otherwise.
     */
    getPermission(origin, targetName) {
        return this.state.subjects[origin]?.permissions[targetName];
    }
    /**
     * Gets all permissions for the specified subject, if any.
     *
     * @param origin - The origin of the subject.
     * @returns The permissions of the subject, if any.
     */
    getPermissions(origin) {
        return this.state.subjects[origin]?.permissions;
    }
    /**
     * Checks whether the subject with the specified origin has the specified
     * permission.
     *
     * @param origin - The origin of the subject.
     * @param target - The target name of the permission.
     * @returns Whether the subject has the permission.
     */
    hasPermission(origin, target) {
        return Boolean(this.getPermission(origin, target));
    }
    /**
     * Checks whether the subject with the specified origin has any permissions.
     * Use this if you want to know if a subject "exists".
     *
     * @param origin - The origin of the subject to check.
     * @returns Whether the subject has any permissions.
     */
    hasPermissions(origin) {
        return Boolean(this.state.subjects[origin]);
    }
    /**
     * Revokes all permissions from the specified origin.
     *
     * Throws an error of the origin has no permissions.
     *
     * @param origin - The origin whose permissions to revoke.
     */
    revokeAllPermissions(origin) {
        this.update((draftState) => {
            if (!draftState.subjects[origin]) {
                throw new UnrecognizedSubjectError(origin);
            }
            delete draftState.subjects[origin];
        });
    }
    /**
     * Revokes the specified permission from the subject with the specified
     * origin.
     *
     * Throws an error if the subject or the permission does not exist.
     *
     * @param origin - The origin of the subject whose permission to revoke.
     * @param target - The target name of the permission to revoke.
     */
    revokePermission(origin, target) {
        this.revokePermissions({ [origin]: [target] });
    }
    /**
     * Revokes the specified permissions from the specified subjects.
     *
     * Throws an error if any of the subjects or permissions do not exist.
     *
     * @param subjectsAndPermissions - An object mapping subject origins
     * to arrays of permission target names to revoke.
     */
    revokePermissions(subjectsAndPermissions) {
        this.update((draftState) => {
            Object.keys(subjectsAndPermissions).forEach((origin) => {
                if (!hasProperty(draftState.subjects, origin)) {
                    throw new UnrecognizedSubjectError(origin);
                }
                subjectsAndPermissions[origin].forEach((target) => {
                    const { permissions } = draftState.subjects[origin];
                    if (!hasProperty(permissions, target)) {
                        throw new PermissionDoesNotExistError(origin, target);
                    }
                    __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_deletePermission).call(this, draftState.subjects, origin, target);
                });
            });
        });
    }
    /**
     * Revokes all permissions corresponding to the specified target for all subjects.
     * Does nothing if no subjects or no such permission exists.
     *
     * @param target - The name of the target to revoke all permissions for.
     */
    revokePermissionForAllSubjects(target) {
        if (this.getSubjectNames().length === 0) {
            return;
        }
        this.update((draftState) => {
            Object.entries(draftState.subjects).forEach(([origin, subject]) => {
                const { permissions } = subject;
                if (hasProperty(permissions, target)) {
                    __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_deletePermission).call(this, draftState.subjects, origin, target);
                }
            });
        });
    }
    /**
     * Checks whether the permission of the subject corresponding to the given
     * origin has a caveat of the specified type.
     *
     * Throws an error if the subject does not have a permission with the
     * specified target name.
     *
     * @template TargetName - The permission target name. Should be inferred.
     * @template CaveatType - The valid caveat types for the permission. Should
     * be inferred.
     * @param origin - The origin of the subject.
     * @param target - The target name of the permission.
     * @param caveatType - The type of the caveat to check for.
     * @returns Whether the permission has the specified caveat.
     */
    hasCaveat(origin, target, caveatType) {
        return Boolean(this.getCaveat(origin, target, caveatType));
    }
    /**
     * Gets the caveat of the specified type, if any, for the permission of
     * the subject corresponding to the given origin.
     *
     * Throws an error if the subject does not have a permission with the
     * specified target name.
     *
     * @template TargetName - The permission target name. Should be inferred.
     * @template CaveatType - The valid caveat types for the permission. Should
     * be inferred.
     * @param origin - The origin of the subject.
     * @param target - The target name of the permission.
     * @param caveatType - The type of the caveat to get.
     * @returns The caveat, or `undefined` if no such caveat exists.
     */
    getCaveat(origin, target, caveatType) {
        const permission = this.getPermission(origin, target);
        if (!permission) {
            throw new PermissionDoesNotExistError(origin, target);
        }
        return findCaveat(permission, caveatType);
    }
    /**
     * Adds a caveat of the specified type, with the specified caveat value, to
     * the permission corresponding to the given subject origin and permission
     * target.
     *
     * For modifying existing caveats, use
     * {@link PermissionController.updateCaveat}.
     *
     * Throws an error if no such permission exists, or if the caveat already
     * exists.
     *
     * @template TargetName - The permission target name. Should be inferred.
     * @template CaveatType - The valid caveat types for the permission. Should
     * be inferred.
     * @param origin - The origin of the subject.
     * @param target - The target name of the permission.
     * @param caveatType - The type of the caveat to add.
     * @param caveatValue - The value of the caveat to add.
     */
    addCaveat(origin, target, caveatType, caveatValue) {
        if (this.hasCaveat(origin, target, caveatType)) {
            throw new CaveatAlreadyExistsError(origin, target, caveatType);
        }
        __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_setCaveat).call(this, origin, target, caveatType, caveatValue);
    }
    /**
     * Updates the value of the caveat of the specified type belonging to the
     * permission corresponding to the given subject origin and permission
     * target.
     *
     * For adding new caveats, use
     * {@link PermissionController.addCaveat}.
     *
     * Throws an error if no such permission or caveat exists.
     *
     * @template TargetName - The permission target name. Should be inferred.
     * @template CaveatType - The valid caveat types for the permission. Should
     * be inferred.
     * @param origin - The origin of the subject.
     * @param target - The target name of the permission.
     * @param caveatType - The type of the caveat to update.
     * @param caveatValue - The new value of the caveat.
     */
    updateCaveat(origin, target, caveatType, caveatValue) {
        if (!this.hasCaveat(origin, target, caveatType)) {
            throw new CaveatDoesNotExistError(origin, target, caveatType);
        }
        __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_setCaveat).call(this, origin, target, caveatType, caveatValue);
    }
    /**
     * Updates all caveats with the specified type for all subjects and
     * permissions by applying the specified mutator function to them.
     *
     * ATTN: Permissions can be revoked entirely by the action of this method,
     * read on for details.
     *
     * Caveat mutators are functions that receive a caveat value and return a
     * tuple consisting of a {@link CaveatMutatorOperation} and, optionally, a new
     * value to update the existing caveat with.
     *
     * For each caveat, depending on the mutator result, this method will:
     * - Do nothing ({@link CaveatMutatorOperation.Noop})
     * - Update the value of the caveat ({@link CaveatMutatorOperation.UpdateValue}). The caveat specification validator, if any, will be called after updating the value.
     * - Delete the caveat ({@link CaveatMutatorOperation.DeleteCaveat}). The permission specification validator, if any, will be called after deleting the caveat.
     * - Revoke the parent permission ({@link CaveatMutatorOperation.RevokePermission})
     *
     * This method throws if the validation of any caveat or permission fails.
     *
     * @param targetCaveatType - The type of the caveats to update.
     * @param mutator - The mutator function which will be applied to all caveat
     * values.
     */
    updatePermissionsByCaveat(targetCaveatType, mutator) {
        if (Object.keys(this.state.subjects).length === 0) {
            return;
        }
        this.update((draftState) => {
            Object.values(draftState.subjects).forEach((subject) => {
                Object.values(subject.permissions).forEach((permission) => {
                    const { caveats } = permission;
                    const targetCaveat = caveats?.find(({ type }) => type === targetCaveatType);
                    if (!targetCaveat) {
                        return;
                    }
                    // The mutator may modify the caveat value in place, and must always
                    // return a valid mutation result.
                    const mutatorResult = mutator(targetCaveat.value);
                    const { operation } = mutatorResult;
                    switch (operation) {
                        case CaveatMutatorOperation.Noop:
                            break;
                        case CaveatMutatorOperation.UpdateValue:
                            // Typecast: `Mutable` is used here to assign to a readonly
                            // property. `targetConstraint` should already be mutable because
                            // it's part of a draft, but for some reason it's not. We can't
                            // use the more-correct `Draft` type here either because it
                            // results in an error.
                            targetCaveat.value =
                                mutatorResult.value;
                            __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_validateCaveat).call(this, targetCaveat, subject.origin, permission.parentCapability);
                            break;
                        case CaveatMutatorOperation.DeleteCaveat:
                            __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_deleteCaveat).call(this, permission, targetCaveatType, subject.origin);
                            break;
                        case CaveatMutatorOperation.RevokePermission:
                            __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_deletePermission).call(this, draftState.subjects, subject.origin, permission.parentCapability);
                            break;
                        default: {
                            // Overriding as `never` is the expected result of exhaustiveness checking,
                            // and is intended to represent unchecked exception cases.
                            // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
                            throw new Error(`Unrecognized mutation result: "${operation}"`);
                        }
                    }
                });
            });
        });
    }
    /**
     * Removes the caveat of the specified type from the permission corresponding
     * to the given subject origin and target name.
     *
     * Throws an error if no such permission or caveat exists.
     *
     * @template TargetName - The permission target name. Should be inferred.
     * @template CaveatType - The valid caveat types for the permission. Should
     * be inferred.
     * @param origin - The origin of the subject.
     * @param target - The target name of the permission.
     * @param caveatType - The type of the caveat to remove.
     */
    removeCaveat(origin, target, caveatType) {
        this.update((draftState) => {
            const permission = draftState.subjects[origin]?.permissions[target];
            if (!permission) {
                throw new PermissionDoesNotExistError(origin, target);
            }
            if (!permission.caveats) {
                throw new CaveatDoesNotExistError(origin, target, caveatType);
            }
            __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_deleteCaveat).call(this, permission, caveatType, origin);
        });
    }
    /**
     * Grants _approved_ permissions to the specified subject. Every permission and
     * caveat is stringently validated—including by calling their specification
     * validators—and an error is thrown if validation fails.
     *
     * ATTN: This method does **not** prompt the user for approval. User consent must
     * first be obtained through some other means.
     *
     * @see {@link PermissionController.requestPermissions} For initiating a
     * permissions request requiring user approval.
     * @param options - Options bag.
     * @param options.approvedPermissions - The requested permissions approved by
     * the user.
     * @param options.requestData - Permission request data. Passed to permission
     * factory functions.
     * @param options.preserveExistingPermissions - Whether to preserve the
     * subject's existing permissions.
     * @param options.subject - The subject to grant permissions to.
     * @returns The subject's new permission state. It may or may not have changed.
     */
    grantPermissions({ approvedPermissions, requestData, preserveExistingPermissions = true, subject, }) {
        return __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_applyGrantedPermissions).call(this, {
            approvedPermissions,
            subject,
            mergePermissions: false,
            preserveExistingPermissions,
            requestData,
        });
    }
    /**
     * Incrementally grants _approved_ permissions to the specified subject. Every
     * permission and caveat is stringently validated—including by calling their
     * specification validators—and an error is thrown if validation fails.
     *
     * ATTN: This method does **not** prompt the user for approval. User consent must
     * first be obtained through some other means.
     *
     * @see {@link PermissionController.requestPermissionsIncremental} For initiating
     * an incremental permissions request requiring user approval.
     * @param options - Options bag.
     * @param options.approvedPermissions - The requested permissions approved by
     * the user.
     * @param options.requestData - Permission request data. Passed to permission
     * factory functions.
     * @param options.subject - The subject to grant permissions to.
     * @returns The subject's new permission state. It may or may not have changed.
     */
    grantPermissionsIncremental({ approvedPermissions, requestData, subject, }) {
        return __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_applyGrantedPermissions).call(this, {
            approvedPermissions,
            subject,
            mergePermissions: true,
            preserveExistingPermissions: true,
            requestData,
        });
    }
    /**
     * Initiates a permission request that requires user approval.
     *
     * Either this or {@link PermissionController.requestPermissionsIncremental}
     * should always be used to grant additional permissions to a subject,
     * unless user approval has been obtained through some other means.
     *
     * Permissions are validated at every step of the approval process, and this
     * method will reject if validation fails.
     *
     * @see {@link ApprovalController} For the user approval logic.
     * @see {@link PermissionController.acceptPermissionsRequest} For the method
     * that _accepts_ the request and resolves the user approval promise.
     * @see {@link PermissionController.rejectPermissionsRequest} For the method
     * that _rejects_ the request and the user approval promise.
     * @param subject - The grantee subject.
     * @param requestedPermissions - The requested permissions.
     * @param options - Additional options.
     * @param options.id - The id of the permissions request. Defaults to a unique
     * id.
     * @param options.preserveExistingPermissions - Whether to preserve the
     * subject's existing permissions. Defaults to `true`.
     * @param options.metadata - Additional metadata about the permission request.
     * @returns The granted permissions and request metadata.
     */
    async requestPermissions(subject, requestedPermissions, options = {}) {
        const { origin } = subject;
        const { id = nanoid(), preserveExistingPermissions = true } = options;
        __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_validateRequestedPermissions).call(this, origin, requestedPermissions);
        const metadata = {
            ...options.metadata,
            id,
            origin,
        };
        const permissionsRequest = {
            metadata,
            permissions: requestedPermissions,
        };
        const approvedRequest = await __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_requestUserApproval).call(this, permissionsRequest);
        return await __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_handleApprovedPermissions).call(this, {
            subject,
            metadata,
            preserveExistingPermissions,
            approvedRequest,
        });
    }
    /**
     * Initiates an incremental permission request that prompts for user approval.
     * Incremental permission requests allow the caller to replace existing and/or
     * add brand new permissions and caveats for the specified subject.
     *
     * Incremental permission request are merged with the subject's existing permissions
     * through a right-biased union, where the incremental permission are the right-hand
     * side of the merger. If both sides of the merger specify the same caveats for a
     * given permission, the caveats are merged using their specification's caveat value
     * merger property.
     *
     * Either this or {@link PermissionController.requestPermissions} should
     * always be used to grant additional permissions to a subject, unless user
     * approval has been obtained through some other means.
     *
     * Permissions are validated at every step of the approval process, and this
     * method will reject if validation fails.
     *
     * @see {@link ApprovalController} For the user approval logic.
     * @see {@link PermissionController.acceptPermissionsRequest} For the method
     * that _accepts_ the request and resolves the user approval promise.
     * @see {@link PermissionController.rejectPermissionsRequest} For the method
     * that _rejects_ the request and the user approval promise.
     * @param subject - The grantee subject.
     * @param requestedPermissions - The requested permissions.
     * @param options - Additional options.
     * @param options.id - The id of the permissions request. Defaults to a unique
     * id.
     * @param options.metadata - Additional metadata about the permission request.
     * @returns The granted permissions and request metadata.
     */
    async requestPermissionsIncremental(subject, requestedPermissions, options = {}) {
        const { origin } = subject;
        const { id = nanoid() } = options;
        __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_validateRequestedPermissions).call(this, origin, requestedPermissions);
        const currentPermissions = this.getPermissions(origin) ?? {};
        const [newPermissions, permissionDiffMap] = __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_mergeIncrementalPermissions).call(this, currentPermissions, requestedPermissions);
        // The second undefined check is just for type narrowing purposes. These values
        // will always be jointly defined or undefined.
        if (newPermissions === undefined || permissionDiffMap === undefined) {
            return [];
        }
        try {
            // It does not spark joy to run this validation again after the merger operation.
            // But, optimizing this procedure is probably not worth it, especially considering
            // that the worst-case scenario for validation degrades to the below function call.
            __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_validateRequestedPermissions).call(this, origin, newPermissions);
        }
        catch (error) {
            if (error instanceof Error) {
                throw new InvalidMergedPermissionsError(origin, error, permissionDiffMap);
            }
            /* istanbul ignore next: This should be impossible */
            throw internalError('Unrecognized error type', { error });
        }
        const metadata = {
            ...options.metadata,
            id,
            origin,
        };
        const permissionsRequest = {
            metadata,
            permissions: newPermissions,
            diff: {
                currentPermissions,
                permissionDiffMap,
            },
        };
        const approvedRequest = await __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_requestUserApproval).call(this, permissionsRequest);
        return await __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_handleApprovedPermissions).call(this, {
            subject,
            metadata,
            preserveExistingPermissions: false,
            approvedRequest,
        });
    }
    /**
     * Accepts a permissions request created by
     * {@link PermissionController.requestPermissions}.
     *
     * @param request - The permissions request.
     */
    async acceptPermissionsRequest(request) {
        const { id } = request.metadata;
        if (!__classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_hasApprovalRequest).call(this, { id })) {
            throw new PermissionsRequestNotFoundError(id);
        }
        if (Object.keys(request.permissions).length === 0) {
            __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_rejectPermissionsRequest).call(this, id, invalidParams({
                message: 'Must request at least one permission.',
            }));
            return;
        }
        try {
            await this.messenger.call('ApprovalController:acceptRequest', id, request);
        }
        catch (error) {
            // If accepting unexpectedly fails, reject the request and re-throw the
            // error
            __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_rejectPermissionsRequest).call(this, id, error);
            throw error;
        }
    }
    /**
     * Rejects a permissions request created by
     * {@link PermissionController.requestPermissions}.
     *
     * @param id - The id of the request to be rejected.
     */
    async rejectPermissionsRequest(id) {
        if (!__classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_hasApprovalRequest).call(this, { id })) {
            throw new PermissionsRequestNotFoundError(id);
        }
        __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_rejectPermissionsRequest).call(this, id, userRejectedRequest());
    }
    /**
     * Gets the subject's endowments per the specified endowment permission.
     * Throws if the subject does not have the required permission or if the
     * permission is not an endowment permission.
     *
     * @param origin - The origin of the subject whose endowments to retrieve.
     * @param targetName - The name of the endowment permission. This must be a
     * valid permission target name.
     * @param requestData - Additional data associated with the request, if any.
     * Forwarded to the endowment getter function for the permission.
     * @returns The endowments, if any.
     */
    async getEndowments(origin, targetName, requestData) {
        if (!this.hasPermission(origin, targetName)) {
            throw unauthorized({ data: { origin, targetName } });
        }
        return __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_getTypedPermissionSpecification).call(this, PermissionType.Endowment, targetName, origin).endowmentGetter({ origin, requestData });
    }
    /**
     * Executes a restricted method as the subject with the given origin.
     * The specified params, if any, will be passed to the method implementation.
     *
     * ATTN: Great caution should be exercised in the use of this method.
     * Methods that cause side effects or affect application state should
     * be avoided.
     *
     * This method will first attempt to retrieve the requested restricted method
     * implementation, throwing if it does not exist. The method will then be
     * invoked as though the subject with the specified origin had invoked it with
     * the specified parameters. This means that any existing caveats will be
     * applied to the restricted method, and this method will throw if the
     * restricted method or its caveat decorators throw.
     *
     * In addition, this method will throw if the subject does not have a
     * permission for the specified restricted method.
     *
     * @param origin - The origin of the subject to execute the method on behalf
     * of.
     * @param targetName - The name of the method to execute. This must be a valid
     * permission target name.
     * @param params - The parameters to pass to the method implementation.
     * @returns The result of the executed method.
     */
    async executeRestrictedMethod(origin, targetName, params) {
        // Throws if the method does not exist
        const methodImplementation = this.getRestrictedMethod(targetName, origin);
        const result = await __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_executeRestrictedMethod).call(this, methodImplementation, { origin }, targetName, params);
        if (result === undefined) {
            throw new Error(`Internal request for method "${targetName}" as origin "${origin}" returned no result.`);
        }
        return result;
    }
}
_PermissionController_caveatSpecifications = new WeakMap(), _PermissionController_permissionSpecifications = new WeakMap(), _PermissionController_unrestrictedMethods = new WeakMap(), _PermissionController_instances = new WeakSet(), _PermissionController_getPermissionSpecification = function _PermissionController_getPermissionSpecification(targetName) {
    return __classPrivateFieldGet(this, _PermissionController_permissionSpecifications, "f")[targetName];
}, _PermissionController_getCaveatSpecification = function _PermissionController_getCaveatSpecification(caveatType) {
    return __classPrivateFieldGet(this, _PermissionController_caveatSpecifications, "f")[caveatType];
}, _PermissionController_expectGetCaveatMerger = function _PermissionController_expectGetCaveatMerger(caveatType) {
    const { merger } = __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_getCaveatSpecification).call(this, caveatType);
    if (merger === undefined) {
        throw new CaveatMergerDoesNotExistError(caveatType);
    }
    return merger;
}, _PermissionController_validatePermissionSpecifications = function _PermissionController_validatePermissionSpecifications(permissionSpecifications, caveatSpecifications) {
    Object.entries(permissionSpecifications).forEach(([targetName, { permissionType, targetName: innerTargetName, allowedCaveats },]) => {
        if (!permissionType || !hasProperty(PermissionType, permissionType)) {
            throw new Error(`Invalid permission type: "${permissionType}"`);
        }
        if (!targetName) {
            throw new Error(`Invalid permission target name: "${targetName}"`);
        }
        if (targetName !== innerTargetName) {
            throw new Error(`Invalid permission specification: target name "${targetName}" must match specification.targetName value "${innerTargetName}".`);
        }
        if (allowedCaveats) {
            allowedCaveats.forEach((caveatType) => {
                if (!hasProperty(caveatSpecifications, caveatType)) {
                    throw new UnrecognizedCaveatTypeError(caveatType);
                }
                const specification = caveatSpecifications[caveatType];
                const isRestrictedMethodCaveat = isRestrictedMethodCaveatSpecification(specification);
                if ((permissionType === PermissionType.RestrictedMethod &&
                    !isRestrictedMethodCaveat) ||
                    (permissionType === PermissionType.Endowment &&
                        isRestrictedMethodCaveat)) {
                    throw new CaveatSpecificationMismatchError(specification, permissionType);
                }
            });
        }
    });
}, _PermissionController_getTypedPermissionSpecification = function _PermissionController_getTypedPermissionSpecification(permissionType, targetName, requestingOrigin) {
    const failureError = permissionType === PermissionType.RestrictedMethod
        ? methodNotFound(targetName, requestingOrigin ? { origin: requestingOrigin } : undefined)
        : new EndowmentPermissionDoesNotExistError(targetName, requestingOrigin);
    if (!__classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_targetExists).call(this, targetName)) {
        throw failureError;
    }
    const specification = __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_getPermissionSpecification).call(this, targetName);
    if (!hasSpecificationType(specification, permissionType)) {
        throw failureError;
    }
    return specification;
}, _PermissionController_deletePermission = function _PermissionController_deletePermission(subjects, origin, target) {
    const { permissions } = subjects[origin];
    if (Object.keys(permissions).length > 1) {
        delete permissions[target];
    }
    else {
        delete subjects[origin];
    }
}, _PermissionController_setCaveat = function _PermissionController_setCaveat(origin, target, caveatType, caveatValue) {
    this.update((draftState) => {
        const subject = draftState.subjects[origin];
        // Unreachable because `hasCaveat` is always called before this, and it
        // throws if permissions are missing. TypeScript needs this, however.
        /* istanbul ignore if */
        if (!subject) {
            throw new UnrecognizedSubjectError(origin);
        }
        const permission = subject.permissions[target];
        /* istanbul ignore if: practically impossible, but TypeScript wants it */
        if (!permission) {
            throw new PermissionDoesNotExistError(origin, target);
        }
        const caveat = {
            type: caveatType,
            value: caveatValue,
        };
        __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_validateCaveat).call(this, caveat, origin, target);
        let addedCaveat = false;
        if (permission.caveats) {
            const caveatIndex = permission.caveats.findIndex((existingCaveat) => existingCaveat.type === caveat.type);
            if (caveatIndex === -1) {
                permission.caveats.push(caveat);
                addedCaveat = true;
            }
            else {
                permission.caveats.splice(caveatIndex, 1, caveat);
            }
        }
        else {
            // At this point, we don't know if the specific permission is allowed
            // to have caveats, but it should be impossible to call this method
            // for a permission that may not have any caveats. If all else fails,
            // the permission validator is also called.
            // @ts-expect-error See above comment
            permission.caveats = [caveat];
            addedCaveat = true;
        }
        // Mutating a caveat does not warrant permission validation, but mutating
        // the caveat array does.
        if (addedCaveat) {
            __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_validateModifiedPermission).call(this, permission, origin, {
                invokePermissionValidator: true,
                performCaveatValidation: false, // We just validated the caveat
            });
        }
    });
}, _PermissionController_deleteCaveat = function _PermissionController_deleteCaveat(permission, caveatType, origin) {
    /* istanbul ignore if: not possible in our usage */
    if (!permission.caveats) {
        throw new CaveatDoesNotExistError(origin, permission.parentCapability, caveatType);
    }
    const caveatIndex = permission.caveats.findIndex((existingCaveat) => existingCaveat.type === caveatType);
    if (caveatIndex === -1) {
        throw new CaveatDoesNotExistError(origin, permission.parentCapability, caveatType);
    }
    if (permission.caveats.length === 1) {
        permission.caveats = null;
    }
    else {
        permission.caveats.splice(caveatIndex, 1);
    }
    __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_validateModifiedPermission).call(this, permission, origin, {
        invokePermissionValidator: true,
        performCaveatValidation: false, // No caveat object was mutated
    });
}, _PermissionController_validateModifiedPermission = function _PermissionController_validateModifiedPermission(permission, origin, validationFlags) {
    /* istanbul ignore if: this should be impossible */
    if (!__classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_targetExists).call(this, permission.parentCapability)) {
        throw new Error(`Fatal: Existing permission target "${permission.parentCapability}" has no specification.`);
    }
    __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_validatePermission).call(this, __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_getPermissionSpecification).call(this, permission.parentCapability), permission, origin, validationFlags);
}, _PermissionController_targetExists = function _PermissionController_targetExists(target) {
    return hasProperty(__classPrivateFieldGet(this, _PermissionController_permissionSpecifications, "f"), target);
}, _PermissionController_applyGrantedPermissions = function _PermissionController_applyGrantedPermissions({ approvedPermissions, subject, mergePermissions, preserveExistingPermissions, requestData, }) {
    const { origin } = subject;
    if (!origin || typeof origin !== 'string') {
        throw new InvalidSubjectIdentifierError(origin);
    }
    const permissions = (preserveExistingPermissions
        ? {
            ...this.getPermissions(origin),
        }
        : {});
    for (const [requestedTarget, approvedPermission] of Object.entries(approvedPermissions)) {
        if (!__classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_targetExists).call(this, requestedTarget)) {
            throw methodNotFound(requestedTarget);
        }
        if (approvedPermission.parentCapability !== undefined &&
            requestedTarget !== approvedPermission.parentCapability) {
            throw new InvalidApprovedPermissionError(origin, requestedTarget, approvedPermission);
        }
        // We have verified that the target exists, and reassign it to change its
        // type.
        const targetName = requestedTarget;
        const specification = __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_getPermissionSpecification).call(this, targetName);
        // The requested caveats are validated here.
        const caveats = __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_constructCaveats).call(this, origin, targetName, approvedPermission.caveats);
        const permissionOptions = {
            caveats,
            invoker: origin,
            target: targetName,
        };
        let permission;
        if (specification.factory) {
            permission = specification.factory(permissionOptions, requestData);
        }
        else {
            permission = constructPermission(permissionOptions);
        }
        if (mergePermissions) {
            permission = __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_mergePermission).call(this, permissions[targetName], permission)[0];
        }
        __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_validatePermission).call(this, specification, permission, origin, {
            invokePermissionValidator: true,
            performCaveatValidation: true,
        });
        permissions[targetName] = permission;
    }
    __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_setValidatedPermissions).call(this, origin, permissions);
    return permissions;
}, _PermissionController_validatePermission = function _PermissionController_validatePermission(specification, permission, origin, { invokePermissionValidator, performCaveatValidation, }) {
    const { allowedCaveats, validator, targetName } = specification;
    if (specification.subjectTypes?.length &&
        specification.subjectTypes.length > 0) {
        const metadata = this.messenger.call('SubjectMetadataController:getSubjectMetadata', origin);
        if (!metadata ||
            metadata.subjectType === null ||
            !specification.subjectTypes.includes(metadata.subjectType)) {
            throw specification.permissionType === PermissionType.RestrictedMethod
                ? methodNotFound(targetName, { origin })
                : new EndowmentPermissionDoesNotExistError(targetName, origin);
        }
    }
    if (hasProperty(permission, 'caveats')) {
        const { caveats } = permission;
        if (caveats !== null && !(Array.isArray(caveats) && caveats.length > 0)) {
            throw new InvalidCaveatsPropertyError(origin, targetName, caveats);
        }
        const seenCaveatTypes = new Set();
        caveats?.forEach((caveat) => {
            if (performCaveatValidation) {
                __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_validateCaveat).call(this, caveat, origin, targetName);
            }
            if (!allowedCaveats?.includes(caveat.type)) {
                throw new ForbiddenCaveatError(caveat.type, origin, targetName);
            }
            if (seenCaveatTypes.has(caveat.type)) {
                throw new DuplicateCaveatError(caveat.type, origin, targetName);
            }
            seenCaveatTypes.add(caveat.type);
        });
    }
    if (invokePermissionValidator && validator) {
        validator(permission, origin, targetName);
    }
}, _PermissionController_setValidatedPermissions = function _PermissionController_setValidatedPermissions(origin, permissions) {
    this.update((draftState) => {
        if (!draftState.subjects[origin]) {
            draftState.subjects[origin] = { origin, permissions: {} };
        }
        draftState.subjects[origin].permissions = castDraft(permissions);
    });
}, _PermissionController_constructCaveats = function _PermissionController_constructCaveats(origin, target, requestedCaveats) {
    const caveatArray = requestedCaveats?.map((requestedCaveat) => {
        __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_validateCaveat).call(this, requestedCaveat, origin, target);
        // Reassign so that we have a fresh object.
        const { type, value } = requestedCaveat;
        return { type, value };
    });
    return caveatArray && isNonEmptyArray(caveatArray)
        ? caveatArray
        : undefined;
}, _PermissionController_validateCaveat = function _PermissionController_validateCaveat(caveat, origin, target) {
    if (!isPlainObject(caveat)) {
        throw new InvalidCaveatError(caveat, origin, target);
    }
    if (Object.keys(caveat).length !== 2) {
        throw new InvalidCaveatFieldsError(caveat, origin, target);
    }
    if (typeof caveat.type !== 'string') {
        throw new InvalidCaveatTypeError(caveat, origin, target);
    }
    const specification = __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_getCaveatSpecification).call(this, caveat.type);
    if (!specification) {
        throw new UnrecognizedCaveatTypeError(caveat.type, origin, target);
    }
    if (!hasProperty(caveat, 'value') || caveat.value === undefined) {
        throw new CaveatMissingValueError(caveat, origin, target);
    }
    if (!isValidJson(caveat.value)) {
        throw new CaveatInvalidJsonError(caveat, origin, target);
    }
    // Typecast: TypeScript still believes that the caveat is a PlainObject.
    specification.validator?.(caveat, origin, target);
}, _PermissionController_validateRequestedPermissions = function _PermissionController_validateRequestedPermissions(origin, requestedPermissions) {
    if (!isPlainObject(requestedPermissions)) {
        throw invalidParams({
            message: `Requested permissions for origin "${origin}" is not a plain object.`,
            data: { origin, requestedPermissions },
        });
    }
    if (Object.keys(requestedPermissions).length === 0) {
        throw invalidParams({
            message: `Permissions request for origin "${origin}" contains no permissions.`,
            data: { requestedPermissions },
        });
    }
    for (const targetName of Object.keys(requestedPermissions)) {
        const permission = requestedPermissions[targetName];
        if (!__classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_targetExists).call(this, targetName)) {
            throw methodNotFound(targetName, { origin, requestedPermissions });
        }
        if (!isPlainObject(permission) ||
            (permission.parentCapability !== undefined &&
                targetName !== permission.parentCapability)) {
            throw invalidParams({
                message: `Permissions request for origin "${origin}" contains invalid requested permission(s).`,
                data: { origin, requestedPermissions },
            });
        }
        // Here we validate the permission without invoking its validator, if any.
        // The validator will be invoked after the permission has been approved.
        __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_validatePermission).call(this, __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_getPermissionSpecification).call(this, targetName), 
        // Typecast: The permission is still a "PlainObject" here.
        permission, origin, { invokePermissionValidator: false, performCaveatValidation: true });
    }
}, _PermissionController_mergeIncrementalPermissions = function _PermissionController_mergeIncrementalPermissions(existingPermissions, incrementalRequestedPermissions) {
    const permissionDiffMap = {};
    // Use immer's produce as a convenience for calculating the new permissions
    // without mutating the existing permissions or committing the results to state.
    const newPermissions = immerProduce(existingPermissions, (draftExistingPermissions) => {
        const leftPermissions = draftExistingPermissions;
        Object.entries(incrementalRequestedPermissions).forEach(([targetName, rightPermission]) => {
            const leftPermission = leftPermissions[targetName];
            const [newPermission, caveatsDiff] = __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_mergePermission).call(this, leftPermission ?? {}, rightPermission);
            if (leftPermission === undefined ||
                Object.keys(caveatsDiff).length > 0) {
                leftPermissions[targetName] = newPermission;
                permissionDiffMap[targetName] = caveatsDiff;
            }
            // Otherwise, leave the left permission as-is; its authority has
            // not changed.
        });
    });
    if (Object.keys(permissionDiffMap).length === 0) {
        return [];
    }
    return [newPermissions, permissionDiffMap];
}, _PermissionController_mergePermission = function _PermissionController_mergePermission(leftPermission, rightPermission) {
    const { caveatPairs, leftUniqueCaveats, rightUniqueCaveats } = collectUniqueAndPairedCaveats(leftPermission, rightPermission);
    const [mergedCaveats, caveatDiffMap] = caveatPairs.reduce(([caveats, diffMap], [leftCaveat, rightCaveat]) => {
        const [newCaveat, diff] = __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_mergeCaveat).call(this, leftCaveat, rightCaveat);
        if (newCaveat !== undefined && diff !== undefined) {
            caveats.push(newCaveat);
            diffMap[newCaveat.type] = diff;
        }
        else {
            caveats.push(leftCaveat);
        }
        return [caveats, diffMap];
    }, [[], {}]);
    const mergedRightUniqueCaveats = rightUniqueCaveats.map((caveat) => {
        const [newCaveat, diff] = __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_mergeCaveat).call(this, undefined, caveat);
        caveatDiffMap[newCaveat.type] = diff;
        return newCaveat;
    });
    const allCaveats = [
        ...mergedCaveats,
        ...leftUniqueCaveats,
        ...mergedRightUniqueCaveats,
    ];
    const newPermission = {
        ...leftPermission,
        ...rightPermission,
        ...(allCaveats.length > 0
            ? { caveats: allCaveats }
            : {}),
    };
    return [newPermission, caveatDiffMap];
}, _PermissionController_mergeCaveat = function _PermissionController_mergeCaveat(leftCaveat, rightCaveat) {
    /* istanbul ignore if: This should be impossible */
    if (leftCaveat !== undefined && leftCaveat.type !== rightCaveat.type) {
        throw new CaveatMergeTypeMismatchError(leftCaveat.type, rightCaveat.type);
    }
    const merger = __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_expectGetCaveatMerger).call(this, rightCaveat.type);
    if (leftCaveat === undefined) {
        return [
            {
                ...rightCaveat,
            },
            rightCaveat.value,
        ];
    }
    const [newValue, diff] = merger(leftCaveat.value, rightCaveat.value);
    return newValue !== undefined && diff !== undefined
        ? [
            {
                type: rightCaveat.type,
                value: newValue,
            },
            diff,
        ]
        : [];
}, _PermissionController_requestUserApproval = 
/**
 * Adds a request to the {@link ApprovalController} using the
 * {@link ApprovalControllerAddRequestAction} action. Also validates the resulting approved
 * permissions request, and throws an error if validation fails.
 *
 * @param permissionsRequest - The permissions request object.
 * @returns The approved permissions request object.
 */
async function _PermissionController_requestUserApproval(permissionsRequest) {
    const { origin, id } = permissionsRequest.metadata;
    const approvedRequest = await this.messenger.call('ApprovalController:addRequest', {
        id,
        origin,
        requestData: permissionsRequest,
        type: MethodNames.RequestPermissions,
    }, true);
    __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_validateApprovedPermissions).call(this, approvedRequest, { id, origin });
    return approvedRequest;
}, _PermissionController_handleApprovedPermissions = 
/**
 * Accepts a permissions request that has been approved by the user. This
 * method should be called after the user has approved the request and the
 * {@link ApprovalController} has resolved the user approval promise.
 *
 * @param options - Options bag.
 * @param options.subject - The subject to grant permissions to.
 * @param options.metadata - The metadata of the approved permissions request.
 * @param options.preserveExistingPermissions - Whether to preserve the
 * subject's existing permissions.
 * @param options.approvedRequest - The approved permissions request to handle.
 * @returns The granted permissions and request metadata.
 */
async function _PermissionController_handleApprovedPermissions({ subject, metadata, preserveExistingPermissions, approvedRequest, }) {
    const { permissions: approvedPermissions, ...requestData } = approvedRequest;
    const approvedMetadata = { ...metadata };
    const sideEffects = __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_getSideEffects).call(this, approvedPermissions);
    if (Object.values(sideEffects.permittedHandlers).length > 0) {
        const sideEffectsData = await __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_executeSideEffects).call(this, sideEffects, approvedRequest);
        approvedMetadata.data = Object.keys(sideEffects.permittedHandlers).reduce((acc, permission, i) => ({ [permission]: sideEffectsData[i], ...acc }), {});
    }
    return [
        this.grantPermissions({
            subject,
            approvedPermissions,
            preserveExistingPermissions,
            requestData,
        }),
        approvedMetadata,
    ];
}, _PermissionController_getSideEffects = function _PermissionController_getSideEffects(permissions) {
    return Object.keys(permissions).reduce((sideEffectList, targetName) => {
        if (__classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_targetExists).call(this, targetName)) {
            const specification = __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_getPermissionSpecification).call(this, targetName);
            if (specification.sideEffect) {
                sideEffectList.permittedHandlers[targetName] =
                    specification.sideEffect.onPermitted;
                if (specification.sideEffect.onFailure) {
                    sideEffectList.failureHandlers[targetName] =
                        specification.sideEffect.onFailure;
                }
            }
        }
        return sideEffectList;
    }, { permittedHandlers: {}, failureHandlers: {} });
}, _PermissionController_executeSideEffects = 
/**
 * Executes the side-effects of the approved permissions while handling the errors if any.
 * It will pass an instance of the {@link messenger} and the request data associated with the permission request to the handlers through its params.
 *
 * @param sideEffects - the side-effect record created by {@link #getSideEffects}
 * @param requestData - the permissions requestData.
 * @returns the value returned by all the `onPermitted` handlers in an array.
 */
async function _PermissionController_executeSideEffects(sideEffects, requestData) {
    const { permittedHandlers, failureHandlers } = sideEffects;
    const params = {
        requestData,
        messenger: this.messenger,
    };
    const promiseResults = await Promise.allSettled(Object.values(permittedHandlers).map((permittedHandler) => permittedHandler(params)));
    // lib.es2020.promise.d.ts does not export its types so we're using a simple type.
    const rejectedHandlers = promiseResults.filter((promise) => promise.status === 'rejected');
    if (rejectedHandlers.length > 0) {
        const failureHandlersList = Object.values(failureHandlers);
        if (failureHandlersList.length > 0) {
            try {
                await Promise.all(failureHandlersList.map((failureHandler) => failureHandler(params)));
            }
            catch (error) {
                throw internalError('Unexpected error in side-effects', { error });
            }
        }
        const reasons = rejectedHandlers.map((handler) => handler.reason);
        reasons.forEach((reason) => {
            console.error(reason);
        });
        throw reasons.length > 1
            ? internalError('Multiple errors occurred during side-effects execution', { errors: reasons })
            : reasons[0];
    }
    // lib.es2020.promise.d.ts does not export its types so we're using a simple type.
    return promiseResults.map(({ value }) => value);
}, _PermissionController_validateApprovedPermissions = function _PermissionController_validateApprovedPermissions(approvedRequest, originalMetadata) {
    const { id, origin } = originalMetadata;
    if (!isPlainObject(approvedRequest) ||
        !isPlainObject(approvedRequest.metadata)) {
        throw internalError(`Approved permissions request for subject "${origin}" is invalid.`, { data: { approvedRequest } });
    }
    const { metadata: { id: newId, origin: newOrigin }, permissions, } = approvedRequest;
    if (newId !== id) {
        throw internalError(`Approved permissions request for subject "${origin}" mutated its id.`, { originalId: id, mutatedId: newId });
    }
    if (newOrigin !== origin) {
        throw internalError(`Approved permissions request for subject "${origin}" mutated its origin.`, { originalOrigin: origin, mutatedOrigin: newOrigin });
    }
    try {
        __classPrivateFieldGet(this, _PermissionController_instances, "m", _PermissionController_validateRequestedPermissions).call(this, origin, permissions);
    }
    catch (error) {
        if (error instanceof Error) {
            // Re-throw as an internal error; we should never receive invalid approved
            // permissions.
            throw internalError(`Invalid approved permissions request: ${error.message}`, error instanceof JsonRpcError ? error.data : undefined);
        }
        /* istanbul ignore next: This should be impossible */
        throw internalError('Unrecognized error type', { error });
    }
}, _PermissionController_hasApprovalRequest = function _PermissionController_hasApprovalRequest(options) {
    return this.messenger.call('ApprovalController:hasRequest', options);
}, _PermissionController_rejectPermissionsRequest = function _PermissionController_rejectPermissionsRequest(id, error) {
    return this.messenger.call('ApprovalController:rejectRequest', id, error);
}, _PermissionController_executeRestrictedMethod = function _PermissionController_executeRestrictedMethod(methodImplementation, subject, method, params = []) {
    const { origin } = subject;
    const permission = this.getPermission(origin, method);
    if (!permission) {
        throw unauthorized({ data: { origin, method } });
    }
    return decorateWithCaveats(methodImplementation, permission, __classPrivateFieldGet(this, _PermissionController_caveatSpecifications, "f"))({ method, params, context: { origin } });
};
