// @metamask/permission-controller/dist/SubjectMetadataController.cjs
"use strict";
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var _a, _SubjectMetadataController_subjectCacheLimit, _SubjectMetadataController_subjectsWithoutPermissionsEncounteredSinceStartup, _SubjectMetadataController_subjectHasPermissions, _SubjectMetadataController_getTrimmedState;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SubjectMetadataController = exports.SubjectType = void 0;
const base_controller_1 = require("@metamask/base-controller");
const controllerName = 'SubjectMetadataController';
const MESSENGER_EXPOSED_METHODS = [
    'clearState',
    'addSubjectMetadata',
    'getSubjectMetadata',
    'trimMetadataState',
];
/**
 * The different kinds of subjects that MetaMask may interact with, including
 * third parties and itself (e.g., when the background communicated with the UI).
 */
var SubjectType;
(function (SubjectType) {
    SubjectType["Extension"] = "extension";
    SubjectType["Internal"] = "internal";
    SubjectType["Unknown"] = "unknown";
    SubjectType["Website"] = "website";
    SubjectType["Snap"] = "snap";
})(SubjectType || (exports.SubjectType = SubjectType = {}));
const stateMetadata = {
    subjectMetadata: {
        includeInStateLogs: true,
        persist: true,
        includeInDebugSnapshot: false,
        usedInUi: true,
    },
};
const defaultState = {
    subjectMetadata: {},
};
/**
 * A controller for storing metadata associated with permission subjects. More
 * or less, a cache.
 */
class SubjectMetadataController extends base_controller_1.BaseController {
    constructor({ messenger, subjectCacheLimit, state = {}, }) {
        if (!Number.isInteger(subjectCacheLimit) || subjectCacheLimit < 1) {
            throw new Error(`subjectCacheLimit must be a positive integer. Received: "${subjectCacheLimit}"`);
        }
        const hasPermissions = (origin) => {
            return messenger.call('PermissionController:hasPermissions', origin);
        };
        super({
            name: controllerName,
            metadata: stateMetadata,
            messenger,
            state: {
                ...__classPrivateFieldGet(_a, _a, "m", _SubjectMetadataController_getTrimmedState).call(_a, state, hasPermissions),
            },
        });
        _SubjectMetadataController_subjectCacheLimit.set(this, void 0);
        _SubjectMetadataController_subjectsWithoutPermissionsEncounteredSinceStartup.set(this, void 0);
        _SubjectMetadataController_subjectHasPermissions.set(this, void 0);
        __classPrivateFieldSet(this, _SubjectMetadataController_subjectHasPermissions, hasPermissions, "f");
        __classPrivateFieldSet(this, _SubjectMetadataController_subjectCacheLimit, subjectCacheLimit, "f");
        __classPrivateFieldSet(this, _SubjectMetadataController_subjectsWithoutPermissionsEncounteredSinceStartup, new Set(), "f");
        this.messenger.registerMethodActionHandlers(this, MESSENGER_EXPOSED_METHODS);
    }
    /**
     * Clears the state of this controller. Also resets the cache of subjects
     * encountered since startup, so as to not prematurely reach the cache limit.
     */
    clearState() {
        __classPrivateFieldGet(this, _SubjectMetadataController_subjectsWithoutPermissionsEncounteredSinceStartup, "f").clear();
        this.update((_draftState) => {
            return { ...defaultState };
        });
    }
    /**
     * Stores domain metadata for the given origin (subject). Deletes metadata for
     * subjects without permissions in a FIFO manner once more than
     * {@link SubjectMetadataController.subjectCacheLimit} distinct origins have
     * been added since boot.
     *
     * In order to prevent a degraded user experience,
     * metadata is never deleted for subjects with permissions, since metadata
     * cannot yet be requested on demand.
     *
     * @param metadata - The subject metadata to store.
     */
    addSubjectMetadata(metadata) {
        const { origin } = metadata;
        const newMetadata = {
            ...metadata,
            extensionId: metadata.extensionId ?? null,
            iconUrl: metadata.iconUrl ?? null,
            name: metadata.name ?? null,
            subjectType: metadata.subjectType ?? null,
        };
        let originToForget = null;
        // We only delete the oldest encountered subject from the cache, again to
        // ensure that the user's experience isn't degraded by missing icons etc.
        if (__classPrivateFieldGet(this, _SubjectMetadataController_subjectsWithoutPermissionsEncounteredSinceStartup, "f").size >=
            __classPrivateFieldGet(this, _SubjectMetadataController_subjectCacheLimit, "f")) {
            const cachedOrigin = __classPrivateFieldGet(this, _SubjectMetadataController_subjectsWithoutPermissionsEncounteredSinceStartup, "f")
                .values()
                .next().value;
            __classPrivateFieldGet(this, _SubjectMetadataController_subjectsWithoutPermissionsEncounteredSinceStartup, "f").delete(cachedOrigin);
            if (!__classPrivateFieldGet(this, _SubjectMetadataController_subjectHasPermissions, "f").call(this, cachedOrigin)) {
                originToForget = cachedOrigin;
            }
        }
        __classPrivateFieldGet(this, _SubjectMetadataController_subjectsWithoutPermissionsEncounteredSinceStartup, "f").add(origin);
        this.update((draftState) => {
            // @ts-expect-error TS2589: Type instantiation is excessively deep and possibly infinite
            draftState.subjectMetadata[origin] = newMetadata;
            if (typeof originToForget === 'string') {
                delete draftState.subjectMetadata[originToForget];
            }
        });
    }
    /**
     * Gets the subject metadata for the given origin, if any.
     *
     * @param origin - The origin for which to get the subject metadata.
     * @returns The subject metadata, if any, or `undefined` otherwise.
     */
    getSubjectMetadata(origin) {
        return this.state.subjectMetadata[origin];
    }
    /**
     * Deletes all subjects without permissions from the controller's state.
     */
    trimMetadataState() {
        this.update(() => {
            return __classPrivateFieldGet(_a, _a, "m", _SubjectMetadataController_getTrimmedState).call(_a, this.state, __classPrivateFieldGet(this, _SubjectMetadataController_subjectHasPermissions, "f"));
        });
    }
}
exports.SubjectMetadataController = SubjectMetadataController;
_a = SubjectMetadataController, _SubjectMetadataController_subjectCacheLimit = new WeakMap(), _SubjectMetadataController_subjectsWithoutPermissionsEncounteredSinceStartup = new WeakMap(), _SubjectMetadataController_subjectHasPermissions = new WeakMap(), _SubjectMetadataController_getTrimmedState = function _SubjectMetadataController_getTrimmedState(state, hasPermissions) {
    const { subjectMetadata = {} } = state;
    return {
        subjectMetadata: Object.keys(subjectMetadata).reduce((newSubjectMetadata, origin) => {
            if (hasPermissions(origin)) {
                newSubjectMetadata[origin] = subjectMetadata[origin];
            }
            return newSubjectMetadata;
        }, {}),
    };
};
