// @metamask/messenger/dist/generate-action-types/parse-source.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.findSourcesWithExposedMethods = exports.parseSourceFile = void 0;
const utils_1 = require("@metamask/utils");
const fs = __importStar(require("node:fs"));
const path = __importStar(require("node:path"));
const typescript_1 = require("typescript");
/**
 * Extracts JSDoc comment from a method declaration.
 *
 * @param node - The method declaration node.
 * @param source - The source file.
 * @returns The JSDoc comment.
 */
function extractJSDoc(node, source) {
    const jsDocTags = (0, typescript_1.getJSDocCommentsAndTags)(node);
    if (jsDocTags.length === 0) {
        return '';
    }
    const jsDoc = jsDocTags[0];
    if ((0, typescript_1.isJSDoc)(jsDoc)) {
        const fullText = source.getFullText();
        const start = jsDoc.getFullStart();
        const end = jsDoc.getEnd();
        const rawJsDoc = fullText.substring(start, end).trim();
        return formatJSDoc(rawJsDoc);
    }
    // istanbul ignore next: defensive check — getJSDocCommentsAndTags always returns JSDoc nodes
    return '';
}
/**
 * Formats JSDoc comments to have consistent indentation for the generated file.
 *
 * @param rawJsDoc - The raw JSDoc comment from the source.
 * @returns The formatted JSDoc comment.
 */
function formatJSDoc(rawJsDoc) {
    const lines = rawJsDoc.split('\n');
    const formattedLines = [];
    for (let i = 0; i < lines.length; i++) {
        const line = lines[i];
        if (i === 0) {
            formattedLines.push('/**');
        }
        else if (i === lines.length - 1) {
            formattedLines.push(' */');
        }
        else {
            const trimmed = line.trim();
            if (trimmed.startsWith('*')) {
                const content = trimmed.substring(1).trim();
                formattedLines.push(content ? ` * ${content}` : ' *');
            }
            else {
                formattedLines.push(trimmed ? ` * ${trimmed}` : ' *');
            }
        }
    }
    return formattedLines.join('\n');
}
/**
 * Visits AST nodes to find exposed methods and controller/service class.
 *
 * @param context - The visitor context.
 * @returns A function to visit nodes.
 */
function createASTVisitor(context) {
    function visitNode(node) {
        if ((0, typescript_1.isVariableStatement)(node)) {
            const declaration = node.declarationList.declarations[0];
            if ((0, typescript_1.isIdentifier)(declaration.name) &&
                declaration.name.text === 'MESSENGER_EXPOSED_METHODS') {
                if (declaration.initializer) {
                    let arrayExpression;
                    if ((0, typescript_1.isArrayLiteralExpression)(declaration.initializer)) {
                        arrayExpression = declaration.initializer;
                    }
                    else if ((0, typescript_1.isAsExpression)(declaration.initializer) &&
                        (0, typescript_1.isArrayLiteralExpression)(declaration.initializer.expression)) {
                        arrayExpression = declaration.initializer.expression;
                    }
                    if (arrayExpression) {
                        context.exposedMethods = arrayExpression.elements
                            .filter(typescript_1.isStringLiteral)
                            .map((element) => element.text);
                    }
                }
            }
        }
        if ((0, typescript_1.isClassDeclaration)(node) && node.name) {
            const classText = node.name.text;
            if (classText.includes('Controller') || classText.includes('Service')) {
                context.className = classText;
                const seenMethods = new Set();
                for (const member of node.members) {
                    if ((0, typescript_1.isMethodDeclaration)(member) &&
                        member.name &&
                        (0, typescript_1.isIdentifier)(member.name)) {
                        const methodName = member.name.text;
                        if (context.exposedMethods.includes(methodName) &&
                            !seenMethods.has(methodName)) {
                            seenMethods.add(methodName);
                            const jsDoc = extractJSDoc(member, context.sourceFile);
                            context.methods.push({
                                name: methodName,
                                jsDoc,
                            });
                        }
                    }
                }
            }
        }
        (0, typescript_1.forEachChild)(node, visitNode);
    }
    return visitNode;
}
/**
 * Create a TypeScript program for the given file by locating the nearest
 * tsconfig.json.
 *
 * @param filePath - Absolute path to the source file.
 * @returns A TypeScript program, or null if no tsconfig was found.
 */
function createProgramForFile(filePath) {
    const configPath = (0, typescript_1.findConfigFile)(path.dirname(filePath), typescript_1.sys.fileExists.bind(typescript_1.sys), 'tsconfig.json');
    if (!configPath) {
        return null;
    }
    const { config, error } = (0, typescript_1.readConfigFile)(configPath, typescript_1.sys.readFile.bind(typescript_1.sys));
    if (error) {
        return null;
    }
    const parsedConfig = (0, typescript_1.parseJsonConfigFileContent)(config, typescript_1.sys, path.dirname(configPath));
    return (0, typescript_1.createProgram)({
        rootNames: parsedConfig.fileNames,
        options: parsedConfig.options,
    });
}
/**
 * Find a class declaration with the given name in a source file.
 *
 * @param source - The source file to search.
 * @param className - The class name to look for.
 * @returns The class declaration node, or null if not found.
 */
function findClassInSourceFile(source, className) {
    return (source.statements.find((node) => (0, typescript_1.isClassDeclaration)(node) && node.name?.text === className) ?? // istanbul ignore next: class is always found when called from parseSourceFile
        null);
}
/**
 * Search through the class hierarchy of a TypeScript type to find the
 * declaration of a method with the given name.
 *
 * @param classType - The class type to search.
 * @param methodName - The method name to look for.
 * @returns The method declaration node, or null if not found.
 */
function findMethodInHierarchy(classType, methodName) {
    const symbol = classType.getProperty(methodName);
    if (!symbol) {
        return null;
    }
    const declarations = symbol.getDeclarations();
    // istanbul ignore next: defensive check — symbols from getProperty always have declarations
    if (!declarations) {
        return null;
    }
    for (const declaration of declarations) {
        if ((0, typescript_1.isMethodDeclaration)(declaration)) {
            return declaration;
        }
    }
    // istanbul ignore next: defensive fallback — property found but not a method declaration
    return null;
}
/**
 * Check if a path is a directory.
 *
 * @param pathValue - The path to check.
 * @returns True if the path is a directory, false otherwise.
 */
async function isDirectory(pathValue) {
    try {
        const stats = await fs.promises.stat(pathValue);
        return stats.isDirectory();
    }
    catch (error) {
        if ((0, utils_1.isObject)(error) &&
            (0, utils_1.hasProperty)(error, 'code') &&
            error.code === 'ENOENT') {
            return false;
        }
        throw error;
    }
}
/**
 * Parses a source file to extract exposed methods and their metadata.
 *
 * @param filePath - Path to the controller/service file to parse.
 * @returns Source information or null if parsing fails.
 */
async function parseSourceFile(filePath) {
    try {
        const content = await fs.promises.readFile(filePath, 'utf8');
        const source = (0, typescript_1.createSourceFile)(filePath, content, typescript_1.ScriptTarget.Latest, true);
        const context = {
            exposedMethods: [],
            className: '',
            methods: [],
            sourceFile: source,
        };
        createASTVisitor(context)(source);
        if (context.exposedMethods.length === 0 || !context.className) {
            return null;
        }
        const foundMethodNames = new Set(context.methods.map((method) => method.name));
        const inheritedMethodNames = context.exposedMethods.filter((name) => !foundMethodNames.has(name));
        if (inheritedMethodNames.length > 0) {
            const program = createProgramForFile(filePath);
            const checker = program?.getTypeChecker();
            const programSourceFile = program?.getSourceFile(filePath);
            (0, utils_1.assert)(checker, `Type checker could not be created for "${filePath}". Ensure a valid tsconfig.json is present.`);
            (0, utils_1.assert)(programSourceFile, `Source file "${filePath}" not found in program.`);
            const classNode = findClassInSourceFile(programSourceFile, context.className);
            (0, utils_1.assert)(classNode, `Class "${context.className}" not found in "${filePath}".`);
            const classType = checker.getTypeAtLocation(classNode);
            for (const methodName of inheritedMethodNames) {
                const methodDeclaration = findMethodInHierarchy(classType, methodName);
                const jsDoc = methodDeclaration
                    ? extractJSDoc(methodDeclaration, methodDeclaration.getSourceFile())
                    : '';
                context.methods.push({ name: methodName, jsDoc });
            }
        }
        return {
            name: context.className,
            filePath,
            methods: context.methods,
        };
    }
    catch (error) {
        console.error(`Error parsing ${filePath}:`, error);
        return null;
    }
}
exports.parseSourceFile = parseSourceFile;
/**
 * Recursively get all files in a directory and its subdirectories.
 *
 * @param directory - The directory to search.
 * @returns An array of file paths.
 */
const EXCLUDED_DIRECTORIES = new Set([
    'node_modules',
    'dist',
    '.git',
    'coverage',
]);
async function getFiles(directory) {
    const entries = await fs.promises.readdir(directory, { withFileTypes: true });
    const files = await Promise.all(entries.map(async (entry) => {
        const fullPath = path.join(directory, entry.name);
        if (entry.isDirectory()) {
            return EXCLUDED_DIRECTORIES.has(entry.name)
                ? []
                : await getFiles(fullPath);
        }
        return fullPath;
    }));
    return files.flat();
}
/**
 * Finds all source files that have MESSENGER_EXPOSED_METHODS constants.
 * Searches recursively through subdirectories.
 *
 * @param sourcePath - Path to the folder where controllers/services are located.
 * @returns A list of source information objects.
 */
async function findSourcesWithExposedMethods(sourcePath) {
    const srcPath = path.resolve(globalThis.process.cwd(), sourcePath);
    const sources = [];
    if (!(await isDirectory(srcPath))) {
        throw new Error(`The specified path is not a directory: ${srcPath}`);
    }
    const srcFiles = await getFiles(srcPath);
    for (const file of srcFiles) {
        if (!file.endsWith('.ts') || file.endsWith('.test.ts')) {
            continue;
        }
        const content = await fs.promises.readFile(file, 'utf8');
        if (content.includes('MESSENGER_EXPOSED_METHODS')) {
            const sourceInfo = await parseSourceFile(file);
            if (sourceInfo) {
                sources.push(sourceInfo);
            }
        }
    }
    return sources;
}
exports.findSourcesWithExposedMethods = findSourcesWithExposedMethods;
